
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="..">
      
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.5.2, mkdocs-material-9.1.21">
    
    
      
        <title>BPP Documentation - BPP Bayesian Phylogenetics & Phylogeography</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.eebd395e.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.ecc896b0.min.css">
      
      

    
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="amber" data-md-color-accent="indigo">
  
    
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#bpp-documentation" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="BPP Bayesian Phylogenetics &amp; Phylogeography" class="md-header__button md-logo" aria-label="BPP Bayesian Phylogenetics & Phylogeography" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            BPP Bayesian Phylogenetics & Phylogeography
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              BPP Documentation
            
          </span>
        </div>
      </div>
    </div>
    
      
    
    
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="BPP Bayesian Phylogenetics &amp; Phylogeography" class="md-nav__button md-logo" aria-label="BPP Bayesian Phylogenetics & Phylogeography" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    BPP Bayesian Phylogenetics & Phylogeography
  </label>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        Quick Start
      </a>
    </li>
  

    
      
      
      

  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          BPP Documentation
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        BPP Documentation
      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#bpp-explained" class="md-nav__link">
    BPP Explained
  </a>
  
    <nav class="md-nav" aria-label="BPP Explained">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#overview" class="md-nav__link">
    Overview
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#possible-types-of-analyses" class="md-nav__link">
    Possible types of analyses
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#model-parameters" class="md-nav__link">
    Model Parameters
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#model-assumptions" class="md-nav__link">
    Model Assumptions
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#nuclear-genomic-data" class="md-nav__link">
    Nuclear genomic data
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#mitochondrial-data" class="md-nav__link">
    Mitochondrial data
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#identical-sequences-and-phylogenetic-signal" class="md-nav__link">
    Identical sequences and phylogenetic signal
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#how-to-cite-bpp" class="md-nav__link">
    How to Cite BPP
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#installing-and-running-bpp" class="md-nav__link">
    Installing and Running BPP
  </a>
  
    <nav class="md-nav" aria-label="Installing and Running BPP">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#obtaining-bpp" class="md-nav__link">
    Obtaining BPP
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#using-the-command-line" class="md-nav__link">
    Using the Command Line
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#compiling-the-bpp-program" class="md-nav__link">
    Compiling the BPP Program
  </a>
  
    <nav class="md-nav" aria-label="Compiling the BPP Program">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#requirements" class="md-nav__link">
    Requirements
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#downloading-and-compiling-the-source-code" class="md-nav__link">
    Downloading and compiling the source code
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#bpp-trial-run" class="md-nav__link">
    BPP Trial Run
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#input-file-formats" class="md-nav__link">
    Input File Formats
  </a>
  
    <nav class="md-nav" aria-label="Input File Formats">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#sequence-file" class="md-nav__link">
    Sequence File
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#imap-file" class="md-nav__link">
    Imap File
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#outgroups-and-constraints-file" class="md-nav__link">
    Outgroups and Constraints File
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#control-file" class="md-nav__link">
    Control File
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#bpp-control-file-variables" class="md-nav__link">
    BPP control file variables
  </a>
  
    <nav class="md-nav" aria-label="BPP control file variables">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1-seed" class="md-nav__link">
    1 seed
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2-usedata" class="md-nav__link">
    2 usedata
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#3-outfile" class="md-nav__link">
    3 outfile
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#4-mcmcfile" class="md-nav__link">
    4 mcmcfile
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#5-seqfile" class="md-nav__link">
    5 seqfile
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#6-finetune" class="md-nav__link">
    6 finetune
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#7-print" class="md-nav__link">
    7 print
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#8-burnin" class="md-nav__link">
    8 burnin
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#9-sampfreq" class="md-nav__link">
    9 sampfreq
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#10-nsample" class="md-nav__link">
    10 nsample
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#11-speciestree" class="md-nav__link">
    11 species&amp;tree
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#12-imapfile" class="md-nav__link">
    12 Imapfile
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#13-speciesdelimitation" class="md-nav__link">
    13 speciesdelimitation
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#14-speciestree" class="md-nav__link">
    14 speciestree
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#15-speciesmodelprior" class="md-nav__link">
    15 speciesmodelprior
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#16-phase" class="md-nav__link">
    16 phase
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#17-nloci" class="md-nav__link">
    17 nloci
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#18-model" class="md-nav__link">
    18 model
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#19-qrates" class="md-nav__link">
    19 Qrates
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#20-basefreqs" class="md-nav__link">
    20 basefreqs
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#21-alphaprior" class="md-nav__link">
    21 alphaprior
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#22-cleandata" class="md-nav__link">
    22 cleandata
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#23-thetaprior" class="md-nav__link">
    23 thetaprior
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#24-tauprior" class="md-nav__link">
    24 tauprior
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#25-phiprior" class="md-nav__link">
    25 phiprior
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#26-locusrate" class="md-nav__link">
    26 locusrate
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#27-clock" class="md-nav__link">
    27 clock
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#28-heredity" class="md-nav__link">
    28 heredity
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#29-checkpoint" class="md-nav__link">
    29 checkpoint
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#30-constraintfile" class="md-nav__link">
    30 constraintfile
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#31-threads" class="md-nav__link">
    31 threads
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#example-control-file" class="md-nav__link">
    Example control file
  </a>
  
    <nav class="md-nav" aria-label="Example control file">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#control-file-variables-explained" class="md-nav__link">
    Control file variables explained
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#substitution-models" class="md-nav__link">
    Substitution Models
  </a>
  
    <nav class="md-nav" aria-label="Substitution Models">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#dna-substitution-models" class="md-nav__link">
    DNA Substitution Models
  </a>
  
    <nav class="md-nav" aria-label="DNA Substitution Models">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#jukes-cantor-model" class="md-nav__link">
    Jukes-Cantor model
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gtr-model" class="md-nav__link">
    GTR model
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#other-dna-substitution-models" class="md-nav__link">
    Other DNA substitution models
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#amino-acid-substitution-models" class="md-nav__link">
    Amino acid substitution models
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#custom-among-locus-model-variation" class="md-nav__link">
    Custom among-locus model variation
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#models-of-substitution-rate-variation" class="md-nav__link">
    Models of Substitution Rate Variation
  </a>
  
    <nav class="md-nav" aria-label="Models of Substitution Rate Variation">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#among-site-rate-variation" class="md-nav__link">
    Among-site rate variation
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#among-locus-rate-variation" class="md-nav__link">
    Among-locus rate variation
  </a>
  
    <nav class="md-nav" aria-label="Among-locus rate variation">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#prior-distribution-of-the-mean-rate-among-loci" class="md-nav__link">
    Prior distribution of the mean rate among loci
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#prior-distribution-for-locus-specific-rates-given-mean-rate-among-loci" class="md-nav__link">
    Prior distribution for locus-specific rates given mean rate among loci
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#control-file-settings-for-locusrate" class="md-nav__link">
    Control file settings for locusrate
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#among-species-rate-variation" class="md-nav__link">
    Among-species rate variation
  </a>
  
    <nav class="md-nav" aria-label="Among-species rate variation">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#variation-in-the-degree-of-departure-from-a-molecular-clock-among-loci" class="md-nav__link">
    Variation in the degree of departure from a molecular clock among loci
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#relaxed-clock-with-iid-rates-among-branches" class="md-nav__link">
    Relaxed clock with i.i.d rates among branches
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#relaxed-clock-with-autocorrelated-rates-among-branches" class="md-nav__link">
    Relaxed clock with autocorrelated rates among branches
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#control-file-settings-for-clock" class="md-nav__link">
    Control file settings for clock
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#introgression-and-migration-models" class="md-nav__link">
    Introgression and Migration Models
  </a>
  
    <nav class="md-nav" aria-label="Introgression and Migration Models">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#the-msc-i-model" class="md-nav__link">
    The MSC-I Model
  </a>
  
    <nav class="md-nav" aria-label="The MSC-I Model">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#autogeneration-of-extended-newick-graphs" class="md-nav__link">
    Autogeneration of extended newick graphs
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#examples-using-msci-create" class="md-nav__link">
    Examples using --msci-create
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#the-msc-m-model" class="md-nav__link">
    The MSC-M Model
  </a>
  
    <nav class="md-nav" aria-label="The MSC-M Model">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#control-file-specification-of-msc-m" class="md-nav__link">
    Control file specification of MSC-M
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#prior-on-migration-rates" class="md-nav__link">
    Prior on migration rates
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#variable-migration-rates-among-loci" class="md-nav__link">
    Variable migration rates among loci
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#priors-and-pseudo-priors" class="md-nav__link">
    Priors and pseudo-priors
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#odds-and-ends" class="md-nav__link">
    Odds and Ends
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#methods-of-analysis" class="md-nav__link">
    Methods of Analysis
  </a>
  
    <nav class="md-nav" aria-label="Methods of Analysis">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#a00-demography-and-divergence-time-estimation" class="md-nav__link">
    A00: Demography and Divergence Time Estimation
  </a>
  
    <nav class="md-nav" aria-label="A00: Demography and Divergence Time Estimation">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#input-a00-single-population" class="md-nav__link">
    Input A00 (single population)
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#output-a00-single-population" class="md-nav__link">
    Output A00 (single population)
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#input-a00-multiple-populations" class="md-nav__link">
    Input A00 (multiple populations)
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#output-a00-multiple-populations" class="md-nav__link">
    Output A00 (multiple populations)
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#mcmc-output-file" class="md-nav__link">
    MCMC output file
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#adjusting-step-lengths-for-mcmc-moves-finetune" class="md-nav__link">
    Adjusting step lengths for MCMC moves (finetune)
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#a01-species-tree-estimation" class="md-nav__link">
    A01: Species Tree Estimation
  </a>
  
    <nav class="md-nav" aria-label="A01: Species Tree Estimation">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#input-a01" class="md-nav__link">
    Input A01
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#output-a01" class="md-nav__link">
    Output A01
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#mcmc-output-file_1" class="md-nav__link">
    MCMC output file
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#a10-species-delimitation" class="md-nav__link">
    A10: Species Delimitation
  </a>
  
    <nav class="md-nav" aria-label="A10: Species Delimitation">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#input-a10" class="md-nav__link">
    Input A10
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#output-a10" class="md-nav__link">
    Output A10
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#mcmc-output-file_2" class="md-nav__link">
    MCMC output file
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#a11-joint-species-delimitation-and-species-tree-estimation" class="md-nav__link">
    A11: Joint Species Delimitation and Species Tree Estimation
  </a>
  
    <nav class="md-nav" aria-label="A11: Joint Species Delimitation and Species Tree Estimation">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#input-a11" class="md-nav__link">
    Input A11
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#output-a11" class="md-nav__link">
    Output A11
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#mcmc-output-file_3" class="md-nav__link">
    MCMC output file
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#advanced-features-of-bpp" class="md-nav__link">
    Advanced Features of BPP
  </a>
  
    <nav class="md-nav" aria-label="Advanced Features of BPP">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#threading-and-checkpointing" class="md-nav__link">
    Threading and Checkpointing
  </a>
  
    <nav class="md-nav" aria-label="Threading and Checkpointing">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#threads" class="md-nav__link">
    Threads
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#checkpointing" class="md-nav__link">
    Checkpointing
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#marginal-likelihood-calculations" class="md-nav__link">
    Marginal Likelihood Calculations
  </a>
  
    <nav class="md-nav" aria-label="Marginal Likelihood Calculations">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#compiling-and-running-bfdriver" class="md-nav__link">
    Compiling and running BFdriver
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#running-the-program" class="md-nav__link">
    Running the program
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#exercise-and-results" class="md-nav__link">
    Exercise and results
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#common-errors-and-problems" class="md-nav__link">
    Common errors and problems
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#simulation-using-bpp" class="md-nav__link">
    Simulation Using BPP
  </a>
  
    <nav class="md-nav" aria-label="Simulation Using BPP">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#b1-control-file-options-for-simulating-without-migration" class="md-nav__link">
    B1 Control-file options for simulating without migration
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#b2-simulating-with-migration-under-the-im-model" class="md-nav__link">
    B2 Simulating with migration (under the IM model)
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#features-new-to-bpp-40" class="md-nav__link">
    Features New To BPP 4.0
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#bpp-explained" class="md-nav__link">
    BPP Explained
  </a>
  
    <nav class="md-nav" aria-label="BPP Explained">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#overview" class="md-nav__link">
    Overview
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#possible-types-of-analyses" class="md-nav__link">
    Possible types of analyses
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#model-parameters" class="md-nav__link">
    Model Parameters
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#model-assumptions" class="md-nav__link">
    Model Assumptions
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#nuclear-genomic-data" class="md-nav__link">
    Nuclear genomic data
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#mitochondrial-data" class="md-nav__link">
    Mitochondrial data
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#identical-sequences-and-phylogenetic-signal" class="md-nav__link">
    Identical sequences and phylogenetic signal
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#how-to-cite-bpp" class="md-nav__link">
    How to Cite BPP
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#installing-and-running-bpp" class="md-nav__link">
    Installing and Running BPP
  </a>
  
    <nav class="md-nav" aria-label="Installing and Running BPP">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#obtaining-bpp" class="md-nav__link">
    Obtaining BPP
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#using-the-command-line" class="md-nav__link">
    Using the Command Line
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#compiling-the-bpp-program" class="md-nav__link">
    Compiling the BPP Program
  </a>
  
    <nav class="md-nav" aria-label="Compiling the BPP Program">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#requirements" class="md-nav__link">
    Requirements
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#downloading-and-compiling-the-source-code" class="md-nav__link">
    Downloading and compiling the source code
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#bpp-trial-run" class="md-nav__link">
    BPP Trial Run
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#input-file-formats" class="md-nav__link">
    Input File Formats
  </a>
  
    <nav class="md-nav" aria-label="Input File Formats">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#sequence-file" class="md-nav__link">
    Sequence File
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#imap-file" class="md-nav__link">
    Imap File
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#outgroups-and-constraints-file" class="md-nav__link">
    Outgroups and Constraints File
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#control-file" class="md-nav__link">
    Control File
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#bpp-control-file-variables" class="md-nav__link">
    BPP control file variables
  </a>
  
    <nav class="md-nav" aria-label="BPP control file variables">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1-seed" class="md-nav__link">
    1 seed
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2-usedata" class="md-nav__link">
    2 usedata
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#3-outfile" class="md-nav__link">
    3 outfile
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#4-mcmcfile" class="md-nav__link">
    4 mcmcfile
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#5-seqfile" class="md-nav__link">
    5 seqfile
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#6-finetune" class="md-nav__link">
    6 finetune
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#7-print" class="md-nav__link">
    7 print
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#8-burnin" class="md-nav__link">
    8 burnin
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#9-sampfreq" class="md-nav__link">
    9 sampfreq
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#10-nsample" class="md-nav__link">
    10 nsample
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#11-speciestree" class="md-nav__link">
    11 species&amp;tree
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#12-imapfile" class="md-nav__link">
    12 Imapfile
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#13-speciesdelimitation" class="md-nav__link">
    13 speciesdelimitation
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#14-speciestree" class="md-nav__link">
    14 speciestree
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#15-speciesmodelprior" class="md-nav__link">
    15 speciesmodelprior
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#16-phase" class="md-nav__link">
    16 phase
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#17-nloci" class="md-nav__link">
    17 nloci
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#18-model" class="md-nav__link">
    18 model
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#19-qrates" class="md-nav__link">
    19 Qrates
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#20-basefreqs" class="md-nav__link">
    20 basefreqs
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#21-alphaprior" class="md-nav__link">
    21 alphaprior
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#22-cleandata" class="md-nav__link">
    22 cleandata
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#23-thetaprior" class="md-nav__link">
    23 thetaprior
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#24-tauprior" class="md-nav__link">
    24 tauprior
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#25-phiprior" class="md-nav__link">
    25 phiprior
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#26-locusrate" class="md-nav__link">
    26 locusrate
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#27-clock" class="md-nav__link">
    27 clock
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#28-heredity" class="md-nav__link">
    28 heredity
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#29-checkpoint" class="md-nav__link">
    29 checkpoint
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#30-constraintfile" class="md-nav__link">
    30 constraintfile
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#31-threads" class="md-nav__link">
    31 threads
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#example-control-file" class="md-nav__link">
    Example control file
  </a>
  
    <nav class="md-nav" aria-label="Example control file">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#control-file-variables-explained" class="md-nav__link">
    Control file variables explained
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#substitution-models" class="md-nav__link">
    Substitution Models
  </a>
  
    <nav class="md-nav" aria-label="Substitution Models">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#dna-substitution-models" class="md-nav__link">
    DNA Substitution Models
  </a>
  
    <nav class="md-nav" aria-label="DNA Substitution Models">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#jukes-cantor-model" class="md-nav__link">
    Jukes-Cantor model
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gtr-model" class="md-nav__link">
    GTR model
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#other-dna-substitution-models" class="md-nav__link">
    Other DNA substitution models
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#amino-acid-substitution-models" class="md-nav__link">
    Amino acid substitution models
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#custom-among-locus-model-variation" class="md-nav__link">
    Custom among-locus model variation
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#models-of-substitution-rate-variation" class="md-nav__link">
    Models of Substitution Rate Variation
  </a>
  
    <nav class="md-nav" aria-label="Models of Substitution Rate Variation">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#among-site-rate-variation" class="md-nav__link">
    Among-site rate variation
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#among-locus-rate-variation" class="md-nav__link">
    Among-locus rate variation
  </a>
  
    <nav class="md-nav" aria-label="Among-locus rate variation">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#prior-distribution-of-the-mean-rate-among-loci" class="md-nav__link">
    Prior distribution of the mean rate among loci
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#prior-distribution-for-locus-specific-rates-given-mean-rate-among-loci" class="md-nav__link">
    Prior distribution for locus-specific rates given mean rate among loci
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#control-file-settings-for-locusrate" class="md-nav__link">
    Control file settings for locusrate
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#among-species-rate-variation" class="md-nav__link">
    Among-species rate variation
  </a>
  
    <nav class="md-nav" aria-label="Among-species rate variation">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#variation-in-the-degree-of-departure-from-a-molecular-clock-among-loci" class="md-nav__link">
    Variation in the degree of departure from a molecular clock among loci
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#relaxed-clock-with-iid-rates-among-branches" class="md-nav__link">
    Relaxed clock with i.i.d rates among branches
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#relaxed-clock-with-autocorrelated-rates-among-branches" class="md-nav__link">
    Relaxed clock with autocorrelated rates among branches
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#control-file-settings-for-clock" class="md-nav__link">
    Control file settings for clock
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#introgression-and-migration-models" class="md-nav__link">
    Introgression and Migration Models
  </a>
  
    <nav class="md-nav" aria-label="Introgression and Migration Models">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#the-msc-i-model" class="md-nav__link">
    The MSC-I Model
  </a>
  
    <nav class="md-nav" aria-label="The MSC-I Model">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#autogeneration-of-extended-newick-graphs" class="md-nav__link">
    Autogeneration of extended newick graphs
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#examples-using-msci-create" class="md-nav__link">
    Examples using --msci-create
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#the-msc-m-model" class="md-nav__link">
    The MSC-M Model
  </a>
  
    <nav class="md-nav" aria-label="The MSC-M Model">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#control-file-specification-of-msc-m" class="md-nav__link">
    Control file specification of MSC-M
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#prior-on-migration-rates" class="md-nav__link">
    Prior on migration rates
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#variable-migration-rates-among-loci" class="md-nav__link">
    Variable migration rates among loci
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#priors-and-pseudo-priors" class="md-nav__link">
    Priors and pseudo-priors
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#odds-and-ends" class="md-nav__link">
    Odds and Ends
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#methods-of-analysis" class="md-nav__link">
    Methods of Analysis
  </a>
  
    <nav class="md-nav" aria-label="Methods of Analysis">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#a00-demography-and-divergence-time-estimation" class="md-nav__link">
    A00: Demography and Divergence Time Estimation
  </a>
  
    <nav class="md-nav" aria-label="A00: Demography and Divergence Time Estimation">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#input-a00-single-population" class="md-nav__link">
    Input A00 (single population)
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#output-a00-single-population" class="md-nav__link">
    Output A00 (single population)
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#input-a00-multiple-populations" class="md-nav__link">
    Input A00 (multiple populations)
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#output-a00-multiple-populations" class="md-nav__link">
    Output A00 (multiple populations)
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#mcmc-output-file" class="md-nav__link">
    MCMC output file
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#adjusting-step-lengths-for-mcmc-moves-finetune" class="md-nav__link">
    Adjusting step lengths for MCMC moves (finetune)
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#a01-species-tree-estimation" class="md-nav__link">
    A01: Species Tree Estimation
  </a>
  
    <nav class="md-nav" aria-label="A01: Species Tree Estimation">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#input-a01" class="md-nav__link">
    Input A01
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#output-a01" class="md-nav__link">
    Output A01
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#mcmc-output-file_1" class="md-nav__link">
    MCMC output file
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#a10-species-delimitation" class="md-nav__link">
    A10: Species Delimitation
  </a>
  
    <nav class="md-nav" aria-label="A10: Species Delimitation">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#input-a10" class="md-nav__link">
    Input A10
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#output-a10" class="md-nav__link">
    Output A10
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#mcmc-output-file_2" class="md-nav__link">
    MCMC output file
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#a11-joint-species-delimitation-and-species-tree-estimation" class="md-nav__link">
    A11: Joint Species Delimitation and Species Tree Estimation
  </a>
  
    <nav class="md-nav" aria-label="A11: Joint Species Delimitation and Species Tree Estimation">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#input-a11" class="md-nav__link">
    Input A11
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#output-a11" class="md-nav__link">
    Output A11
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#mcmc-output-file_3" class="md-nav__link">
    MCMC output file
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#advanced-features-of-bpp" class="md-nav__link">
    Advanced Features of BPP
  </a>
  
    <nav class="md-nav" aria-label="Advanced Features of BPP">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#threading-and-checkpointing" class="md-nav__link">
    Threading and Checkpointing
  </a>
  
    <nav class="md-nav" aria-label="Threading and Checkpointing">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#threads" class="md-nav__link">
    Threads
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#checkpointing" class="md-nav__link">
    Checkpointing
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#marginal-likelihood-calculations" class="md-nav__link">
    Marginal Likelihood Calculations
  </a>
  
    <nav class="md-nav" aria-label="Marginal Likelihood Calculations">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#compiling-and-running-bfdriver" class="md-nav__link">
    Compiling and running BFdriver
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#running-the-program" class="md-nav__link">
    Running the program
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#exercise-and-results" class="md-nav__link">
    Exercise and results
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#common-errors-and-problems" class="md-nav__link">
    Common errors and problems
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#simulation-using-bpp" class="md-nav__link">
    Simulation Using BPP
  </a>
  
    <nav class="md-nav" aria-label="Simulation Using BPP">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#b1-control-file-options-for-simulating-without-migration" class="md-nav__link">
    B1 Control-file options for simulating without migration
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#b2-simulating-with-migration-under-the-im-model" class="md-nav__link">
    B2 Simulating with migration (under the IM model)
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#features-new-to-bpp-40" class="md-nav__link">
    Features New To BPP 4.0
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


<h1 id="bpp-documentation">BPP Documentation</h1>
<h2 id="bpp-explained">BPP Explained</h2>
<h3 id="overview">Overview</h3>
<p>Bayesian Phylogenetics &amp; Phylogeography (BPP) is a Bayesian Markov chain Monte Carlo (MCMC) program
for analyzing DNA sequence alignments of multiple loci from multiple
closely-related species under the multispecies coalescent (MSC) model
(<a href="https://doi.org/10.1093/genetics/162.4.1811">Yang 2002</a>; <a href="https://doi.org/10.1093/genetics/164.4.1645">Rannala and Yang 2003</a>). See <a href="https://doi.org/10.1534/genetics.116.190173">Xu 2016</a> and <a href="https://inria.hal.science/PGE/hal-02535622">Rannala et al 2020</a> for reviews
of the MSC model. The program requires 3 input files:</p>
<ol>
<li>
<p>Sequence data file (see <a href="#sequence-file">Sequence File</a>).</p>
</li>
<li>
<p>Imap file specifying the population source of each sequence
    (see <a href="#imap-file">Imap file</a>).</p>
</li>
<li>
<p>Control file specifying other variables needed to run the program
    (e.g., initial species tree, prior distributions,
    etc)(see <a href="#control-file">Control File</a>).</p>
</li>
</ol>
<p>This manual will provide details for creating these input files
(describing syntax and permissible values for variables). The input
files should be in text format and edited using a text format editor
(for example, a Unix editor such as emacs, or vi). A basic set of map
and control files using sensible default priors based on the sequence
data can also be created using the web tool <em>Minimalist BPP</em> available
at</p>
<p><a href="https://brannala.github.io/bpps/">https://brannala.github.io/bpps/</a> </p>
<p>These auto-produced files can
then be used as a template for more editing.</p>
<h3 id="possible-types-of-analyses">Possible types of analyses</h3>
<p>The program can be used to conduct four different types of analyses,
each specified using two variables in the control file:</p>
<ul>
<li>
<p><strong>A00</strong> (<code>speciesdelimitation = 0</code>, <code>speciestree = 0</code>): estimation
    of species divergence times and population sizes under the MSC, IM
    and MSC-I models when the species tree is specified by the user
    (<a href="https://doi.org/10.1093/genetics/164.4.1645">Rannala and Yang 2003</a>; <a href="https://inria.hal.science/PGE/hal-02536475">Flouri et al 2020</a>)</p>
</li>
<li>
<p><strong>A01</strong> (<code>speciesdelimitation = 0</code>, <code>speciestree = 1</code>): inference of
    the species tree when sequences are assigned to species by the user
    (<a href="https://doi.org/10.1093/sysbio/syw119">Rannala and Yang 2017</a>)</p>
</li>
<li>
<p><strong>A10</strong> (<code>speciesdelimitation = 1</code>, <code>speciestree = 0</code>): species
    delimitation using a user-specified guide tree
    (<a href="https://doi.org/10.1073/pnas.0913022107">Yang and Rannala 2010</a>; <a href="https://doi.org/10.1534/genetics.112.149039">Rannala and Yang 2013</a>)</p>
</li>
<li>
<p><strong>A11</strong> (<code>speciesdelimitation = 1</code>, <code>speciestree = 1</code>): joint
    species delimitation and species tree inference using unguided
    species delimitation <a href="https://doi.org/10.1093/molbev/msu279">Yang and Rannala 2014</a></p>
</li>
</ul>
<p>A detailed guide to the use of each of these 4 methods of analysis, as
well as example control files, are provided in
<a href="#methods-of-analysis">Methods of Analysis</a>. The BPP tutorial also
provides examples illustrating the four types of analyses
(<a href="https://doi.org/10.1093/czoolo/61.5.854">Yang 2015</a>; <a href="https://inria.hal.science/PGE/hal-02536475">Flouri et al 2020</a>).</p>
<h3 id="model-parameters">Model Parameters</h3>
<p>The basic parameters in the MSC model include the <strong>species divergence times</strong> 
( <span class="arithmatex">\(\tau\)</span> ) and mutation scaled <strong>population sizes</strong></p>
<p><span class="arithmatex">\(\theta = 4N\mu\)</span></p>
<p>where <span class="arithmatex">\(N\)</span> is the effective population size and <span class="arithmatex">\(\mu\)</span>
is the mutation rate per site per generation. Note that <span class="arithmatex">\(\theta\)</span>
specifies the average proportion of sites that have different bases when
comparing two sequences sampled at random from the population. Both
<span class="arithmatex">\(\tau\)</span> and <span class="arithmatex">\(\theta\)</span> are measured in units of expected number of
mutations per site ( <span class="arithmatex">\(\tau\)</span> can be converted to units of years if the
substitution rate per site per year is known). For example, if <span class="arithmatex">\(\tau\)</span> is
an estimated divergence time in units of expected substitutions per site
(from a BPP analysis) and the substitution rate per site per year is
<span class="arithmatex">\(\mu\)</span>, then the estimated divergence time in years is, <span class="arithmatex">\(\tau'\)</span>, is</p>
<p><span class="arithmatex">\(\tau' = \frac{\tau}{\mu}\)</span> </p>
<p>and the diploid effective population size is</p>
<p><span class="arithmatex">\(N = \frac{\theta}{4\mu}\)</span> </p>
<p>For a species tree with <span class="arithmatex">\(s\)</span> species,
there are (<span class="arithmatex">\(s – 1\)</span>) divergence times (<span class="arithmatex">\(\tau\)</span>) and at most (<span class="arithmatex">\(2s – 1\)</span>)
population size parameters (<span class="arithmatex">\(\theta\)</span>). Analysis A00 estimates those
parameters when the species delimitation and species tree are fixed
(specified by the user). Analyses A01, A10, and A11 estimate parameters
and model probabilities of the different MSC models. The
multispecies-coalescent-with-introgression (MSC-I) model, first
implemented in BPP version 4.1 and described in
<a href="#the-msc-i-model">The MSC-I Model</a>, adds introgression nodes to the tree, each
with a new parameter, the <strong>introgression probability</strong> (<span class="arithmatex">\(\varphi\)</span>) see <a href="https://doi.org/10.1093/molbev/msz296">Flouri et al 2020</a>. 
The multispecies-coalescent-with--migration (MSC-M) model, first implemented in BPP version 4.6 and described in <a href="#the-msc-m-model">The MSC-M Model</a>, adds a matrix of <strong>instantaneous migration rates</strong> (<strong>M</strong>). 
For reviews of the MSC model, see <a href="https://global.oup.com/academic/product/molecular-evolution-9780199602605?lang=es&amp;cc=gb">Yang 2014</a> (Chapter 9), <a href="https://doi.org/10.1534/genetics.116.190173">Xu 2016</a> and <a href="https://inria.hal.science/PGE/hal-02535622">Rannala et al 2020</a>.</p>
<h3 id="model-assumptions">Model Assumptions</h3>
<p>The MSC model implemented in BPP makes two basic assumptions about the
data: <em>no recombination between sites of the same locus</em>, and <em>free
recombination between loci</em>. The original MSC model (<a href="https://doi.org/10.1093/genetics/164.4.1645">Rannala and Yang 2003</a>) also
assumes complete genetic isolation between populations, but more recent
implementations allow gene flow between species (either continuous
migration or episodic introgression); the MSC-I model implemented in BPP
allows episodic introgression <a href="https://doi.org/10.1093/molbev/msz296">Flouri et al 2020</a> and the IM model allows
continuous migration (see
<a href="#introgression-and-migration-models">Introgression and Migration Models</a> and <a href="#isolation-with-migration-models">Isolation With Migration Models</a>). The default model assumes a strict
molecular clock (constant substitution rates among lineages) and
relaxed-clocks (with substitution rates varying among lineages) are
implemented through variable clock models (see
<a href="#substitution-models">Substitution Models</a>).</p>
<h3 id="nuclear-genomic-data">Nuclear genomic data</h3>
<p>Ideal data for analysis using the BPP program are loosely-linked short
genomic segments. Such data are likely to satisfy the model assumptions
because intra-locus recombination is rare for short segments (500 to
1000bp) and distantly spaced segments undergo frequent recombination and
thus have nearly independent histories. Genealogical trees are assumed
to be a product of neutral evolution (not influenced by natural
selection). However, protein-coding gene sequences appear to be useable
in BPP analyses since most proteins are performing similar
functions in closely related species and the main effect of purifying
selection on nonsynonymous mutations is a reduction of the neutral
mutation rate (<a href="https://doi.org/10.1093/molbev/msx277">Shi and Yang 2018</a>; <a href="https://doi.org/10.1093/molbev/msy158">Thawornwattana et al 2018</a>). 
It is a good idea to separate the noncoding and coding regions of the genome into two
datasets.</p>
<p>The program allows the option of either a constant rate (strict
molecular clock) or variable rates of substitution (relaxed molecular
clocks) among lineages. Different substitution models are available,
with the most complex (parameter rich) being the General Time-Reversible
(GTR) model and the simplest the Jukes-Cantor (JC69) mutation model. All
the models correct for multiple hits at individual sites (see <a href="#substitution-models">Substitution Models</a>). 
The use of the JC69 model and the strict molecular clock model should be limited to
closely related species with sequence divergence not much higher than
10%. Sequences from distantly related species should be analyzed using a
GTR substitution model in combination with one of the relaxed clock
models (see <a href="#among-species-rate-variation">Among-species rate variation</a>. Models allowing substitution rate variation among sites 
(see <a href="#among-site-rate-variation">Among-site rate variation</a>) and
among loci (see <a href="#among-locus-rate-variation">Among-locus rate variation</a>)are also implemented in BPP (see <a href="#models-of-substitution-rate-variation">Models of Substitution Rate Variation</a>). </p>
<h3 id="mitochondrial-data">Mitochondrial data</h3>
<p>The MSC model accommodates genealogical heterogeneity across the genome.
Thus, different regions of the autosomal genome may have different gene
tree topologies and branch lengths (coalescent times). While the
mutational process may also vary along the genome, this heterogeneity is
believed to be much less important. Thus multiple genes from the
mitochondrial genome, which in most species does not undergo
recombination, should be treated as one 'locus' in the MSC-based
analysis. The mitochondrial genome often has a mutation rate that
differs from the nuclear genome, as well as a different effective
population size from that of autosomes. While the model of locus-rate
variation (option variable <code>locusrate</code>) and the heredity scalar (option
variable <code>heredity</code>) are designed to deal with this, it may be prudent
to analyze the loci from the nuclear genome separately from the single
mitochondrial locus.</p>
<h3 id="identical-sequences-and-phylogenetic-signal">Identical sequences and phylogenetic signal</h3>
<p>The model implemented in BPP assumes that the sequences
represent random samples from the different species. <em>Sequences from the
same species that are identical should all be used</em>. It is incorrect to
use only the unique haplotypes, which will lead to biased parameter
estimates. Similarly, it is incorrect to filter loci based on bootstrap
support values and use only those loci with a high "phylogenetic
information" signal.</p>
<h3 id="how-to-cite-bpp">How to Cite BPP</h3>
<p>If you publish results obtained using BPP you should cite the version
number of the BPP release you used as well as the canonical citation,
which as of release v4.3.8 is</p>
<pre><code>Tomáš Flouri, Xiyun Jiao, Bruce Rannala, Ziheng Yang (2018) Species
Tree Inference with BPP Using Genomic Sequences and the Multispecies
Coalescent, Molecular Biology and Evolution 35: 2585--2593.
</code></pre>
<p>You can also cite the one of the BPP tutorials and the
original papers describing the methods you used (see above). Be sure to
state the priors that you used since they are necessary for
reproducibility. If you conduct a joint analysis of species delimitation
and species tree inference, your method description may look like the
following (replace the specific values in bold with those you used):</p>
<blockquote>
<p>"Joint Bayesian species delimitation and species tree estimation was conducted using the program
BPP (Flouri et al., 2018; Yang, 2015). The method uses the multispecies coalescent model to compare
different models of species delimitation (Yang and Rannala, 2010; Rannala and Yang, 2013) and species
phylogeny (Yang and Rannala, 2014; Rannala and Yang, 2017) in a Bayesian framework, accounting
for incomplete lineage sorting due to ancestral polymorphism and gene tree-species tree discordance.
The population size parameters (<span class="arithmatex">\(\theta\)</span>s) are assigned the <strong>inverse-gamma prior IG(3, 0.002)</strong>, with mean
<strong>0.002/(3–1) = 0.001</strong>. The divergence time at the root of the species tree (<span class="arithmatex">\(\tau_0\)</span>) is assigned the 
<strong>inverse-gamma prior IG(3, 0.004)</strong>, with mean <strong>0.002</strong>, while the other divergence time parameters are specified
by the uniform Dirichlet distribution (Yang and Rannala, 2010, eq. 2). Each analysis is run at least twice
to confirm consistency between runs.”</p>
</blockquote>
<h2 id="installing-and-running-bpp">Installing and Running BPP</h2>
<p>This manual applies to BPP versions 4.1 and later. The
program is written in C and executables are available for Windows, Mac
OSX and Linux. Alternatively, the BPP program can be
compiled for Unix (Linux, BSD, etc), Mac OSX, or Windows. If you are
using a Windows, Mac OSX or Linux operating system (and you are not a
programmer who prefers to compile from source) you should download the
latest release executables for your machine (go to <a href="#obtaining-bpp">Obtaining BPP</a>). If you
have not used a command line program before, read
<a href="#using-the-command-line">Using the Command Line</a> then go to
<a href="#obtaining-bpp">Obtaining BPP</a>. If
you want to compile the program yourself (you must do this if you are
using a Unix variant other than Linux or Mac OSX) go to
<a href="#compiling-the-bpp-program">Compiling the BPP Program</a>. Note in the
examples below that the <code>$</code> symbol represents the shell prompt, type
the command listed <em>after</em> this prompt symbol (your prompt symbol may
be different).</p>
<h3 id="obtaining-bpp">Obtaining BPP</h3>
<p>Executables and source code for the most recent BPP
release are always available at:</p>
<p><a href="https://github.com/bpp/bpp/releases/latest">https://github.com/bpp/bpp/releases/latest</a></p>
<p>Download the distribution file containing an executable for your
operating system and uncompress the file. Change the current directory
to be the root of the subdirectory created by uncompressing the file.
For example, on a Linux machine you could type the following to download
and install bpp version 4.4.1 (the latest version may be different from
this):</p>
<pre><code class="language-shell">$ wget https://github.com/bpp/bpp/releases/download/\
$ v4.4.1/bpp-4.4.1-linux-x86_64.tar.gz
$ tar -xvzf bpp-4.4.1-linux-x86_64.tar.gz
$ cd bpp-4.4.1-linux-x86_64/
</code></pre>
<p>You are now ready to proceed to <a href="#bpp-trial-run">BPP Trial Run</a></p>
<h3 id="using-the-command-line">Using the Command Line</h3>
<p>BPP is a command-line program, so the preferred way of
running it is to open a terminal application and execute the program
from the terminal command line, rather than double-clicking on the
executable file in your file explorer. If you have not used the command
line before, please work through one of the following short tutorials
first:</p>
<p>Windows:<br />
<a href="http://abacus.gene.ucl.ac.uk/software/CommandLine.Windows.pdf">http://abacus.gene.ucl.ac.uk/software/CommandLine.Windows.pdf</a></p>
<p>Mac OSX:<br />
<a href="http://abacus.gene.ucl.ac.uk/software/CommandLine.MACosx.pdf">http://abacus.gene.ucl.ac.uk/software/CommandLine.MACosx.pdf</a></p>
<h3 id="compiling-the-bpp-program">Compiling the BPP Program</h3>
<h4 id="requirements">Requirements</h4>
<p>To compile the program you will need to have a C compiler and the Make
program installed on your machine. To test whether this software is
installed (on a Unix machine) you can type:</p>
<pre><code class="language-shell">$ cc --version; make --version
</code></pre>
<p>which should produce output similar to the following if a compiler is
installed:</p>
<pre><code>cc (Ubuntu 9.3.0-10ubuntu2) 9.3.0
Copyright (C) 2019 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

GNU Make 4.2.1
Built for x86_64-pc-linux-gnu
Copyright (C) 1988-2016 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later &lt;http://gnu.org/licenses/gpl.html&gt;
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
</code></pre>
<p>If instead you see an error message such as:</p>
<pre><code>Command 'cc' not found
</code></pre>
<p>then you (or your system) administrator will need to install the
compiler software (using a package manager such as apt on Ubuntu, for
example) before you can compile BPP.</p>
<h4 id="downloading-and-compiling-the-source-code">Downloading and compiling the source code</h4>
<p>If you have the necessary software for compiling you can download the C
source code for the latest release from:</p>
<p><a href="https://github.com/bpp/bpp/releases/latest">https://github.com/bpp/bpp/releases/latest</a></p>
<p>The link to the source file on github is named <code>Source code</code>. You will
need to download and uncompress the file, then change to the source code
subdirectory before executing the commands outlined below. The program
needs to be compiled only once. For example, the following commands use
the gcc compiler to compile the program and move the generated
executable file (bpp) into the bin/ folder.</p>
<pre><code class="language-shell">$ cd bpp
$ mkdir bin
$ cd src
$ make
$ mv bpp ../bin
</code></pre>
<p>If you use git you can instead clone the bpp repository and check out
the master branch (which contains source code for the latest stable
version of bpp) and then compile the program:</p>
<pre><code class="language-shell">$ git clone https://github.com/bpp/bpp.git
$ cd bpp
$ git checkout master
$ cd src
$ make
</code></pre>
<p>You are now ready to proceed to <a href="#bpp-trial-run">BPP Trial Run</a></p>
<h3 id="bpp-trial-run">BPP Trial Run</h3>
<p>Run the program from the command line (rather than double-clicking the
executable) so that you will see any potential error messages. Change
your current directory to the top level of the directory created by
uncompressing the bpp distribution file. For example, in Linux the
following commands will uncompress the distribution file and move you to
the top level of the bpp directory:</p>
<pre><code class="language-shell">$ tar -xvzf bpp-4.4.1-linux-x86_64.tar.gz
$ cd bpp-4.4.1-linux-x86_64
</code></pre>
<p>In the bpp/ subdirectory, run the program in Windows by typing the
following command within the Terminal application:</p>
<pre><code>bin\bpp --cfile examples\frogs\A00.bpp.ctl
</code></pre>
<p>In Linux or Mac OSX type the following commands within a terminal:</p>
<pre><code class="language-shell">$ cd examples/frogs/
$ ../../bin/bpp --cfile examples/frogs/A00.bpp.ctl
</code></pre>
<p>If the program executed successfully, you should see initial output
similar to the following:</p>
<pre><code>bpp v4.4.1_linux_x86_64, 31GB RAM, 12 cores
https://github.com/bpp/bpp

Auto-selected SIMD ISA: AVX2


Starting timer..
Using seed: -1
Parsing species tree... Done
Parsing phylip file... Done
</code></pre>
<p>Alternatively, you may see one or more error messages. This may be due
to a spelling error, or you may not be in the correct subdirectory. If
this does not appear to be the case and the problem persists you can ask
for help on the BPP discussion group at:</p>
<p><a href="https://groups.google.com/forum/#!forum/bpp-discussion-group">https://groups.google.com/forum/#!forum/bpp-discussion-group</a></p>
<p>When posting on the forum, please specify the exact command you typed
and the error message you received (preferably by copying and pasting
this information from your terminal or attaching a screen shot).</p>
<h2 id="input-file-formats">Input File Formats</h2>
<p>Two input files are required for every BPP analysis, the sequence file
and the control file. If more than one species is being analyzed an Imap
file is also required. A few additional input files are required only
when specific options are specified in the control file. Here
we describe the content and format specifications of the sequence, Imap
and control files that will be needed for most BPP analyses.</p>
<h3 id="sequence-file">Sequence File</h3>
<p>The sequence alignments must be in the phylip/paml format, with one
alignment following the other, all in one file. An alignment is called a
locus. Every locus must have at least 2 sequences, different loci can
have different numbers of sequences, and some species may have no
sequences present at some loci. In the phylip/paml format the first line
of the entry for each locus contains two numbers specifying the number
of sequences and the number of sites, respectively, that are present at
the locus. A sequence name must precede each sequence entry and be
separated from it by either a line break or two or more spaces. Each
sequence name must contain an individual ID, which is unique among the
sequences listed for a given locus, preceded by a caret <code>^</code> symbol. For
example, the sequence name <code>GI01234567^Specimen1</code> has the individual ID
<code>Specimen1</code>. The label <code>GI01234567</code> that precedes the <code>^</code> is optional, so
<code>^Specimen1</code> is also a valid sequence name but <code>Specimen1</code> is not. The
sequences may be in interleaved format.</p>
<p>DNA bases can be specified using either upper or lower case letters,
missing data can be specified using a <code>?</code> symbol, and IUPAC ambiguity
codes may be used. The interpretation of an ambiguity code (as either an
uncertain base or a diploid genotype) depends on whether the data are
specified as phased in the control file. By default (control file
variable <code>cleandata = 0</code>, see <a href="#control-file">Control File</a>),
alignment gaps and ambiguity nucleotides are used in the likelihood
calculation, with gaps treated as question marks 
(see <a href="https://global.oup.com/academic/product/molecular-evolution-9780199602605?lang=es&amp;cc=gb">Yang 2014</a> pp. 111-112). 
If <code>cleandata = 1</code>, all columns with gaps or ambiguity
characters are removed before analysis. An example of a sequence data
file for 2 loci and 4 sequences per locus is given below (you can also
examine the sequence files <code>frogs.txt</code> and <code>yu2001.txt</code> in the
<code>examples</code> subdirectory).</p>
<pre><code>    4 20

    dog^alfy1    attgtgccctctctctctca
    dog^alfy2    attgtgccctctctctctca
    cat^smoot    attgtgccctctagctctca
    human^ben    attgtgccctctgactctcc

    4 10

    dog^alfy1    attgtgccct
    dog^alfy2    attgtgccct
    fox^sample1  attgtgccct
    ^ben         attgtgaagt
</code></pre>
<h2 id="imap-file">Imap File</h2>
<p>Each sequence name has an individual ID tag after the caret <code>^</code> symbol.
For example, the sequence name <code>GI01234567^Specimen1</code> has the individual
ID <code>Specimen1</code>. An <strong>Imap file</strong> maps the individual IDs (specimens) to
their species/populations. For example, the following Imap file maps
<code>Specimen1</code> to species A, <code>Specimen4</code> to species B, and so on.</p>
<pre><code>    Specimen1   A
    Specimen2   A
    Specimen3   A
    Specimen4   B
    Specimen5   C
    Specimen6   C
</code></pre>
<p>See also the map files <code>frogs.Imap.txt</code> and
<code>Rokas2003-5species-Imap.txt</code> in the <code>examples</code> subdirectory. The MSC
model implemented in BPP infers the relationships among
species/populations not individuals and so it uses only the population
ID for a sequence (such as A, B, ... in the example), and does not use
the individual ID. Also when BPP reads the sequence names,
it use the individual ID to retrieve the species ID for each sequence
but then ignores the sequence name. One motivation for the use of the
Imap file and this two-layer design is that one may wish to analyze the
same sequence data with different population/species assignments of
individual IDs. This can be achieved by editing the small Imap file
rather than editing the larger sequence data file.</p>
<p><img alt="imap" src="https://github.com/bpp/bpp-manual/blob/main/figures/fig-imap.png?raw=true" /><br />
An example Imap file showing the sequence ID in the Imap file
corresponding to that in the sequence file. Note that the caret <code>^</code>
symbol is not allowed in the sequence ID of the Imap file but is
required in the sequence label in the sequence data
file.</p>
<h2 id="outgroups-and-constraints-file">Outgroups and Constraints File</h2>
<p>The <strong>outgroups and constraints file</strong> is optional and its name is
specified in the control file by setting variable <code>constraintfile</code> to
the name of the constraint file. Topological constraints on species
trees can only be specified for analyses A01 (species tree inference
with a fixed species delimitation) and A11 (joint species tree inference
and species delimitation). See <a href="#methods-of-analysis">Methods of Analysis</a>
for a detailed description of the A01 and
A11 methods of analysis. Note that species trees are always rooted in
BPP, whether you use the clock or relaxed clock. The
control variable <code>constraintfile</code> in the control file has the format</p>
<pre><code>    constraintfile = constraints.txt
</code></pre>
<p>Inside the constraint file, three keywords are allowed: <code>define</code>,
<code>constraint</code>, and <code>outgroup</code>. The <code>define</code> keyword is used to assign a
name or alias to a clade, <code>constraint</code> defines a clade or subtree, and
<code>outgroup</code> means that species not on the list (the ingroup species) form
a clade.</p>
<pre><code>    define g1 as (G,H);
    # identical to constraint (D,E,F,(G,H));
    constraint (D,E,F,g1);   
    outgroup A,B,C;
    constraint (A,B);
</code></pre>
<p>Either the equal sign (=) or spaces can be used to separate the keyword
from the specified value. In the example above, there are eight species
in the dataset: A, B, C, D, E, F, G, and H. Line 1 defines an alias <code>g1</code>
for clade (G,H), which may be used in subsequent <code>define</code> or
<code>constraint</code> statements. Line 3 specifies A, B, C as the outgroups; this
is interpreted to mean that species not on the list, (that is, D, E, F,
G and H) form a clade. Since this information is already in line 2, line
3 has no effect in this case.</p>
<p>The modifier keyword <code>NOT</code> (case-insensitive) can be used with <code>define</code>
to instead specify the complement of the species on the list. Suppose in
the data we have 10 species of ratites (flightless birds), and two
outgroup species, chicken and ostrich. You can specify the constraint as
follows:</p>
<pre><code>    define ratites as NOT (chicken, ostrich);
    constraint = (chicken, (ostrich, ratites));
</code></pre>
<p>Note that BPP interprets <code>outgroup</code> to mean that all
species not on the list form a clade. In theory the <code>outgroup</code> keyword
is unnecessary because you can always achieve the same thing by using
<code>define</code> (especially with the <code>NOT</code> option) and <code>constraint</code>. However
<code>outgroup</code> may be more convenient or explicit in some situations. If
there are four species (A,B,C,D),</p>
<pre><code>    outgroup = B,C,D;
</code></pre>
<p>achieves nothing since all the 15 rooted species trees are allowed,
whereas</p>
<pre><code>    outgroup = C,D;
</code></pre>
<p>means (A,B) form a clade so that 3 rooted species trees are allowed.
Finally if multiple compatible constraints are specified,
BPP merges them into one so that</p>
<pre><code>    constraint (E,F,G,H);
    constraint (F,G,H);
    constraint (G,H);
</code></pre>
<p>is equivalent to</p>
<pre><code>    constraint (E,(F,(G,H)));
</code></pre>
<p>If any two constraints are in conflict, BPP will abort
with an error message. For example the following will cause an error.</p>
<pre><code>    constraint = (E,F,G,H);
    constraint = (G,H);
    constraint = (F,G);
</code></pre>
<h2 id="control-file">Control File</h2>
<p>The <strong>control file</strong> specifies most aspects of a BPP analysis, including
the names of input and output files, choices of models, prior
distributions on parameters, etc. The format is line oriented, with each
line specifying the value of a particular variable. The default control
file name is <code>bpp.ctl</code>. Lines beginning with <code>*</code> or <code>#</code> are comments and
are ignored by the program. Most often the order of the lines is
unimportant, but there are exceptions (as explained below). Each
non-commented line of the control file contains a variable name and an
assignment of one or more values to the variable. The general syntax of
a line in the control file is:</p>
<pre><code>    variable = value
</code></pre>
<p>where <code>variable</code> is the name of a valid variable and <code>value</code> is a list
of one or more permissible values for the variable. Most variable
assignments follow this format --- the exceptions are described later.
BPP has complicated dependencies among variables, both in terms of
whether the variable needs to be defined (or has a default value when
undefined) and what its permissible values are. Here we begin by
defining all the variable names, the syntax of their arguments and
permissible types (integers, floating point numbers, etc), and outline
the structure of dependencies among variables</p>
<table>
<thead>
<tr>
<th>Symbol/Rule</th>
<th>Definition</th>
</tr>
</thead>
<tbody>
<tr>
<td>b</td>
<td>Boolean (0,1)</td>
</tr>
<tr>
<td>+d</td>
<td>Positive integer</td>
</tr>
<tr>
<td>f</td>
<td>floating point number</td>
</tr>
<tr>
<td>s</td>
<td>String (or character)</td>
</tr>
<tr>
<td>t</td>
<td>Tree in Newick format</td>
</tr>
<tr>
<td>x*</td>
<td>List of one or more elements of type x</td>
</tr>
<tr>
<td>y [z]</td>
<td>z is a set of variables whose form is determined by y</td>
</tr>
<tr>
<td>(w,z)</td>
<td>w and z are two different permissible types</td>
</tr>
<tr>
<td>y [(w,z)]</td>
<td>whether w or z is the permissible type depends on value of y</td>
</tr>
</tbody>
</table>
<p><strong>Table 1.</strong> Symbols used to represent syntax and value types of variables in variable definitions.</p>
<p>Detailed definitions of all the control file variables are
provided at the end of this section, ordered according to their indexes
in Table 2 below. If you are impatient to get started using
the program with worked examples, you might want to now skip ahead to
<a href="#example-control-file">Example control file</a> which explains the control file
variables in the context of a simple example dataset analysis. You can
later consult this section regarding particular control file variables
as the need arises.</p>
<table>
<thead>
<tr>
<th>Index</th>
<th>Variable</th>
<th>Values</th>
<th>Condition</th>
<th>Dependencies</th>
</tr>
</thead>
<tbody>
<tr>
<td><a href="#1-seed">1</a></td>
<td><em>seed</em></td>
<td>(-d,+d)</td>
<td>True</td>
<td>None</td>
</tr>
<tr>
<td><a href="#2-usedata">2</a></td>
<td><em>usedata</em></td>
<td>b</td>
<td>True</td>
<td>None</td>
</tr>
<tr>
<td><a href="#3-outfile">3</a></td>
<td><strong>outfile</strong></td>
<td>s</td>
<td>True</td>
<td>None</td>
</tr>
<tr>
<td><a href="#4-mcmcfile">4</a></td>
<td><strong>mcmcfile</strong></td>
<td>s</td>
<td>True</td>
<td>None</td>
</tr>
<tr>
<td><a href="#5-seqfile">5</a></td>
<td><strong>seqfile</strong></td>
<td>s</td>
<td>True</td>
<td>&lt;-17</td>
</tr>
<tr>
<td><a href="#6-finetune">6</a></td>
<td><strong>finetune</strong></td>
<td>b [f*]</td>
<td>True</td>
<td>-&gt;8</td>
</tr>
<tr>
<td><a href="#7-print">7</a></td>
<td><strong>print</strong></td>
<td>b b b b b</td>
<td>True</td>
<td>None</td>
</tr>
<tr>
<td><a href="#8-burnin">8</a></td>
<td><code>burnin</code></td>
<td>+d</td>
<td>True</td>
<td>&lt;-6</td>
</tr>
<tr>
<td><a href="#9-sampfreq">9</a></td>
<td><strong>sampfreq</strong></td>
<td>+d</td>
<td>True</td>
<td>None</td>
</tr>
<tr>
<td><a href="#10-nsample">10</a></td>
<td><strong>nsample</strong></td>
<td>+d</td>
<td>True</td>
<td>None</td>
</tr>
<tr>
<td><a href="#11-species&amp;tree">11</a></td>
<td><strong>species&amp;tree</strong></td>
<td>+d s<em> +d</em> t</td>
<td>True</td>
<td>-&gt;{12,16,25}</td>
</tr>
<tr>
<td><a href="#12-imapfile">12</a></td>
<td><code>Imapfile</code></td>
<td>s</td>
<td>11[1]&gt;1</td>
<td>&lt;-11</td>
</tr>
<tr>
<td><a href="#13-speciesdelimitation">13</a></td>
<td><em>speciesdelimitation</em></td>
<td>b [b f*]</td>
<td>True</td>
<td>-&gt;15</td>
</tr>
<tr>
<td><a href="#14-speciestree">14</a></td>
<td><em>speciestree</em></td>
<td>b</td>
<td>True</td>
<td>-&gt;14</td>
</tr>
<tr>
<td><a href="#15-speciesmodelprior">15</a></td>
<td><code>speciesmodelprior</code></td>
<td>+d</td>
<td>13[1]=1 OR 14[1]=1</td>
<td>&lt;-{13,14}</td>
</tr>
<tr>
<td><a href="#16-phase">16</a></td>
<td><strong>phase</strong></td>
<td>b*</td>
<td>11[1]</td>
<td>&lt;-11</td>
</tr>
<tr>
<td><a href="#17-nloci">17</a></td>
<td><em>nloci</em></td>
<td>+d</td>
<td>True</td>
<td>-&gt;5 &lt;-31</td>
</tr>
<tr>
<td><a href="#18-model">18</a></td>
<td><em>model</em></td>
<td>s [s]</td>
<td>True</td>
<td>None</td>
</tr>
<tr>
<td><a href="#19-qrates">19</a></td>
<td><em>Qrates</em></td>
<td>b +f +f +f +f +f +f</td>
<td>18=''7''</td>
<td>&lt;-18</td>
</tr>
<tr>
<td><a href="#20-basefreqs">20</a></td>
<td><em>basefreqs</em></td>
<td>b +f +f +f +f</td>
<td>18=''7''</td>
<td>&lt;-18</td>
</tr>
<tr>
<td><a href="#21-alphaprior">21</a></td>
<td><em>alphaprior</em></td>
<td>f f +d</td>
<td>True</td>
<td>None</td>
</tr>
<tr>
<td><a href="#22-cleandata">22</a></td>
<td><em>cleandata</em></td>
<td>b</td>
<td>True</td>
<td>None</td>
</tr>
<tr>
<td><a href="#thetaprior">23</a></td>
<td><em>thetaprior</em></td>
<td>s [(f*, f f s)]</td>
<td>True</td>
<td>None</td>
</tr>
<tr>
<td><a href="#tauprior">24</a></td>
<td><em>tauprior</em></td>
<td>s f f</td>
<td>True</td>
<td>None</td>
</tr>
<tr>
<td><a href="#phiprior">25</a></td>
<td><code>phiprior</code></td>
<td>f f</td>
<td>11[4]</td>
<td>&lt;-11</td>
</tr>
<tr>
<td><a href="#locusrate">26</a></td>
<td><em>locusrate</em></td>
<td>+d [(f f f s, s, )]</td>
<td>True</td>
<td>&lt;-27 -&gt;27</td>
</tr>
<tr>
<td><a href="#clock">27</a></td>
<td><em>clock</em></td>
<td>+d [f f f s s]</td>
<td>True</td>
<td>&lt;-26 -&gt;26</td>
</tr>
<tr>
<td><a href="#heredity">28</a></td>
<td><em>heredity</em></td>
<td>+d [(f f, s)]</td>
<td>True</td>
<td>None</td>
</tr>
<tr>
<td><a href="#checkpoint">29</a></td>
<td><em>checkpoint</em></td>
<td>+d [(+d)]</td>
<td>True</td>
<td>None</td>
</tr>
<tr>
<td><a href="#constraintfile">30</a></td>
<td><em>constraintfile</em></td>
<td>s</td>
<td>True</td>
<td>None</td>
</tr>
<tr>
<td><a href="#threads">31</a></td>
<td><code>threads</code></td>
<td>+d [ +d +d ]</td>
<td>True</td>
<td>-&gt;17</td>
</tr>
</tbody>
</table>
<p><strong>Table 2.</strong> Complete list of BPP control file variables. The column <code>Variable</code> contains the
  control file variable name, the column <code>Values</code> specifies (using symbols) the format of the values (symbols are defined in
  Table 1). The column <code>Condition</code> describes the conditions (if any) needed for the variable to be required. An entry
  of True in this column means that the variable is always required/defined although possibly not explicitly (there may be a
  default value). The column <code>Dependencies</code> lists the variables that either influence the variable (&lt;-) or are influenced by the variable
  (-&gt;) with the variables denoted using their index from column <code>Index</code>.</p>
<p>Table 2 lists all the variables that can be defined
in the BPP control file. The first column lists an index used to
identify the variable in other columns of the table and to determine the
order in which the variables are defined in subsequent text. The second
column lists the variable name. The variable names are highlighted to
indicate the structure of their dependencies among variables as follows:</p>
<ul>
<li>
<p>Bold = variable always needs to be defined, has no default value, and
    does not depend on other variables.</p>
</li>
<li>
<p>Italic = variable always needs to be defined, has a default value if
    no value is provided in control file, and does not depend on other
    variables.</p>
</li>
<li>
<p>Solid background  = whether variable needs to defined (and possibly also its
    permissible values) depends on the value of another variable.</p>
</li>
<li>
<p>Regular = variable always needs to be defined and has no default value
    but its permissible values depend on the value of another variable.</p>
</li>
</ul>
<p>Column 3 lists the permissible types of each variable and the syntax for
specifying arguments in defining the variable. The variable types are
defined in Table 1.</p>
<p>Column 4 lists the condition (state of another variable) that either
determines whether the variable on that line exists (solid background variables)
or determines its permissible values (italic or regular variables). x [y]
refers to value of argument y of the variable with index x. Column 5
lists indexes of variables that either are influenced by the variable
(the notation -&gt;[c,d] indicates that the variable influences
variables c and d) or variables that influence the variable (the
notation &lt;-[a] indicates that variable a influences the variable). If
you are writing a control file from scratch, these dependencies
necessitate that you specify the bold variables first, followed by the
italic (only needed if changing defaults) and then the solid background and regular.</p>
<h2 id="bpp-control-file-variables">BPP control file variables</h2>
<h3 id="1-seed">1 seed</h3>
<hr />
<pre><code>seed = (-d,+d)
</code></pre>
<p><strong>DESCRIPTION</strong><br />
Specifies the seed used for the random number generator.<br />
<strong>VALUES</strong><br />
<code>-d</code>, use the wall clock (in Windows) or white noise (in Linux) to
generate a seed (store seed in file <code>SeedUsed</code>).<br />
<code>+d</code>, use +d as the seed.<br />
<strong>DEFAULT</strong><br />
<code>-1</code><br />
<strong>COMMENTS</strong><br />
Using the same positive integer seed will produce identical results in
different runs, which is useful for debugging, and using different
positive integers in different runs produces different results. Using
<span class="arithmatex">\(-d\)</span>, for example, <span class="arithmatex">\(-1\)</span>, the computer's entropy generator is used as a
source for the seed, and different runs will produce different results.
When evaluating MCMC convergence by comparing results across different
runs be sure to use -d. The positive integer seed automatically
generated when using option -d is stored in a file named SeedUsed and
can be used explicitly to replicate a result. It is recommended that you
run each analysis at least twice using different seeds to confirm that
the results are stable across runs.<br />
<strong>EXAMPLES</strong></p>
<pre><code>seed = -1
seed = 278
</code></pre>
<h3 id="2-usedata">2 usedata</h3>
<hr />
<pre><code>usedata = b
</code></pre>
<p><strong>DESCRIPTION</strong><br />
Specifies whether data (likelihood+priors) are used in calculating
probabilities during MCMC or only priors.<br />
<strong>VALUES</strong><br />
<code>0</code>, use only the priors to calculate probabilities (likelihood
constant).<br />
<code>1</code>, use likelihood and prior probabilities.<br />
<strong>DEFAULT</strong><br />
<code>1</code><br />
<strong>COMMENTS</strong><br />
The <code>0</code> option can be used for debugging, or examining priors. When
using option <code>0</code> a MCMC run produces samples from the prior for each
variable.<br />
<strong>EXAMPLES</strong></p>
<pre><code>usedata = 0
usedata = 1
</code></pre>
<h3 id="3-outfile">3 outfile</h3>
<hr />
<pre><code>outfile = s
</code></pre>
<p><strong>DESCRIPTION</strong><br />
Sets the path/name of the output file to be the string <code>s</code><br />
<strong>VALUES</strong><br />
<code>s</code>, a string of characters specifying the directory path and name
of file that will receive the program output.<br />
<strong>EXAMPLES</strong></p>
<pre><code>outfile = /home/mickey/mouse_out.txt
</code></pre>
<h3 id="4-mcmcfile">4 mcmcfile</h3>
<hr />
<pre><code>mcmcfile = s
</code></pre>
<p><strong>DESCRIPTION</strong><br />
Sets the name of the file receiving MCMC output to be the string
<code>s</code><br />
<strong>VALUES</strong><br />
<code>s</code>, a string of characters specifying the directory path and name
of file that will receive the MCMC output.<br />
<strong>EXAMPLES</strong></p>
<pre><code>mcmcfile = /home/mickey/mouse_mcmc.txt
mcmcfile = mcmc.txt
</code></pre>
<h3 id="5-seqfile">5 seqfile</h3>
<hr />
<pre><code>seqfile = s
</code></pre>
<p><strong>DESCRIPTION</strong><br />
Sets the name of the file containing the sequence data to be the string
<code>s</code><br />
<strong>VALUES</strong><br />
<code>s</code>, a string of characters specifying the directory path and name
of file that contains the sequence data<br />
<strong>EXAMPLES</strong></p>
<pre><code>seqfile = /home/mickey/mouse_seq.txt
seqfile = sequences.txt
</code></pre>
<h3 id="6-finetune">6 finetune</h3>
<hr />
<pre><code>finetune = b: [f*]
</code></pre>
<p><strong>DESCRIPTION</strong><br />
Determines whether step lengths in MCMC proposals are automatically
optimized or manually set to fixed values as well as setting either the
fixed step lengths (manual) or the starting step lengths (optimized).
There are up to 15 proposals with step lengths that can be adjusted. Not
all proposals are defined for any given choice of model and analysis,
however 15 values should always be specified when using manual as the
order is fixed, step lengths specified for unused proposals will be
ignored by the program. The sizes of proposal steps for each parameter
are provided (from left to right) in the same order as listed (from top
to bottom in table. The proposals are as follows:</p>
<table>
<thead>
<tr>
<th>Abbreviation</th>
<th>Parameter Proposal Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>Gage</td>
<td>Node age on gene tree</td>
</tr>
<tr>
<td>Gspr</td>
<td>Subtree pruning and regrafting move</td>
</tr>
<tr>
<td>thet</td>
<td>Theta</td>
</tr>
<tr>
<td>tau</td>
<td>Tau</td>
</tr>
<tr>
<td>mix</td>
<td>Mixing step (jointly changing mu, tau and gene tree node ages</td>
</tr>
<tr>
<td>lrht</td>
<td>empty</td>
</tr>
<tr>
<td>phi</td>
<td>Introgression probability</td>
</tr>
<tr>
<td>pi</td>
<td>Stationary frequencies of substitution model</td>
</tr>
<tr>
<td>qmat</td>
<td>Q matrix elements of substitution model</td>
</tr>
<tr>
<td>alfa</td>
<td>Alpha parameter of gamma model for rate variation among sites</td>
</tr>
<tr>
<td>mubr</td>
<td>empty</td>
</tr>
<tr>
<td>nubr</td>
<td>empty</td>
</tr>
<tr>
<td>mu_i</td>
<td>empty</td>
</tr>
<tr>
<td>nu_i</td>
<td>empty</td>
</tr>
<tr>
<td>brte</td>
<td>empty</td>
</tr>
</tbody>
</table>
<p><strong>VALUES</strong><br />
<code>0: Gage Gspr thet tau mix lrht phi pi qmat alfa mubr nubr mu_i nu_i brte</code><br />
Use fixed step lengths in MCMC proposals as specified, where Gage is the
specified step length for proposal Gage, and so on.<br />
<code>1: Gage Gspr thet tau mix lrht phi pi qmat alfa mubr nubr mu_i nu_i brte</code><br />
Automatically optimize step lengths in MCMC proposals using specified
initial step lengths.<br />
<strong>DEFAULT</strong><br />
<code>0</code><br />
Manually fixes all step lengths to be <span class="arithmatex">\(0.1\)</span>.<br />
<code>1</code><br />
Fixes all initial proposal step lengths to <span class="arithmatex">\(0.1\)</span> prior to optimization.<br />
<strong>DEPENDENCIES</strong><br />
If <code>b</code> is set to 1 (automatic optimization) then <code>burnin</code> has to
be <span class="arithmatex">\(&gt;200\)</span>.<br />
<strong>COMMENTS</strong><br />
The first value, before the colon, is a switch, with 0 meaning no
automatic adjustments by the program and 1 meaning automatic adjustments
by the program. Following the colon are the step lengths for the
proposals used in the program. and then the step lengths specified here
will be the initial step lengths, and the program will try to adjust
them using the information collected during the burnin step. This option
appears to work fine. Some notes about manually adjusting those finetune
step lengths are provided below in section 3.2.<br />
<strong>EXAMPLES</strong></p>
<pre><code>finetune = 0: .01 .02 .03 .04 .05 .01 .01 .01 .01 .01 .01 .01 .01 .01 .01
finetune = 1
</code></pre>
<h3 id="7-print">7 print</h3>
<hr />
<pre><code>print = b b b b b
</code></pre>
<p><strong>DESCRIPTION</strong><br />
Specifies what information is saved to the output files during the
MCMC.<br />
<strong>VALUES</strong><br />
There are 5 boolean (0,1) variables that specify whether particular
variables are saved to file (1) or not saved (0). From left to right:
variable 1 specifies MCMC samples; variable 2 specifies locus-specific
rate parameters (rate <span class="arithmatex">\(\mu_i\)</span>, variance parameter <span class="arithmatex">\(\nu_i\)</span> and
species-tree branch rates for locus <span class="arithmatex">\(r_{ij}\)</span>); variable 3 specifies
locus-specific heredity scalars if those are estimated from the data;
variable 4 specifies locus-specific gene trees; and variable 5 specifies
locus-specific substitution-rate parameters (qmat for the <span class="arithmatex">\(Q\)</span> matrix,
freqs for base frequencies, and alpha for gamma rates for sites).<br />
<strong>EXAMPLES</strong></p>
<pre><code>print = 0 1 1 1 0
print = 1 1 1 1 1
</code></pre>
<h3 id="8-burnin">8 burnin</h3>
<hr />
<pre><code>burnin = +d
</code></pre>
<p><strong>DESCRIPTION</strong><br />
Specifies the number of burn-in iterations that are executed in the MCMC
run before sampling begins.<br />
<strong>VALUES</strong><br />
<code>+d</code>, a positive integer specifying the number of burn-in
iterations.<br />
<strong>DEPENDENCIES</strong><br />
If <code>finetune = 1</code> (automatic optimization) then <code>burnin</code> has to
be <span class="arithmatex">\(&gt;200\)</span>.<br />
<strong>COMMENTS</strong><br />
The total number of MCMC iterations is <code>burnin</code> + <code>nsample</code> <span class="arithmatex">\(\times\)</span> <code>sampfreq</code>.<br />
<strong>EXAMPLES</strong></p>
<pre><code>burnin = 100
burnin = 10000
</code></pre>
<h3 id="9-sampfreq">9 sampfreq</h3>
<hr />
<pre><code>sampfreq = +d
</code></pre>
<p><strong>DESCRIPTION</strong><br />
Specifies the interval at which samples from the MCMC are to be written
to the output file specified by <code>mcmcfile</code>.<br />
<strong>VALUES</strong><br />
<code>+d</code>, a positive integer specifying the interval between samples
from the MCMC.<br />
<strong>COMMENTS</strong><br />
The total number of MCMC iterations is <code>burnin</code> + <code>nsample</code> <span class="arithmatex">\(\times\)</span> <code>sampfreq</code>.<br />
<strong>EXAMPLES</strong></p>
<pre><code>sampfreq = 2
sampfreq = 10
</code></pre>
<h3 id="10-nsample">10 nsample</h3>
<hr />
<pre><code>nsample = +d
</code></pre>
<p><strong>DESCRIPTION</strong><br />
Specifies the number of samples from the MCMC that are to be written to
the output file specified by <code>mcmcfile</code>.<br />
<strong>VALUES</strong><br />
<code>+d</code>, a positive integer specifying the number of MCMC samples.<br />
<strong>COMMENTS</strong><br />
The total number of MCMC iterations is <code>burnin</code> + <code>nsample</code> <span class="arithmatex">\(\times\)</span> <code>sampfreq</code>.<br />
<strong>EXAMPLES</strong></p>
<pre><code>nsample = 5000
nsample = 25499
</code></pre>
<h3 id="11-speciestree">11 species&amp;tree</h3>
<hr />
<pre><code>species&amp;tree = +d s*
+d*
t
</code></pre>
<p><strong>DESCRIPTION</strong><br />
Specifies the number of species, the species names, the maximum number
of sequences (at any locus) for each species, and the species tree in
Newick format (or species graph in extended Newick format in models with
introgression).<br />
<strong>VALUES</strong><br />
<code>+d</code>, a positive integer specifying the total number of species.
<code>s*</code>, a list of <code>+d</code> strings specifying the name of each
species. <code>+d*</code>, a list of <code>+d</code> positive integers specifying the
maximum number of sequences (at any locus) for each species. <code>t</code>, a
tree (or graph) in Newick format (or Newick extended format) specifying
the relationships among species (and possibly introgression events).<br />
<strong>DEPENDENCIES</strong><br />
The variable <code>Imapfile</code> must be specified only when the number of
species <code>+d</code> &gt; 1. The length of the boolean array
specified for variable <code>phase</code> is equal to the number of species
<code>+d</code>. The variable <code>phiprior</code> must be specified only if <code>t</code>
is an introgression graph.<br />
<strong>COMMENTS</strong><br />
See the subsection <em>Notation for trees and introgression graphs</em> for
details on the format of <code>t</code>. The role of <code>t</code> in the MCMC
analysis depends on the value of the <code>speciestree</code> variable. If
<code>speciestree=1</code> the variable <code>t</code> specifies the starting tree for
the MCMC analysis, otherwise if <code>speciestree=0</code> it is the fixed tree
used for estimating parameters or the fixed guide tree for species
delimitation.<br />
<strong>EXAMPLES</strong></p>
<pre><code>species&amp;tree = 4 H C G O
1 1 1 1
(((H, C), G), O);
</code></pre>
<h3 id="12-imapfile">12 Imapfile</h3>
<hr />
<pre><code>Imapfile = s
</code></pre>
<p><strong>DESCRIPTION</strong><br />
Sets the path/name of the Imap file to be the string <code>s</code>.<br />
<strong>VALUES</strong><br />
<code>s</code>, a string of characters specifying the directory path and name
of a file that contains the species/sequence map information.<br />
<strong>DEPENDENCIES</strong><br />
The variable <code>Imapfile</code> must be specified only when the number of
species defined by the <code>species&amp;tree</code> variable
<code>+d</code> &gt; 1.<br />
<strong>COMMENTS</strong><br />
See <a href="#imap-file">Imap File</a> for a complete description of the Imap file format.<br />
<strong>EXAMPLES</strong></p>
<pre><code>Imapfile = /home/foo/bar_Imap.txt
Imapfile = Imap.txt
</code></pre>
<h3 id="13-speciesdelimitation">13 speciesdelimitation</h3>
<hr />
<pre><code>speciesdelimitation = b [b f*]
</code></pre>
<p><strong>DESCRIPTION</strong><br />
Specifies whether species delimitation is performed and if so specifies
the parameters of the species delimitation rjMCMC proposal algorithm.<br />
<strong>VALUES</strong><br />
<code>0</code>, specifies no species delimitation.<br />
<code>1 0 epsilon</code>, specifies species delimitation with rjMCMC algorithm
0, with parameter <span class="arithmatex">\(\epsilon\)</span> as specified in equations 3 and 4 of
<a href="https://doi.org/10.1073/pnas.0913022107">Yang and Rannala 2010</a>. Reasonable values for <span class="arithmatex">\(\epsilon\)</span> are 1, 2, 5, etc.<br />
<code>1 1 alpha m</code>, specifies rjMCMC algorithm 1, with parameters
<span class="arithmatex">\(\alpha\)</span> and <span class="arithmatex">\(m\)</span> as specified in equations 6 and 7 of <a href="https://doi.org/10.1073/pnas.0913022107">Yang and Rannala 2010</a>.
Reasonable values are <span class="arithmatex">\(\alpha = 1, 1.5, 2\)</span>, etc. and <span class="arithmatex">\(m = 0.5, 1, 2\)</span>.<br />
<strong>DEFAULT</strong><br />
<code>0</code><br />
<strong>DEPENDENCIES</strong><br />
If <code>speciesdelimitation = 1</code> then <code>speciesmodelprior</code> must be defined.<br />
<strong>EXAMPLES</strong></p>
<pre><code>speciesdelimitation = 1 0 2
speciesdelimitation = 1 1 2 1
</code></pre>
<h3 id="14-speciestree">14 speciestree</h3>
<hr />
<pre><code>speciestree = b
</code></pre>
<p><strong>DESCRIPTION</strong><br />
Specifies whether the species tree is estimated or fixed in the MCMC
analysis.<br />
<strong>VALUES</strong><br />
<code>0</code>, specifies that species tree is fixed.<br />
<code>1</code>, specifies that species tree is estimated.<br />
<strong>DEFAULT</strong><br />
<code>0</code><br />
<strong>DEPENDENCIES</strong><br />
If <code>speciestree = 1</code> then <code>speciesmodelprior</code> must be defined.  </p>
<p><strong>COMMENTS</strong><br />
This option invokes the nearest-neighbor interchange (NNI) or
subtree-pruning-regrafting (SPR) algorithm to propose changes to the
species tree topology during the MCMC run.<br />
<strong>EXAMPLES</strong></p>
<pre><code>speciestree = 1
speciestree = 0
</code></pre>
<h3 id="15-speciesmodelprior">15 speciesmodelprior</h3>
<hr />
<pre><code>speciesmodelprior
</code></pre>
<p><strong>DESCRIPTION</strong><br />
Specifies the prior distribution for rooted species trees and species
delimitations.<br />
<strong>VALUES</strong><br />
<code>0</code>, specifies equal probabilities for the labeled histories (rooted
trees with the internal nodes ordered by age).<br />
<code>1</code>, specifies equal probabilities for the rooted species trees.<br />
<code>2</code>, specifies equal probabilities for the numbers of species (<span class="arithmatex">\(1/s\)</span>
each for <span class="arithmatex">\(1,2,\ldots,s\)</span> species given <span class="arithmatex">\(s\)</span> populations) and divides the
probability for any specific number of species among the compatible
models (of species delimitation and species phylogeny) in proportion to
the number of labeled histories.<br />
<code>3</code>, specifies equal probabilities for the numbers of species (<span class="arithmatex">\(1/s\)</span>
each for <span class="arithmatex">\(1,2,\ldots,s\)</span> species given <span class="arithmatex">\(s\)</span> populations) and divides the
probability for any specific number of species among the compatible
models (of species delimitation and species phylogeny) uniformly.<br />
<strong>DEPENDENCIES</strong><br />
The variable <code>speciesmodelprior</code> must be defined if either (or both)
<code>speciestree = 1</code> or <code>speciesdelimitation = 1</code>. If
<code>speciestree = 1</code> and <code>speciesdelimitation = 0</code> then only
options <code>speciesmodelprior = 0</code> or <code>speciesmodelprior = 1</code> may
be used.<br />
<strong>COMMENTS</strong><br />
The priors specified by options 2 and 3 are described in <a href="https://doi.org/10.1093/molbev/msu279">Yang and Rannala 2014</a>.
The prior specified by option 3 may be suitable when there are many populations.<br />
<strong>EXAMPLES</strong></p>
<pre><code>speciesmodelprior = 0
speciesmodelprior = 3
</code></pre>
<h3 id="16-phase">16 phase</h3>
<hr />
<pre><code>phase = b*
</code></pre>
<p><strong>DESCRIPTION</strong><br />
Specifies whether the sequence data for each species is phased.<br />
<strong>VALUES</strong><br />
<code>1</code>, in position <span class="arithmatex">\(i\)</span> of the list indicates that data is unphased for
species <span class="arithmatex">\(i\)</span> in the list of species in variable <code>species&amp;tree</code>,
otherwise <code>0</code> indicates it is phased.<br />
<strong>DEFAULT</strong><br />
<code>0</code><br />
<strong>DEPENDENCIES</strong><br />
The number of boolean (0,1) variables in the list must equal the number
of species <code>d+</code> specified in the <code>species&amp;tree</code> variable.<br />
<strong>COMMENTS</strong><br />
If at position <span class="arithmatex">\(i\)</span> the <code>phase</code> variable is 1, each sequence from
species <span class="arithmatex">\(i\)</span> is treated as an unphased diploid sequence, with
heterozygous sites represented using the ambiguity characters <code>YRMKSW</code>.
The missing data tokens <code>N-?</code> are allowed and are all treated as ?, but
other ambiguity characters (<code>HBVD</code>) are not allowed.</p>
<p>For phasing, each sequence is expanded/resolved into two sequences, and
the program averages over possible phase resolutions according to the
approach of @Gronau2011. For example, a sequence <code>R...Y</code> with two
heterozygous sites, <code>R</code> (meaning <code>A/G</code>) and <code>Y</code> (meaning <code>C/T</code>), has two
possible phase resolutions:</p>
<ol>
<li>
<p><code>A...T</code> and <code>G...C</code></p>
</li>
<li>
<p><code>A...C</code> and <code>G...T</code></p>
</li>
</ol>
<p>A sequence with no heterozygous sites will be phased/resolved into two
identical sequences. The MSC model effectively averages over different
phase resolutions of the heterozygous sites in performing the likelihood
calculation. The option may increase memory usage and CPU time
considerably if many sequences exist with many heterozygote sites at a
locus. If the <code>phase</code> variable for a species is 0, all sequences
from that species are treated as resolved haplotype sequences, and
ambiguities are interpreted in the usual way. For example, an ambiguity
code <code>Y</code> at a position is interpreted to mean that one sequence exists
with an uncertain nucleotide at that position that is either a T or a C.
The program does not allow some sequences from a species to be phased
and other sequences from that same species to be unphased.<br />
<strong>EXAMPLES</strong></p>
<pre><code>phase = 0 0 0 0 1
phase = 1 1 1 1 1
</code></pre>
<h3 id="17-nloci">17 nloci</h3>
<hr />
<pre><code>nloci = +d
</code></pre>
<p><strong>DESCRIPTION</strong><br />
Specifies the number of loci present in the sequence data file specified
by variable <code>seqfile</code> that will be analysed.<br />
<strong>VALUES</strong><br />
<code>+d</code>, a positive integer specifying the number of loci to be
analyzed.<br />
<strong>DEFAULT</strong><br />
If <code>nloci</code> is not specified, all the loci present in the sequence
file specified by <code>seqfile</code> will be analysed.<br />
<strong>DEPENDENCIES</strong><br />
<code>nloci</code> must be less than or equal to the number of loci available
in the sequence file specified by <code>seqfile</code>.<br />
<strong>COMMENTS</strong><br />
If <code>nloci</code> is less than the number of loci present in the file
specified by <code>seqfile</code> only the first <code>nloci</code> present in that
file will be analysed.<br />
<strong>EXAMPLES</strong></p>
<pre><code>nloci = 1
nloci = 2000
</code></pre>
<h3 id="18-model">18 model</h3>
<hr />
<pre><code>model = s [s]
</code></pre>
<p><strong>DESCRIPTION</strong><br />
Specifies the DNA substitution model (or amino acid substitution model)
that will be used. For DNA sequences the simplest (default) model that
can be chosen is the Jukes-Cantor (JC69) model and the most complex is
the General Time-Reversible (GTR) model. A range of models of
intermediate complexity are also available. For amino acid sequences,
most of the commonly used amino acid substitution matrix models are
available. If a model other than <code>Custom</code> is specified it will apply to
all loci. If <code>Custom filename</code> is specified, an additional file
<code>filename</code> must be provided that contains one or more lines with the
format:</p>
<pre><code>loci_indices data_type model
</code></pre>
<p>where <code>loci_indices</code> is an index or range of indexes for loci (indexed
according to their order of occurrence in the sequence data file),
<code>data_type</code> is either <code>DNA</code> or <code>AA</code>, and model specifies the
substitution model variable for the indexed loci as described above. For
example, the entry</p>
<pre><code>1-10 DNA JC69
11 AA DAYHOFF
</code></pre>
<p>specifies that loci 1 to 10 comprise DNA sequences that will have the
<code>JC69</code> model applied to them, while locus 11 is an amino acid sequence
that will have the <code>DAYHOFF</code> model applied to it.<br />
<strong>VALUES</strong><br />
For DNA sequences: <code>JC69</code>, <code>K80</code>, <code>F81</code>, <code>HKY</code>,
<code>T92</code>, <code>TN93</code>, <code>F84</code>, <code>GTR</code>.<br />
For amino acid sequences: <code>DAYHOFF</code>, <code>LG</code>, <code>DCMUT</code>,
<code>JTT</code>, <code>MTREV</code>, <code>WAG</code>, <code>RTREV</code>, <code>CPREV</code>, <code>VT</code>,
<code>BLOSUM62</code>, <code>MTMAM</code>, <code>MTART</code>, <code>MTZOA</code>, <code>PMB</code>,
<code>HIVB</code>, <code>HIVW</code>, <code>JTTDCMUT</code>, <code>FLU</code>, and <code>STMTREV</code>.<br />
Among-locus model variation: <code>Custom filename</code>.<br />
<strong>DEFAULT</strong><br />
<code>JC69</code><br />
<strong>DEPENDENCIES</strong><br />
If <code>model = GTR</code> the variables <code>Qrates</code> and <code>basefreqs</code> can
be optionally defined which provide priors on the rate matrix and base
frequencies, respectively (default priors will be used otherwise).<br />
<strong>EXAMPLES</strong></p>
<pre><code>model = GTR
model = CPREV
model = Custom models.txt
</code></pre>
<h3 id="19-qrates">19 Qrates</h3>
<hr />
<pre><code>Qrates = b +f +f +f +f +f +f
</code></pre>
<p><strong>DESCRIPTION</strong><br />
Specifies whether the exchangeability parameters in the GTR model are
fixed or variable and either their fixed values or prior relative mean
values.<br />
<strong>VALUES</strong><br />
<code>b</code>, specifies either fixed (1) or variable (0) exchangeability
parameters in the GTR model.<br />
<code>+f +f +f +f +f +f</code>, specify the exchangeability parameters in the
order <span class="arithmatex">\(a, b, c, d, e, f\)</span> for TC, TA, TG, CA, CG, and AG, as described in
<a href="https://global.oup.com/academic/product/molecular-evolution-9780199602605?lang=es&amp;cc=gb">Yang 2014</a>.<br />
<strong>DEFAULT</strong><br />
<code>0 1 2 1 1 2 1</code><br />
<strong>DEPENDENCIES</strong><br />
The variable <code>Qrates</code> is defined only if <code>model = GTR</code>.<br />
<strong>COMMENTS</strong><br />
Two example cases are discussed here with either fixed, or variable
rates. The following specifies fixed rates:</p>
<pre><code>Qrates = 1  2 1 1 1 1 2
</code></pre>
<p>This specifies fixed rates at <span class="arithmatex">\(a = 2, b = c = d = e = 1\)</span> and <span class="arithmatex">\(f = 2\)</span>, so
that the transition rate is twice as high as the transversion rate, and
the model corresponds to K80 or HKY. Note that only the relative rates
matter, because the rate matrix (<span class="arithmatex">\(Q\)</span>) is scaled so that the average rate
(over the base frequencies) is 1 and branch lengths are measured in the
expected number of mutations/substitutions per site. Nevertheless the
program expects six rate parameters.</p>
<pre><code>Qrates = 0  10 5 5 5 5 10 
</code></pre>
<p>This specifies variable rates with a prior on <span class="arithmatex">\(a, b, c, d, e\)</span> that is a
Dirichlet distribution for every locus. The above specifies Dir(10, 5,
5, 5, 5, 10), with parameters
<span class="arithmatex">\(\alpha_a = 10, \alpha_b = \alpha_c = \alpha_d = \alpha_e = 10\)</span>, and
<span class="arithmatex">\(\alpha_f = 10\)</span>. The rates generated from the Dirichlet sum to 1, and
they are rescaled. Note that larger values for those parameters mean
less variance, while the mean for <span class="arithmatex">\(a\)</span>, say, is given by
<span class="arithmatex">\(\alpha_a/\alpha\)</span> with
<span class="arithmatex">\(\alpha = \alpha_a + \alpha_b + \alpha_c + \alpha_d + \alpha_e +
\alpha_f\)</span>. This specifies an average transition/transversion rate ratio
of 2 (if base frequencies are all fixed at <span class="arithmatex">\(\frac{1}{4}\)</span>).<br />
<strong>EXAMPLES</strong></p>
<pre><code>Qrates = 1 2 1 1 1 1 2
Qrates = 0 1 1 1 1 1 1
</code></pre>
<h3 id="20-basefreqs">20 basefreqs</h3>
<hr />
<pre><code>basefreqs = b +f +f +f +f
</code></pre>
<p><strong>DESCRIPTION</strong><br />
Specifies whether the base frequency parameters are fixed or variable
and either their fixed values or prior relative mean values.<br />
<strong>VALUES</strong><br />
<code>b</code>, specifies either fixed (1) or variable (0) base frequency
parameters in the GTR model.<br />
<code>+f +f +f +f</code>, specify the base frequency parameters in the order
TCAG.<br />
<strong>DEFAULT</strong><br />
<code>0 1 1 1 1</code><br />
<strong>DEPENDENCIES</strong><br />
The variable <code>basefreqs</code> is defined only if <code>model = GTR</code>.<br />
<strong>COMMENTS</strong><br />
The parameters <span class="arithmatex">\(\pi_T, \pi_C, \pi_A, \pi_G\)</span> can be either fixed or
variable with a prior specified by the Dirichlet distribution with
<span class="arithmatex">\(\alpha\)</span> parameters specifying the relative mean frequencies.<br />
<strong>EXAMPLES</strong></p>
<pre><code>basefreqs = 1 0.15 0.35 0.15 0.35
basefreqs = 0 10 10 10 10
</code></pre>
<h3 id="21-alphaprior">21 alphaprior</h3>
<hr />
<pre><code>alphaprior = f f +d
</code></pre>
<p><strong>DESCRIPTION</strong><br />
Specifies the prior probability density for the shape parameter
controlling the pattern of among-site rate variation.<br />
<strong>VALUES</strong><br />
<code>f f +d</code>, specify the <span class="arithmatex">\(\alpha_p\)</span> and <span class="arithmatex">\(\beta_p\)</span> parameters of the
prior on the Gamma shape parameter and the number of rate categories
ncatG, respectively.<br />
<strong>DEFAULT</strong><br />
<span class="arithmatex">\(\alpha=\infty\)</span> (no rate variation among sites)<br />
<strong>COMMENTS</strong><br />
Among-site rate variation is assumed to follow a Gamma distribution with
a mean of 1 and shape parameter <span class="arithmatex">\(\alpha\)</span>. The variance of rates among
sites is <span class="arithmatex">\(1/\alpha\)</span> and therefore a smaller <span class="arithmatex">\(\alpha\)</span> specifies more rate
variation. A discrete distribution is used to efficiently approximate
the Gamma distribution and ncatG is the number of rate categories used
in the approximation -- more rate categories provide a better
approximation but incur greater computational expense. A value for ncatG
of 4 is recommended. The prior on <span class="arithmatex">\(\alpha\)</span> is also a Gamma distribution
with parameters <span class="arithmatex">\(\alpha_p\)</span> and <span class="arithmatex">\(\beta_p\)</span>. The variance and mean of the
prior on <span class="arithmatex">\(\alpha\)</span> are: </p>
<p><span class="arithmatex">\(\textrm{Mean}(\alpha) = \frac{\alpha_p}{\beta_p}\)</span>,</p>
<p><span class="arithmatex">\(\textrm{Var}(\alpha) = \frac{\alpha_p}{\beta_p^2}\)</span></p>
<p><strong>EXAMPLES</strong></p>
<pre><code>alphaprior = 1 1 4
</code></pre>
<h3 id="22-cleandata">22 cleandata</h3>
<hr />
<pre><code>cleandata = b
</code></pre>
<p><strong>DESCRIPTION</strong><br />
Specifies whether columns in the alignment which have gaps or ambiguity
characters will be removed prior to analysis, or will instead be
retained for use in the likelihood calculation.<br />
<strong>VALUES</strong><br />
<code>0</code>, specifies that columns in the alignment which have gaps or
ambiguity characters will be treated as missing data and used in the
likelihood calculation.<br />
<code>1</code>, specifies the removal of columns in the alignment which have
gaps or ambiguity characters.<br />
<strong>DEFAULT</strong><br />
<code>0</code><br />
<strong>EXAMPLES</strong></p>
<pre><code>cleandata = 1
cleandata = 0
</code></pre>
<h3 id="23-thetaprior">23 thetaprior</h3>
<hr />
<pre><code>thetaprior = s [(f*, f f s)]
</code></pre>
<p><strong>DESCRIPTION</strong><br />
Specifies the distribution model and parameters of the prior
distribution on the contemporary and ancestral population parameters
<span class="arithmatex">\(\theta\)</span> and whether the MCMC integrates over the parameters or the
integration is instead performed analytically.<br />
<strong>VALUES</strong><br />
<code>s</code>, specifies the form of the prior distribution on theta and
should be either <code>invgamma</code>, <code>gamma</code> or <code>beta</code> to specify either an
inverse gamma, gamma, or beta distribution, respectively.
<code>[(f*, f f s)]</code>, specifies the parameters of the prior distribution.
The form for the parameters depends on the type of distribution
specified in the first argument. The permissible combinations of
distributions and parameters are:</p>
<pre><code>    invgamma a b
    invgamma a b e
</code></pre>
<p>which specifies an inverse gamma prior with <span class="arithmatex">\(\alpha\)</span> and <span class="arithmatex">\(\beta\)</span>
parameters a and b, respectively. The first entry specifies that
<span class="arithmatex">\(\theta\)</span>s are integrated over analytically and the second entry (with
trailing e) specifies that <span class="arithmatex">\(\theta\)</span>s are estimated.</p>
<pre><code>    gamma a b
</code></pre>
<p>which specifies a gamma prior with <span class="arithmatex">\(\alpha\)</span> and <span class="arithmatex">\(\beta\)</span> parameters a and
b, respectively.</p>
<pre><code>    beta p q l u
</code></pre>
<p>which specifies a beta prior constrained to the interval <span class="arithmatex">\((l,u)\)</span> with
parameters p and q, respectively. All the distributions above, except
the inverse Gamma distribution (without the e argument) integrate over
the <span class="arithmatex">\(\theta\)</span> parameters numerically as part of the MCMC, estimating the
posterior distribution of <span class="arithmatex">\(\theta\)</span> for each population.<br />
<strong>COMMENTS</strong><br />
The inverse-gamma prior <code>invgamma a b</code>, where a and b are the parameters
<span class="arithmatex">\(\alpha\)</span> and <span class="arithmatex">\(\beta\)</span>, has mean and variance: </p>
<p><span class="arithmatex">\(\textrm{Mean}(\theta) = \frac{\beta}{\alpha - 1}\)</span>,</p>
<p><span class="arithmatex">\(\textrm{Var}(\theta) = \frac{\beta^2}{(\alpha - 1)^2 (\alpha - 2)}\)</span>.</p>
<p>For example, if <span class="arithmatex">\(\alpha=3\)</span> and <span class="arithmatex">\(\beta=0.002\)</span> the mean is
<span class="arithmatex">\(0.002/(3 – 1) = 0.001\)</span> (one variable site per kb on average). Note that
all <span class="arithmatex">\(\theta\)</span> parameters in the MSC model (for both modern species and
extinct ancestral species) are assigned the same specified prior
distribution with identical parameters. The inverse-gamma is a conjugate
prior for <span class="arithmatex">\(\theta\)</span> (Hey and Nielsen, 2007), which means that both the
prior and the posterior of <span class="arithmatex">\(\theta\)</span> will be inverse-gamma. Use of the
conjugate prior allows the <span class="arithmatex">\(\theta\)</span> parameters to be integrated out
analytically, and thus the dimension of the parameter space is reduced.
This typically leads to improved mixing of the MCMC. However, with this
option, the posterior of the <span class="arithmatex">\(\theta\)</span> parameters will not be produced.
To estimate the <span class="arithmatex">\(\theta\)</span> parameters when using an inverse gamma prior,
add the letter e (or E) on the line, as follows</p>
<pre><code>    thetaprior = 3 0.002 e
</code></pre>
<p>Whether <span class="arithmatex">\(\theta\)</span>s are integrated out or estimated, other results (such
as the posterior probability of species trees or species-delimitation
models) should be identical if the same prior is used. The analytical
integration of <span class="arithmatex">\(\theta\)</span> is only possible with an inverse gamma prior
distribution, if any other prior is used the <span class="arithmatex">\(\theta\)</span>s are automatically
estimated and the E flag is not needed and should not be used. The gamma
prior <code>gamma a b</code>, where a and b are the parameters <span class="arithmatex">\(\alpha\)</span> and
<span class="arithmatex">\(\beta\)</span>, has mean and variance: </p>
<p><span class="arithmatex">\(\textrm{Mean}(\theta) = \frac{\alpha}{\beta}\)</span>,</p>
<p><span class="arithmatex">\(\textrm{Var}(\theta) = \frac{\alpha}{\beta^2}\)</span>.</p>
<p>For example, if <span class="arithmatex">\(\alpha=0.001\)</span> and <span class="arithmatex">\(\beta=1\)</span> the mean is
<span class="arithmatex">\(0.001/1 = 0.001\)</span> (one variable site per kb on average). The beta prior
<code>beta p q l u</code> has mean and variance: </p>
<p><span class="arithmatex">\(\textrm{Mean}(\theta) = \frac{p u + q l}{p + q}\)</span>,</p>
<p><span class="arithmatex">\(\textrm{Var}(\theta) = \frac{p q (u - l)^2}{(p + q)^2 (p + q + 1)}\)</span>.</p>
<p>For example, with <code>beta 2 18 1e-6 0.1</code> the mean is: </p>
<p><span class="arithmatex">\(\frac{2(0.1)+18(10^{-6})}{2+18} = 0.0100009\)</span> </p>
<p>which is 10 variable sites per kb on average.<br />
<strong>EXAMPLES</strong></p>
<pre><code>thetaprior = invgamma 3 0.002
thetaprior = invgamma 4 0.001 e
thetaprior = beta 2 18 1e-6 0.1
thetaprior = beta 2 18 1e-6 0.01
thetaprior = beta 1 10 0.0001 0.2
thetaprior = gamma 0.001 1
</code></pre>
<h3 id="24-tauprior">24 tauprior</h3>
<hr />
<pre><code>tauprior = s f f
</code></pre>
<p><strong>DESCRIPTION</strong><br />
Specifies the prior probability distribution of the divergence time
parameter for the root in the species tree.
<strong>VALUES</strong>\
<code>s f f</code>, specifies either the gamma or inverse gamma prior
distribution and the <span class="arithmatex">\(\alpha\)</span> and <span class="arithmatex">\(\beta\)</span> parameters, respectively for
<span class="arithmatex">\(\tau_0\)</span>, the divergence time parameter for the root in the species
tree.</p>
<pre><code>    gamma a b
    invgamma a b
</code></pre>
<p>specifies a gamma distribution, or an inverse gamma distribution,
respectively with a and b to be the parameters <span class="arithmatex">\(\alpha\)</span> and <span class="arithmatex">\(\beta\)</span>,
respectively.<br />
<strong>COMMENTS</strong><br />
If the specified prior on <span class="arithmatex">\(\tau_0\)</span>, the age of the root of the tree is
an inverse-gamma distribution IG(<span class="arithmatex">\(\alpha, \beta\)</span>) with <span class="arithmatex">\(\alpha &gt; 2\)</span> the
mean and variance are: </p>
<p><span class="arithmatex">\(\textrm{Mean}(\tau_0) = \frac{\beta}{\alpha - 1}\)</span>,</p>
<p><span class="arithmatex">\(\textrm{Var}(\tau_0) = \frac{\beta^2}{(\alpha - 1)^2 (\alpha - 2)}\)</span>.</p>
<p>If the prior is a gamma distribution the mean and variance are:</p>
<p><span class="arithmatex">\(\textrm{Mean}(\tau_0) = \frac{\alpha}{\beta}\)</span>,</p>
<p><span class="arithmatex">\(\textrm{Var}(\tau_0) = \frac{\alpha}{\beta^2}\)</span>.</p>
<p>The remaining species divergence times are generated from the uniform
Dirichlet distribution [@Yang2010 eq. 2]. As an example, suppose that
<code>tauprior = 3 0.03</code> specifies the prior for <span class="arithmatex">\(\tau_0\)</span>, the divergence
time parameter for the root in the species tree. The mean of <span class="arithmatex">\(\tau_0\)</span> is
then <span class="arithmatex">\(0.03/(3 – 1) = 0.015\)</span> (which means 1.5% of sequence divergence
between the root of the species tree and the present time). If the
mutation rate is <span class="arithmatex">\(10^{–9}\)</span> mutations/site/year, this distance will
translate to an absolute prior mean divergence time of 15 MY.<br />
<strong>EXAMPLES</strong></p>
<pre><code>tauprior = invgamma 3 0.03
tauprior = gamma 0.01 1
</code></pre>
<h3 id="25-phiprior">25 phiprior</h3>
<hr />
<pre><code>phiprior = f f
</code></pre>
<p><strong>DESCRIPTION</strong><br />
Specifies the parameters of the prior distribution on <span class="arithmatex">\(\varphi\)</span>, the
introgression probability under the MSC-I model.<br />
<strong>VALUES</strong><br />
<code>f f</code>, specify the parameters <span class="arithmatex">\(\alpha\)</span> and <span class="arithmatex">\(\beta\)</span>, respectively, of
the beta distribution prior <span class="arithmatex">\(\textrm{B}(\alpha,\beta)\)</span> for <span class="arithmatex">\(\varphi\)</span>.<br />
<strong>DEPENDENCIES</strong><br />
If the <code>species&amp;tree</code> variable <code>t</code> specifies a tree with
introgression nodes then <code>phiprior</code> must be defined.<br />
<strong>COMMENTS</strong><br />
The introgression probability <span class="arithmatex">\(\varphi\)</span> takes values on the interval
<span class="arithmatex">\((0,1)\)</span> and the prior distribution for <span class="arithmatex">\(\varphi\)</span> is assumed to be beta
<span class="arithmatex">\(\textrm{B}(\alpha, \beta)\)</span> which has mean and variance:</p>
<p><span class="arithmatex">\(\textrm{Mean}(\varphi) = \frac{\alpha}{\alpha+\beta}\)</span>,</p>
<p><span class="arithmatex">\(\textrm{Var}(\varphi) = \frac{\alpha \beta}{(\alpha + \beta)^2 (\alpha + \beta + 1)}\)</span>.</p>
<p>For example, <code>phiprior = 1 1</code> specifies the beta prior
<span class="arithmatex">\(\textrm{B}(\alpha,\beta)\)</span> with <span class="arithmatex">\(\alpha = 1\)</span> and <span class="arithmatex">\(\beta = 1\)</span> for <span class="arithmatex">\(\varphi\)</span>, which has a mean of <span class="arithmatex">\(1/2\)</span> and a variance of <span class="arithmatex">\(1/12\)</span>.<br />
<strong>EXAMPLES</strong></p>
<pre><code>phiprior = 1 1
phiprior = 1.5 1.5
</code></pre>
<h3 id="26-locusrate">26 locusrate</h3>
<hr />
<pre><code>locusrate = d+[(f f f s, s)]
</code></pre>
<p><strong>DESCRIPTION</strong><br />
Specifies the model of substitution rate variation among loci, as well
as the parameters, and the prior on the parameters of the model.<br />
<strong>VALUES</strong><br />
<code>d+[(f f f s, s)]</code>, specifies the model and parameters. The variable
d+ takes one of 3 possible values with different numbers of additional
variables depending on d+. The permissible combinations are:</p>
<table>
<thead>
<tr>
<th><strong>d+</strong></th>
<th><strong>[args]</strong></th>
<th><strong>Description</strong></th>
</tr>
</thead>
<tbody>
<tr>
<td>0</td>
<td></td>
<td>Loci have identical rates</td>
</tr>
<tr>
<td>1</td>
<td>f f f s</td>
<td>Locus rates variable and estimated</td>
</tr>
<tr>
<td></td>
<td><span class="arithmatex">\(\alpha_{\bar{\mu}} \,\, \beta_{\bar{\mu}} \,\, \alpha_{\mu_i}\)</span> prior</td>
<td></td>
</tr>
<tr>
<td>2</td>
<td>s</td>
<td>Locus rates variable and specified</td>
</tr>
<tr>
<td></td>
<td>Filename of file containing locus rates</td>
<td></td>
</tr>
</tbody>
</table>
<p>where <span class="arithmatex">\(\alpha_{\bar{\mu}}\)</span> and <span class="arithmatex">\(\beta_{\bar{\mu}}\)</span> are the parameters of
the Gamma distribution prior on the average rate among loci <span class="arithmatex">\(\bar{\mu}\)</span>
and <span class="arithmatex">\(\alpha_{\mu_i}\)</span> is the parameter of the prior on the rate at the
<span class="arithmatex">\(i\)</span>th locus given <span class="arithmatex">\(\bar{\mu}\)</span>. The prior variable must be either <code>iid</code>
for a conditional iid (hierarchical) prior on <span class="arithmatex">\(\mu_i\)</span> or <code>dir</code> for a
Gamma-Dirichlet prior. The prior variable is optional, if it is not
specified <code>iid</code> is used.<br />
<strong>DEFAULT</strong><br />
<code>0</code><br />
<strong>DEPENDENCIES</strong><br />
If both the <code>prior</code> option for the variable <code>locusrate</code> and the <code>prior</code>
option for the variable <code>clock</code> are set, they should take the same
value.<br />
<strong>COMMENTS</strong><br />
The setting <code>locusrate = 0</code> (default) means that all loci have the same
mutation rate. Setting <code>locusrate = 1 f f f s</code> specifies a model with
rate variation among loci in which rates are estimated from the sequence
data. Including the first integer (1) which species the model there are
5 arguments, the last one (prior) being optional. Parameters
<span class="arithmatex">\(\alpha_{\bar{\mu}}\)</span> and <span class="arithmatex">\(\beta_{\bar{\mu}}\)</span> specify the shape and rate
parameters of the gamma distribution for the mean rate across loci
(<span class="arithmatex">\(\bar\mu\)</span>), while <span class="arithmatex">\(\alpha_{\mu_i}\)</span> and prior are used to specify the
locus rates (<span class="arithmatex">\(\mu_i\)</span>) given the mean rate (<span class="arithmatex">\(\bar\mu\)</span>). The option prior
can take two values dir (for Gamma-Dirichlet rates for loci), and iid
(for conditional i.i.d. or hierarchical prior).
<code>locusrate = 2 LocusRateFileName</code> specifies the fixed-rates model of
locus-rate variation (<a href="https://doi.org/10.1093/molbev/msn148">Burgess and Yang 2008</a>). This is the strategy used by
<a href="https://doi.org/10.1093/genetics/162.4.1811">Yang 2002</a>, with the relative rates estimated by the distance to an
outgroup species. The relative locus rates are listed in the file: there
should be as many numbers in the file, separately by spaces or line
returns, as the number of loci (nloci). The program re-scales those
rates so that the average across all loci is 1 and then use those
relative rates as fixed constants. Specifically the mean rate across
loci (<span class="arithmatex">\(\bar{\mu}\)</span>) is assigned a gamma prior:</p>
<p><span class="arithmatex">\(\bar{\mu} = \mathrm{G}(\alpha_{\bar{\mu}},\beta_{\bar{\mu}}),\)</span></p>
<p>with mean and variance:</p>
<p><span class="arithmatex">\(\textrm{Mean}(\bar{\mu}) = \frac{\alpha_{\bar{\mu}}}{\beta_{\bar{\mu}}}\)</span>,</p>
<p><span class="arithmatex">\(\textrm{Var}(\bar{\mu}) = \frac{\alpha_{\bar{\mu}}}{\beta_{\bar{\mu}}^2}\)</span>.</p>
<p>When there are no fossil calibrations in the species tree, the rates
should all be relative. In this case we suggest fixing
<span class="arithmatex">\(\alpha_{\bar{\mu}}=\beta_{\bar{\mu}}=0\)</span> (with 0 causing the program to
fix both parameters at <span class="arithmatex">\(\infty\)</span>), and the program will fix the mean rate
across loci at <span class="arithmatex">\(\bar\mu = 1\)</span>. Otherwise one can use equal and large
values such as <span class="arithmatex">\(\alpha_{\bar{\mu}}=100\)</span> and <span class="arithmatex">\(\beta_{\bar{\mu}}=100\)</span>, so
that <span class="arithmatex">\(\bar\mu\)</span> is nearly fixed at 1. Given the mean rate <span class="arithmatex">\(\bar\mu\)</span>, two
priors are available to specify the locus rates <span class="arithmatex">\(\mu_i\)</span>, with
<span class="arithmatex">\(i = 1, 2, \cdots, L\)</span>, where <span class="arithmatex">\(L\)</span> is the number of loci (<code>nloci</code>). If
<code>prior = dir</code> (for Gamma-Dirichlet distribution of locus rates), the
total rate <span class="arithmatex">\(L\bar\mu\)</span>, given the mean rate (<span class="arithmatex">\(\bar\mu\)</span>), is partitioned
into locus rates (<span class="arithmatex">\(\mu_i\)</span>), using the concentration parameter
<span class="arithmatex">\(\alpha_{\mu_i}\)</span> The model and notation follow <a href="https://doi.org/10.1093/molbev/msn148">Burgess and Yang 2008</a> [eq. 4] and
<a href="https://doi.org/10.1093/sysbio/syu020">Dos Reis et al 2014</a> [eqs. 3-5]. If <code>prior = iid</code> (for conditional-i.i.d. or
hierarchical prior of locus rates), the locus rates (<span class="arithmatex">\(\mu_i\)</span>) are
i.i.d. given the mean rate (<span class="arithmatex">\(\bar\mu\)</span>):</p>
<p><span class="arithmatex">\(\mu_i \sim \mathrm{G}(\alpha_{\mu_i},\alpha_{\mu_i}/\bar{\mu})\)</span>. </p>
<p>This model is described in <a href="https://doi.org/10.1093/sysbio/syu109">Zhu et al 2015</a> [eq. 8] and is also implemented in
[mcmctree]{.smallcaps}. In both the Gamma-Dirichlet and the conditional
i.i.d. models, parameter <span class="arithmatex">\(\alpha_{\mu_i}\)</span> is inversely related to the
extent of rate variation among loci, with a large <span class="arithmatex">\(\alpha_{\mu_i}\)</span>
meaning similar rates among loci. If all loci are noncoding, the rates
are probably similar, so <span class="arithmatex">\(\alpha_{\mu_i} =\)</span> 10 or 20 may be reasonable,
while for coding loci or exons, <span class="arithmatex">\(\alpha_{\mu_i} =\)</span> 2 or 1 may be
appropriate. The <span class="arithmatex">\(\alpha_{\mu_i}\)</span> parameter may affect the estimates of
the population size parameter (<span class="arithmatex">\(\theta\)</span>) for the root node on the
species tree (<a href="https://doi.org/10.1093/molbev/msn148">Burgess and Yang 2008</a>). The <span class="arithmatex">\(L\)</span> locus rates (<span class="arithmatex">\(\mu_i\)</span>) are
parameters in the model. If <span class="arithmatex">\(\alpha_{\bar{\mu}} &gt; 0\)</span>, the mean rate
(<span class="arithmatex">\(\bar\mu\)</span>) is a parameter as well.<br />
<strong>EXAMPLES</strong></p>
<pre><code>locusrate = 0
locusrate = 2 rates.txt
locusrate = 1 0 0 2
locusrate = 1 2 3 2 dir
locusrate = 1 1 1 1 iid
</code></pre>
<h3 id="27-clock">27 clock</h3>
<hr />
<pre><code>clock = +d [f f f s s]
</code></pre>
<p><strong>DESCRIPTION</strong><br />
Specifies whether a strict molecular clock (equal substitution rates
among lineages) is used or instead a specific variable clock model
(allowing variation in substitution rates among lineages). If a variable
clock model is specified, the priors on the parameters of the variable
clock model are also specified.<br />
<strong>VALUES</strong><br />
<code>+d</code>, specifies a strict clock (1), a variable clock with
independent rates among branches (2), or a variable clock with
autocorrelated rates between ancestral and descendent branches (3).<br />
<code>f f f s s</code>, are required when <code>clock =</code> 2 or 3 and specify
respectively the parameters <span class="arithmatex">\(\alpha_{\bar{\nu}}\)</span>, <span class="arithmatex">\(\beta_{\bar{\nu}}\)</span>
and <span class="arithmatex">\(\alpha_{\nu_i}\)</span>, the prior distribution for the locus rate
(<span class="arithmatex">\(\mu_i\)</span>) which is either <code>dir</code> for Dirichlet or <code>iid</code> for conditional
i.i.d (hierarchical) prior, and the distribution for the branch rates
(<span class="arithmatex">\(r_{ij}\)</span>) given the locus rate (<span class="arithmatex">\(\mu_i\)</span>) which is either <code>G</code> for Gamma
or <code>LN</code> for log-normal. Parameters <span class="arithmatex">\(\alpha_{\bar{\nu}}\)</span> and
<span class="arithmatex">\(\beta_{\bar{\nu}}\)</span> are the parameters of the gamma distribution for the
average of variances across loci (<span class="arithmatex">\(\bar\nu\)</span>), while <span class="arithmatex">\(\alpha_{\nu_i}\)</span> and
<code>prior</code> are used to specify the variance (<span class="arithmatex">\(\nu_i\)</span>) for locus <span class="arithmatex">\(i\)</span> given
the average (<span class="arithmatex">\(\bar\nu\)</span>). The specification of the distribution of
<span class="arithmatex">\(\nu_i\)</span> given <span class="arithmatex">\(\bar\nu\)</span> follows the same procedure as the specification
of the distribution of <span class="arithmatex">\(\mu_i\)</span> given the mean rate <span class="arithmatex">\(\bar\mu\)</span> when
modeling among-locus rate variation; see notes above about the
<code>locusrate</code> variable.<br />
<strong>DEFAULT</strong><br />
<code>1</code><br />
<strong>DEPENDENCIES</strong><br />
If both the <code>prior</code> option for the variable <code>locusrate</code> and the <code>prior</code>
option for the variable <code>clock</code> are set, they should take the same
value.<br />
<strong>COMMENTS</strong><br />
As an example, the specification (<code>clock = 2 10.0 100.0 5.0 iid G</code>)
specifies a local-clock model in which the mutation rate drifts over
branches, independently among loci. A detailed explanation is as
follows. First <span class="arithmatex">\(\bar\nu\)</span> is assigned a gamma distribution
<span class="arithmatex">\(G(10.0, 100.0)\)</span>, with mean 0.1. Given <span class="arithmatex">\(\bar\nu\)</span>, the conditional
i.i.d. prior means that
<span class="arithmatex">\(\nu_i \sim G(\alpha_{\nu_i}, \alpha_{\nu_i}/\bar\nu)\)</span> with shape
parameter <span class="arithmatex">\(\alpha_{\nu_i} = 5.0\)</span> and mean <span class="arithmatex">\(\bar\nu\)</span>, for
<span class="arithmatex">\(i = 1, 2, \cdots, L\)</span>. Note that <span class="arithmatex">\(\alpha_{\nu_i}\)</span> is inversely related
to the variance of <span class="arithmatex">\(\nu_i\)</span>: use small values of <span class="arithmatex">\(\alpha_{\nu_i}\)</span> (2, 1,
or 0.5) if you believe that <span class="arithmatex">\(\nu_i\)</span> varies among loci (meaning that the
clock nearly holds at some loci but is seriously violated at others).</p>
<p>As another example, the specification
(<code>clock = 2 10.0 100.0 5.0 dir LN</code>) means that <span class="arithmatex">\(\bar\nu\)</span> is assigned a
gamma prior <span class="arithmatex">\(G(10.0, 100.0)\)</span>, and then the sum <span class="arithmatex">\(L\bar\nu\)</span> is partitioned
into <span class="arithmatex">\(\nu_i\)</span> (for <span class="arithmatex">\(i = 1, 2, \cdots, L\)</span>) according to the Dirichlet
distribution (<code>prior = dir</code>) with concentration parameter
<span class="arithmatex">\(\alpha_{\nu_i} = 5.0\)</span>. Again large <span class="arithmatex">\(\alpha_{\nu_i}\)</span> means the same
extent of clock violation at different loci.</p>
<p>Given the locus rate <span class="arithmatex">\(\mu_i\)</span> (specified using the <code>locusrate</code> variable)
and the variance <span class="arithmatex">\(\nu_i\)</span> (specified using the <code>clock</code> variable) for
locus <span class="arithmatex">\(i\)</span>, the different lineages may have different rates at the locus,
and those rates are independent among loci. If distribution specified is
<code>G</code> (for gamma), the rate for species-tree branch <span class="arithmatex">\(j\)</span> at locus <span class="arithmatex">\(i\)</span> has
the following gamma distribution</p>
<p><span class="arithmatex">\(r_{ij} | \mu_i, \nu_i \sim \mathrm{G}(\mu_i^2/\nu_i, \mu_i/\nu_i)\)</span>.</p>
<p>This has mean <span class="arithmatex">\(\mu_i\)</span> and variance <span class="arithmatex">\(\nu_i\)</span>. Alternatively if
distribution = <code>LN</code> (for log-normal), the rate for species-tree branch
<span class="arithmatex">\(j\)</span> at locus <span class="arithmatex">\(i\)</span> has the following the log-normal distribution</p>
<p><span class="arithmatex">\(r_{ij} | \mu_i, \nu_i \sim \mathrm{LN}(\mu_i, \nu_i)\)</span>,</p>
<p>where <span class="arithmatex">\(\mu_i\)</span>
is the mean of the LN distribution and <span class="arithmatex">\(\nu_i\)</span> is the variance parameter
of the lognormal. Note that the rate-drift model specifies rates for
branches on the species tree (rather than on the gene tree) for each
locus, and gene-tree branches residing in the same population or species
have the same rate. For example, if all sequences at a locus are from
the same species and all coalescent events occur in that species (before
reaching an ancestral species), all branches on the gene tree will have
the same rate even if the relaxed-clock model allows rates among
species. In contrast, if a gene-tree branch passes several species or
populations, the different segments of the branch will have different
rates. The branch length on the gene tree is calculated by summing up
the lengths of those segments (with the length of each segment being the
product of the rate and the time duration for the segment).</p>
<p>The option <code>clock = 3</code> is similar to clock 2 but specifies the
autocorrelated-rates model. <code>clock = 3</code> with the log-normal distribution
(LN) specifies the geometric Brownian motion model of <a href="https://doi.org/10.1080/10635150701420643">Rannala and Yang 2007</a>. This
assigns a rate to each species-tree branch, that is, to the mid-point of
the branch. Given the rate at the species-tree root (<span class="arithmatex">\(\mu_i\)</span> at locus
<span class="arithmatex">\(i\)</span>), the rates for the two branches around the root are specified. Then
given the rate for each ancestral branch, the rates for its two daughter
branches are specified, by integrating over the rate at the internal
node that is ancestral to the daughter branches. See figure 1 and
equations 3-8 in <a href="https://doi.org/10.1080/10635150701420643">Rannala and Yang 2007</a>. The rates for all species-tree branches
are thus assigned through a pre-order tree traversal, starting from the
root moving to the tips, until all branches are visited. If <code>clock = 3</code>
is specified with the gamma distribution (option <code>G</code>), the model works
as follows, using a similar pre-order tree traversal. First the two
branches at the species-tree root have the gamma distribution with mean
<span class="arithmatex">\(\mu_i\)</span> and variance <span class="arithmatex">\(\nu_i\)</span>. Then given the rate for each ancestral
branch, the rates for its two daughter branches are specified as
independent gamma variables with the mean to be the rate of the parental
branch and with the variance to be <span class="arithmatex">\(\nu_i\)</span>. Clock 3 is currently
implemented for the MSC model only and is unavailable under the MSC-I
model.</p>
<p>Note that a larger variance <span class="arithmatex">\(\nu_i\)</span> implies a more serious violation of
the molecular clock at locus <span class="arithmatex">\(i\)</span>. Also note that <span class="arithmatex">\(\nu_i\)</span> will be similar
to <span class="arithmatex">\(\bar\nu\)</span>, especially if <span class="arithmatex">\(\alpha_{\nu_i}\)</span> is large, and <span class="arithmatex">\(\bar\nu\)</span> has
prior mean <span class="arithmatex">\(\alpha_{\bar{\nu}} / \beta_{\bar{\nu}}\)</span>. If
<span class="arithmatex">\(\alpha_{\bar{\nu}} = 10\)</span> and <span class="arithmatex">\(\beta_{\bar{\nu}} = 100\)</span>, the prior mean
will be 0.1. Thus, for the log-normal model <span class="arithmatex">\(\nu = 0.5\)</span> represents a
serious violation of the clock while <span class="arithmatex">\(\nu &lt; 0.1\)</span> represents only a
slight violation.<br />
<strong>EXAMPLES</strong></p>
<pre><code>clock = 1
clock = 2 10.0 100.0 5.0 dir LN
clock = 3 10.0 50.0 3.0 dir G
</code></pre>
<h3 id="28-heredity">28 heredity</h3>
<hr />
<pre><code>heredity = +d [(f f, s, )]
</code></pre>
<p><strong>DESCRIPTION</strong><br />
Specifies whether proportional differences exist between all <span class="arithmatex">\(\theta\)</span>
values for different loci, and if so whether the proportionality
multipliers (called inheritance scalars) are estimated from the data, or
set to specified fixed values. If inheritance scalars are estimated, the
parameters of the prior also must be specified.<br />
<strong>VALUES</strong><br />
<code>+d</code>, specifies whether no differences in <span class="arithmatex">\(\theta\)</span> exist among loci
(0), proportional differences exist and inheritance scalars are
estimated (1), or proportional differences exist and inheritance scalars
are fixed to specified values (2).<br />
<code>f f</code>, is only specified when <code>+d</code> is 1 and specifies the <span class="arithmatex">\(\alpha\)</span>
and <span class="arithmatex">\(\beta\)</span> parameters, respectively, of the Gamma prior on the scalar
<span class="arithmatex">\(s_i\)</span> for locus <span class="arithmatex">\(i\)</span>.<br />
<code>s</code>, is only specified when <code>+d</code> is 2 and specifies the name of a
file containing <span class="arithmatex">\(L\)</span> (number of loci) heredity scalars (which must be
positive numbers) separated by spaces or linebreaks.<br />
<strong>DEFAULT</strong><br />
<code>0</code><br />
<strong>COMMENTS</strong><br />
<code>heredity = 0</code> is the default and constrains <span class="arithmatex">\(\theta\)</span> to be equal across
loci. <code>heredity = 1</code> or <code>2</code> specifies two models that allow <span class="arithmatex">\(\theta\)</span> to
vary among loci, which may be useful for analyses combining data from
autosomal loci, loci on sex chromosomes, and/or mitochondrial loci. Such
mixed data, may be expected to have proportional differences in
effective population sizes among loci so that inheritance scalars
<a href="https://doi.org/10.1534/genetics.103.024182">Hey and Nielsen 2004</a> should be applied. For example, in diploid sexual organisms
with strict maternal inheritance of mtDNA we expect the effective
population size of a mitochondrial locus to be <span class="arithmatex">\(1/4\)</span> that of a nuclear
locus. Other factors such as natural selection may also cause <span class="arithmatex">\(\theta\)</span>
to deviate from the neutral expectation even among autosomal loci.
BPP implements two options for allowing such variations.
The first option (<code>heredity = 1</code>) specifies that locus-specific
inheritance scalars <span class="arithmatex">\(s_i\)</span> be estimated, using a Gamma prior with
parameters <span class="arithmatex">\(\alpha\)</span> and <span class="arithmatex">\(\beta\)</span> specified by the user. The prior mean
and variance of <span class="arithmatex">\(s_i\)</span> are:</p>
<p><span class="arithmatex">\(\textrm{Mean}(s_i) = \frac{\alpha}{\beta}\)</span>,</p>
<p><span class="arithmatex">\(\textrm{Var}(s_i) = \frac{\alpha}{\beta^2}\)</span>.</p>
<p>For example, <code>(heredity = 1 4 4)</code> specifies a Gamma prior
<span class="arithmatex">\(\textrm{G}(4, 4)\)</span>, with mean <span class="arithmatex">\(4/4 = 1\)</span> and variance <span class="arithmatex">\(1/4\)</span>, for the
inheritance scalar at each locus. The MCMC then generates the posterior
distribution of <span class="arithmatex">\(s_i\)</span> for each locus. The second option (<code>heredity = 2</code>)
allows the user to specify the <span class="arithmatex">\(s_i\)</span> in a file so they remain fixed
constants during the MCMC run.<br />
<strong>EXAMPLES</strong></p>
<pre><code>heredity = 0
heredity = 1 4 4
heredity = 2 scalars.txt
</code></pre>
<h3 id="29-checkpoint">29 checkpoint</h3>
<hr />
<pre><code>checkpoint = +d [(+d)]
</code></pre>
<p><strong>DESCRIPTION</strong><br />
Specifies whether one or more checkpoint files are created to allow the
BPP program to resume using the current state of the MCMC at the
iteration during which the checkpoint file was created.<br />
<strong>VALUES</strong><br />
<code>+d[(+d)]</code>, specifies the iteration at which the first checkpoint
file is created (first argument) and how frequently checkpoint files are
subsequently created (second optional argument).<br />
<strong>DEFAULT</strong><br />
If the checkpoint value is undefined no checkpoint file is created.<br />
<strong>COMMENTS</strong><br />
The checkpoint files are named OUTFILE.Z.chk where outfile is the name
specified by the <code>outfile</code> variable and Z is the number of the
checkpoint file (the first checkpoint file is labeled 1, the second 2,
and so on). To resume execution of the program starting from a
checkpoint use the BPP <code>–resume</code> switch followed by the checkpoint file
name. For example, <code>bpp –resume outrun1.1.chk</code>.<br />
<strong>EXAMPLES</strong></p>
<pre><code>checkpoint = 100000
checkpoint = 100000 10000
</code></pre>
<h3 id="30-constraintfile">30 constraintfile</h3>
<hr />
<pre><code>constraintfile = s
</code></pre>
<p><strong>DESCRIPTION</strong><br />
Specifies the name of a file containing information for placing
topological constraints on inferred trees and specifying outgroups for
tree rooting.<br />
<strong>VALUES</strong><br />
<code>s</code>, a string specifying the name of the constraint file.<br />
<strong>DEFAULT</strong><br />
If <code>constraintfile</code> is undefined no constraints or outgroups are used in
the analysis.<br />
<strong>EXAMPLES</strong></p>
<pre><code>constraintfile = myconstraints.txt
</code></pre>
<h3 id="31-threads">31 threads</h3>
<hr />
<pre><code>threads = +d [+d +d]
</code></pre>
<p><strong>DESCRIPTION</strong><br />
Specifies the number of CPU threads to be used and optionally the
specific threads to be used.<br />
<strong>VALUES</strong><br />
<code>+d [+d +d\]</code>, either 1 or 3 positive integers. A single integer
argument specifies the number of threads to use. If three integers are
given the first specifies the number of threads, the second specifies
the index of the first thread used and the third specifies the
increment. For example, <code>12 4 1</code> specifies that 12 threads should be
used, the first with index 4 and incrementing by one so that threads
4-15 are used.<br />
<strong>DEFAULT</strong><br />
If <code>threads</code> is undefined, the calculations for each locus are placed on
different threads when possible.<br />
<strong>DEPENDENCIES</strong><br />
The number of specified threads cannot exceed the number of loci
specified.<br />
<strong>EXAMPLES</strong></p>
<pre><code>threads = 4
threads = 8 4 1
</code></pre>
<h2 id="example-control-file">Example control file</h2>
<p>To examine the structure of a typical BPP control file, we consider the
example file <code>A00.bpp.ctl</code> contained in the BPP distribution
subdirectory <code>examples/frogs</code>. The goal here is to provide an example of a
correctly formatted control file and briefly explain the syntax and
meaning of specified variables, as well as proposing some conventions
for writing optional comments in control files. An exhaustive
description of all control file variables, their effects, and
permissible values is provided in the definitions of
<a href="#bpp-control-file-variables">BPP Control File Variables</a>. 
In depth descriptions of specific model and prior choices can be found in
<a href="#substitution-models">Substitution Models</a>, <a href="#introgression-and-migration-models">Introgression and Migration Models</a>, and <a href="">Isolation with Migration Models</a>. 
Control files illustrating the 4 methods of analysis A00, A01, A10 and A11 are found in
<a href="#methods-of-analysis">Methods of Analysis</a>. The content of the control file <code>A00.bpp.ctl</code> is displayed below:</p>
<pre><code>    seed =  -1

    seqfile = frogs.txt
    Imapfile = frogs.Imap.txt
    outfile = out.txt
    mcmcfile = mcmc.txt

    # fixed species tree and delimitation
    speciesdelimitation = 0 
    speciestree = 0

    species&amp;tree = 4  K  C  L  H
                   9  7 14  2
                   (((K, C), L), H);

    # sequence data are unphased for all 4 populations
    phase =   1  1  1  1

    # 0: no data (prior); 1:seq like
    usedata = 1  

    # number of data sets in seqfile
    nloci = 5  

    # remove sites with ambiguity data (1:yes, 0:no)?
    cleandata = 0    

    # invgamma(a, b) for theta
    thetaprior = invgamma 3 0.004 E  

    # invgamma(a, b) for root tau &amp; Dirichlet(a) for other tau's
    tauprior = invgamma 3 0.002    

    *     heredity = 1 4 4
    *     locusrate = 1 5

    # finetune for GBtj, GBspr, theta, tau, mix, locusrate, seqerr
    finetune =  1: 5 0.001 0.001  0.001 0.3 0.33 1.0  

    # MCMC samples, locusrate, heredityscalars, Genetrees
    print = 1 0 0 0   

    burnin = 8000

    sampfreq = 2

    nsample = 100000
</code></pre>
<p><strong>A comment about comments</strong><br />
Optional comments may be added to the control file. If either a '#' or
a '*' symbol occurs on a line, everything following the symbol up to
the end of the line will be ignored by the BPP program. This allows
optional comments to be added by the user either to explain choices (or
meanings) of variables, or to retain optional alternative variable
settings for future use. The control files presented in this
documentation use the different comment-initiating symbols for distinct
purposes:</p>
<ol>
<li>
<p>'#' initializes a comment that lists and/or explains the possible
    settings of the variable found immediately below the comment.</p>
</li>
<li>
<p>'*' initializes a comment that retains deactivated variable
    settings (for future use). If an alternative to a defined variable
    it follows that variable.</p>
</li>
</ol>
<p>It is good practice to make use of these conventions for comments in
your own bpp control files.</p>
<h3 id="control-file-variables-explained">Control file variables explained</h3>
<p>Here we walk through the variables in the above control file example
explaining the meaning of each and the valid values that can be
specified. An exhaustive description of all control file variables and
permissible values is given in <a href="#control-file">Control File</a>.</p>
<pre><code>    seed = -1
</code></pre>
<p>This line specifies the seed value used to initialize the random number
generator. Because BPP uses pseudorandom numbers to drive a stochastic
algorithm for inference the results will vary between analyses initiated
with different seeds, but will be identical if the same seed is used
(because pseudorandom numbers are generated using a deterministic
algorithm started with the value of seed). Setting <code>seed = -1</code> specifies
that BPP use the clock on the computer to generate a seed. With this
setting, every run of the program will automatically be started with a
different seed. The seed that was used for a run is stored in a file
generated by the program called SeedUsed.</p>
<pre><code>    seqfile = frogs.txt
    Imapfile = frogs.Imap.txt
    outfile = out.txt
    mcmcfile = mcmc.txt
</code></pre>
<p>These lines specify the names of the files used to run the program and
summarize the output. The program will assume that all these files are
located in the current working directory since no path is specified. The
line <code>seqfile = frogs.txt</code> specifies that the sequence data are in the
file named <code>frogs.txt</code>. See <a href="#sequence-file">Sequence File</a> for a 
description of the sequence data file format.
The line <code>Imapfile = frogs.Imap.txt</code> specifies that the map file is
named <code>frogs.Imap.txt</code>. See <a href="#imap-file">Imap File</a> for
a description of the map file format. The lines <code>outfile = out.txt</code> and
<code>mcmcfile = mcmc.txt</code> specify that the program output and mcmc samples
are printed to files named <code>out.txt</code> and <code>mcmc.txt</code>, respectively. The
format of the output file depends on the type of analysis being
performed and is described in <a href="#methods-of-analysis">Methods of Analysis</a>.</p>
<pre><code>    # fixed species tree
    speciesdelimitation = 0 
    speciestree = 0
</code></pre>
<p>The line <code>speciesdelimitation = 0</code> specifies that the number of species
is fixed (not delimited) and the line <code>speciestree = 0</code> specifies that
the species phylogeny (topology) is fixed (not estimated). This
combination specifies an analysis of type A00, which is parameter
estimation under a MSC using multiple-loci and multiple-species data on
a fixed species tree (<a href="https://doi.org/10.1093/genetics/164.4.1645">Rannala and Yang 2003</a>; <a href="https://doi.org/10.1093/molbev/msn148">Burgess and Yang 2008</a>). Analysis A00 is
described in detail in <a href="#a00-demography-and-divergence-time-estimation">A00: Demography and Divergence Time Estimation</a>.</p>
<pre><code>    species&amp;tree = 4  K  C  L  H
                   9  7 14  2
                   (((K, C), L), H);
</code></pre>
<p>The line <code>species&amp;tree = 4 K C L H</code> specifies the number of
species/populations and their names. This example specifies 4 species in
the data, which are K, C, L and H. The next line <code>9 7 14 2</code> specifies
the maximum number of sequences at any locus for each species/population
(in the same order as the species listing on the previous line). This
example specifies a maximum of 9 sequences for K, 7 for C, 14 for L and
2 for H. The program interprets these numbers as follows. First, if the
number of sequences for a species is 2 or greater, <span class="arithmatex">\(\theta\)</span> for that
species will be a parameter estimated by the program. Second, the sum of
the numbers of sequences over all species specifies the maximum number
of sequences at a locus. However, specifying 2 sequences for a species
when there are actually 10 sequences for that species at some loci in
the data file has no harmful effects. Note that if there are <span class="arithmatex">\(s\)</span> species
in the species tree, the model will minimally contain the following
parameters: <span class="arithmatex">\(s – 1\)</span> species divergence times (<span class="arithmatex">\(\tau\)</span>s) and <span class="arithmatex">\(s – 1\)</span>
ancestral species <span class="arithmatex">\(\theta\)</span>s. A <span class="arithmatex">\(\theta\)</span> parameter will also exist for
each contemporary species with at least two sequences present at some
locus. If the data are for a single species/population, the model will
contain one parameter only, <span class="arithmatex">\(\theta\)</span> for that species.</p>
<p>The line <code>(((K, C), L), H);</code> specifies the (fixed) species tree in
Newick format ending with a semicolon (;). This example specifies a tree
with no introgression (the MSC model). Specification of the MSC-I model
with introgression is different (see
<a href="#introgression-and-migration-models">Introgression and Migration Models</a> for a description of the MSC-I model). The
Newick and extended Newick formats for specifying phylogenetic trees are
described in <a href="#introgression-and-migration-models">Introgression and Migration Models</a> and a good overview is found at
<a href="https://en.wikipedia.org/wiki/Newick_format">https://en.wikipedia.org/wiki/Newick_format</a>.</p>
<pre><code>    # sequence data are unphased for all 4 populations
    phase =   1  1  1  1
</code></pre>
<p>The <code>phase</code> variable is specified using a boolean (0/1) indicator
variable for each species/population, with 0 (the default) specifying
that sequences from that species are fully phased haplotypes and 1
specifying they are unphased diploid sequences. If this line is omitted,
all sequences are assumed to be fully phased (that is, value 0 for all
species). The program does not allow some sequences from a species to be
phased while other sequences from that same species are unphased. If the
variable is 1, each sequence from that species is treated as an unphased
diploid sequence, with heterozygous sites represented using ambiguity
characters <code>YRMKSW</code>. The chararcters <code>N</code>, <code>-</code> and <code>?</code> are allowed in the
sequence file and are all treated as <code>?</code>, but other ambiguities (HBVD)
are not allowed. Each sequence is then expanded/resolved into two
sequences, and the program averages over all the possible phase
resolutions according to the approach of <a href="https://doi.org/10.1038/ng.937">Gronau et al 2011</a>. For example a
sequence with two heterozygous sites R and Y is resolved into two
possible haplotypes: (i) A...T and G...C, and (ii) A...C and G...T.
Note that this option forces a sequence with no heterozygote sites to be
phased/resolved into two identical sequences.</p>
<p>The MSC model averages over different phase resolutions of the
heterozygote sites in the likelihood calculation. This option may
increase the memory usage and CPU time considerably if there are many
sequences with many heterozygote sites at one locus. The simulation
program MCcoal has been modified to simulate diploid unphased data as
well. This allows one to use simulations to examine the difference
between the analysis of the full data and the unphased data. If the
phase flag for a species is 0, all sequences from that species are
treated as resolved haplotype sequences, and ambiguities are interpreted
in the usual way (for example, Y means one nucleotide that is either T
or C).</p>
<pre><code>    # 0: no data (prior); 1:seq like
    usedata = 1  
</code></pre>
<p>The line <code>usedata = 1</code> specifies that the likelihood is used in the MCMC
run to generate the posterior distribution of parameters. Setting this
to 0 allows one to run the MCMC without sequence data to generate the
prior (mostly used for debugging).</p>
<pre><code>    # number of data sets in seqfile
    nloci = 5  
</code></pre>
<p>The line <code>nloci = 5</code> specifies the number of loci (alignments) to be 5.
There may be more loci in the sequence data file than is specified here.
For example, if 200 loci exist in the sequence data file and you specify
<code>nloci = 2</code> in the control file, BPP will read the first
two loci only.</p>
<pre><code>    # remove sites with ambiguity data (1:yes, 0:no)?
    cleandata = 0    
</code></pre>
<p>The line <code>cleandata = 0</code> specifies that columns in the alignment which
have gaps or ambiguity characters are retained and used in the
likelihood calculation. Setting <code>cleandata = 1</code> instead specifies that
such data are removed prior to analysis.</p>
<pre><code>    # invgamma(a, b) for theta
    thetaprior = invgamma 3 0.004 E  
</code></pre>
<p>The line <code>thetaprior = 3 0.004 E</code> specifies the inverse-gamma prior
IG(<span class="arithmatex">\(\alpha, \beta\)</span>) for the <span class="arithmatex">\(\theta\)</span> parameters, with the mean to be
<span class="arithmatex">\(\beta/(\alpha-1)\)</span>. In the example, the mean is <span class="arithmatex">\(0.004/(3 – 1) = 0.002\)</span>
(two substitutions per kb on average) and the option E specifies that
the program analytically integrates over <span class="arithmatex">\(\theta\)</span>s. Note that all
<span class="arithmatex">\(\theta\)</span> parameters in the MSC model (for both modern species and
extinct ancestral species) are assigned the inverse-gamma prior with the
same parameters.</p>
<p><strong>Analytical integration of <span class="arithmatex">\(\theta\)</span> and the inverse gamma prior</strong><br />
The inverse-gamma is a conjugate prior for <span class="arithmatex">\(\theta\)</span> (<a href="https://doi.org/10.1073/pnas.0611164104">Hey and Nielsen 2007</a>)
which means that both the prior and the posterior of <span class="arithmatex">\(\theta\)</span> will be
inverse-gamma. Use of the conjugate priors allows the <span class="arithmatex">\(\theta\)</span>
parameters to be integrated out analytically, and thus the dimension of
the parameter space is reduced. This typically leads to improved mixing
of the MCMC. However, with this option, the posterior of the <span class="arithmatex">\(\theta\)</span>
parameters will not be produced. To estimate the <span class="arithmatex">\(\theta\)</span> parameters,
add the letter e (or E) on the line, as follows</p>
<pre><code>    thetaprior = invgamma 3 0.002 e
</code></pre>
<p>Whether <span class="arithmatex">\(\theta\)</span>s are integrated out or estimated, other results (such
as the posterior of the <span class="arithmatex">\(\theta\)</span> parameters or the posterior probability
of species trees or species-delimitation models) should be identical. A
useful strategy may be to run the A01, A10, and A11 analyses without
estimating the <span class="arithmatex">\(\theta\)</span> parameters, and after the MAP model (the best
species tree or the best species delimitation model) is identified, run
the A00 analysis with the model fixed to estimate all parameters.</p>
<p>Some notes about the inverse-gamma distribution. Since
BPP3.4, both the <span class="arithmatex">\(\theta\)</span> and <span class="arithmatex">\(\tau\)</span> parameters are
assigned the inverse gamma priors rather than the gamma priors in
version 3.3 or earlier. One difference is that the gamma is light-tailed
while the inverse-gamma is heavy-tailed, so that the inverse-gamma may
be less influential than the gamma if your prior mean is much too small.
The inverse-gamma distribution IG(<span class="arithmatex">\(\alpha, \beta\)</span>) has mean
<span class="arithmatex">\(m = \alpha/(\beta-1)\)</span> if <span class="arithmatex">\(\alpha &gt; 1\)</span> and variance
<span class="arithmatex">\(s^2 = \beta^2/[(\alpha-1)^2(\alpha-2)]\)</span> if <span class="arithmatex">\(\alpha &gt; 2\)</span>, while the
coefficient of variation is <span class="arithmatex">\(s/m = \sqrt{1/(\alpha-2)}\)</span>. If little
information is available about the parameters, you can use <span class="arithmatex">\(\alpha = 3\)</span>
for a diffuse prior and then adjust the <span class="arithmatex">\(\beta\)</span> so that the mean looks
reasonable. For example, for the human species,
<span class="arithmatex">\(\theta_H \approx 0.0006\)</span>, which means that two random sequences from
the human population are different at  0.06% of sites, less than 1
difference per kb. A sensible diffuse prior is then
<code>thetaprior = 3 0.002</code>, with mean <span class="arithmatex">\(0.002/(3 – 1) = 0.001\)</span>.</p>
<pre><code>    # invgamma(a, b) for root tau &amp; Dirichlet(a) for other tau's
    tauprior = invgamma 3 0.002    
</code></pre>
<p>The line <code>tauprior = 3 0.002</code> specifies the inverse-gamma prior
IG(<span class="arithmatex">\(\alpha, \beta\)</span>) for <span class="arithmatex">\(\tau_0\)</span>, the divergence time parameter for the
root in the species tree. Other divergence times are generated from the
uniform Dirichlet distribution (<a href="https://doi.org/10.1073/pnas.0913022107">Yang and Rannala 2010, eq.2</a>). In the example, the
mean is <span class="arithmatex">\(0.002/(3 – 1) = 0.001\)</span> (which means 0.1% average sequence
divergence between the root of the species tree and the present time).
If the mutation rate is <span class="arithmatex">\(10^{–9}\)</span> mutations/site/year, this distance
will translate to a divergence time of 1 MY.</p>
<pre><code>    # finetune for GBtj, GBspr, theta, tau, mix, locusrate, seqerr
    finetune =  1: 5 0.001 0.001  0.001 0.3 0.33 1.0  
</code></pre>
<p>The line <code>finetune = 1: 5 0.001 0.001 0.001 0.3 0.33 1.0</code> specifies the
step lengths used in proposals for the parameters of the MCMC algorithm.
These step lengths can affect both mixing and convergence of the MCMC.
The first value, before the colon (1 in the example), is a boolean
switch, with 1 specifying automatic adjustments of step lengths by the
program and 0 specifying no automatic adjustments. Following the colon
are the step lengths for the proposals used in the program. If you
choose to let the program adjust the step lengths, burnin has to be
&gt;200, and the step lengths specified here will then be the initial step
lengths, with the program adjusting/optimizing step lengths using the
information collected during the burnin step. Automatic adjustment works
well in most cases and is usually the best option. Some notes about
manually adjusting the step lengths are provided below in section ?.</p>
<pre><code>    # MCMC samples, locusrate, heredityscalars, Genetrees
    print = 1 0 0 0   

    burnin = 8000

    sampfreq = 2

    nsample = 100000
</code></pre>
<p>The line <code>print = 1 0 0 0</code> specifies the information that is printed out
during the MCMC run. It has several boolean variables (0 for off and 1
for on) that control the printouts during the MCMC. The first variable
specifies whether the MCMC samples are written into the main sample file
(mcmc.txt). The second flag is for locus-specific rate parameters (rate
<span class="arithmatex">\(\mu_i\)</span>, variance parameter <span class="arithmatex">\(\nu_i\)</span> and species-tree branch rates for
locus <span class="arithmatex">\(r_{ij}\)</span>). The third flag is for locus-specific heredity scalars
if those are estimated from the data. The fourth flag is for printing
locus-specific gene trees. The fifth flag is for printing locus-specific
substitution-rate parameters (qmat for the <span class="arithmatex">\(Q\)</span> matrix, freqs for base
frequencies, and alpha for gamma rates for sites).</p>
<p>MCMC samples are taken after the burnin (8000 iterations), and in this
example, are taken every 2 iterations, with a total of 100,000 samples
taken. The total number of MCMC iterations is <code>burnin</code> + <code>sampfreq</code>
<span class="arithmatex">\(\times\)</span> <code>nsample</code>. The resulting file can be large. For analysis A00
(<code>speciesdelimitation = 0, speciestree = 0</code>), this file can be analysed
using the R package or tracer. For other analyses (A01, A10, and A11),
the dimension of the sample is changing and cannot be analysed using
such traceplot programs.</p>
<p><strong>Combining results from different runs</strong><br />
The variable setting <code>print = -1</code> specifies that BPP will
not perform an MCMC analysis and will instead read an existing MCMC
sample file and summarize the results. Thus, with setting <code>print = 1</code>,
the <code>mcmc.txt</code> file receives the MCMC program output, but with
<code>print = -1</code>, it instead provides the program input. When using this
option be careful not to overwrite files that you intended to keep. The
<code>print = -1</code> option is useful for combining MCMC samples from multiple
runs to produce the posterior summary. Suppose you run the same analysis
3 times in different directories -- you can merge the mcmc.txt files
from the three runs into one file and run the program using the
<code>print = -1</code> option to summarize the posterior for the combined sample.
Note that if you forced one or more of the jobs to terminate (using the
Unix kill command for example) the last line of each mcmc.txt file may
incomplete. If that is the case, you should delete the incomplete line.
The header lines of any files being concatenated should also be deleted.
Do not combine the MCMC samples from different types of analyses (i.e.,
using different data files and/or variable settings or priors).</p>
<h2 id="substitution-models">Substitution Models</h2>
<p>BPP includes several models that allow more realistic representations of
the processes underlying the evolution of DNA and amino acid sequences.
Fundamental to all molecular phylogeographic analyses is the model of
the nucleotide or amino acid substitution process itself. DNA
substitution models can accomodate different rates of substitution
between nucleotides due to factors such as transition versus
transversion bias and unequal stationary nucleotide frequencies. BPP
allows the use of many different DNA base substitution and amino acid
substitution models. For DNA sequences, the models range from a simple
Jukes-Cantor model (computationally efficient) to a sophisticated
General Time-Reversible (GTR) model (computationally expensive but more
realistic). For closely related species the Jukes-Cantor model is often
adequate, while the GTR often performs better for more distantly related
species. The remaining DNA substitution models (K80, F81, HKY, T92,
TN93, and F84) are all special cases (submodels) of the GTR model and
are of intermediate complexity, usually having one or two free
parameters (<a href="https://global.oup.com/academic/product/molecular-evolution-9780199602605?lang=es&amp;cc=gb">see Yang 2014</a>). Many fixed substitution matrix models (no
free parameters) for modeling substitutions among amino acid sequences
are also available.</p>
<p>Another factor to consider is variation in overall (or average)
substitution rates. The overall rate of substitution can vary in
multiple ways and, based on empirical evidence, this type of variation
is one of the most important factors to accommodate during
phylogeographic inference. There are 3 distinct types of substitution
rate variation that can be accommodated using the models implemented in
BPP: rate variation among sites, rate variation among loci, and rate
variation among species (or populations). Substitution rate variation
among sites, or among loci, is a common pattern and is likely due to the
pervasive effects of negative (purifying) selection. Variation of
substitution rates among species (or populations) is more often observed
among distantly related species and may reflect the effects of
differential selection, evolution of DNA repair mechanisms, or other
factors.</p>
<p>This Chapter briefly describes the models of DNA substitution, and of
substitution rate variation, implemented in BPP, the parameters and
priors associated with these models, and how to specify the different
models in the BPP control file.</p>
<h3 id="dna-substitution-models">DNA Substitution Models</h3>
<h4 id="jukes-cantor-model">Jukes-Cantor model</h4>
<p>One of the earliest and simplest models of DNA substitution, the
Jukes-Cantor model, assumes that all possible changes between the 4
nucleotides have equal rates and thus the stationary frequencies of
bases are <span class="arithmatex">\(1/4\)</span> each. This model is the default in BPP and will be used
if no model is specified in the control file. The control file variable
<code>model</code> specifies the substitution model and the explicit specification
of the JC69 model is:</p>
<pre><code>    model = JC69
</code></pre>
<p>The JC69 model appears exceedingly unrealistic. However, for closely
related sequences (with few multiple substitutions) it often performs as
well as more sophisticated models while incurring much less
computational expense. There are no parameters associated with the JC69
model (other than the substitution rate) so no additional control file
variables associated with the JC69 substitution model need to be
specified.</p>
<h4 id="gtr-model">GTR model</h4>
<p>The General Time-Reversible (GTR) model implemented in BPP allows
different rates of substitution between different nucleotide pairs, with
the constraint that the process is "reversible" so that, for example,
the rate of transition from A to T equals the rate of transition from T
to A and so on. The GTR model accommodates factors such as
transition-transversion bias and should be used when sequences are more
divergent (distantly related species). The control file specification of
the GTR model is:</p>
<pre><code>    model = GTR
</code></pre>
<p>Note that if the GTR model is used, two other variables may be specified
in the control file: <code>Qrates</code> and <code>basefreqs</code>. The <code>Qrates</code> variable
specifies whether the exchangeability parameters of the GTR model are
fixed or variable as well as either their fixed values, or prior mean
values, respectively. The control file specification is:</p>
<pre><code>    Qrates = B a b c d e f
</code></pre>
<p>The variable <code>B</code> is either 0 (variable exchangeability parameters) or 1
(fixed exchangeability parameters). The variables <code>a b c d e f</code> are
positive numbers specifying the exchangeability parameters for TC, TA,
TG, CA, CG and AG substitutions respectively (see <a href="https://doi.org/10.1007/BF00178256">Yang 1994</a>). Note that
TC is the rate of either T-&gt;C or C-&gt;T substitutions. Note that only
the relative rates matter because the rate matrix is scaled to have an
overall average substitution rate of 1. Thus, for the model of fixed
rates if x is any positive number a b c d e f and xa xb xc xd xe xf are
equivalent specifications. An example with fixed rates is as follows:</p>
<pre><code>    Qrates = 1  2 1 1 1 1 2
</code></pre>
<p>which specifies <span class="arithmatex">\(a=2,b=c=d=e=1,f=2\)</span> so that the transition rate is twice
as high as the transversion rate, corresponding to a K80 or HKY model.
In most cases, users will want to use the GTR model with exchangeability
parameters variable. Using this specification, the posterior
distribution of the parameters will be estimated from the data. An
example with variable rates is as follows:</p>
<pre><code>    Qrates = 0  10 5 5 5 5 10
</code></pre>
<p>The absolute magnitude of the numbers <em>does matter</em> in this case as it
determines the variance of the prior probability distribution on
exchangeability parameters. The variance is inversely proportional to
the absolute magnitude of the variables. The prior on the free
parameters <span class="arithmatex">\(a, b, c, d, e\)</span> is a Dirichlet distribution with parameters</p>
<p><span class="arithmatex">\(\alpha_a = 10, \alpha_b = \alpha_c = \alpha_d = \alpha_e = 5, \texttt{and} \, \alpha_f = 10\)</span></p>
<p>The rates generated from the Dirichlet sum to 1 and are rescaled. If we
define</p>
<p><span class="arithmatex">\(\alpha = \alpha_a + \alpha_b + \alpha_c + \alpha_d + \alpha_e + \alpha_f\)</span></p>
<p>then the mean for <span class="arithmatex">\(a\)</span> is <span class="arithmatex">\(\alpha_a/\alpha\)</span> and the variance for <span class="arithmatex">\(a\)</span> is</p>
<p><span class="arithmatex">\(\textrm{Var}(a) = \frac{\frac{\alpha_a}{\alpha}\left(1-\frac{\alpha_a}{\alpha}\right)}{\alpha + 1}\)</span>.</p>
<p>The means and variances of the other parameters are the same
(substituting subscripts). Thus, in this example, the mean for <span class="arithmatex">\(a\)</span> is
<span class="arithmatex">\(\alpha_a/\alpha = 10/40 = 1/4\)</span> and the variance is
<span class="arithmatex">\((1/4)(3/4)/41 = 4.6 \times 10^{-3}\)</span>. The prior specified above has an
average transition-transversion rate ratio of 2 (if base frequencies are
all fixed at 1/4).</p>
<p>The other control file variable that must be specified when using the
GTR model is the <code>basefreqs</code> variable which specifies whether the
stationary nucleotide (base) frequencies are fixed, or variable, and
either their fixed values, or relative prior means, respectively. The
control file specification is:</p>
<pre><code>    basefreqs = B T C A G
</code></pre>
<p>where B is either 0 (variable base frequencies) or 1 (fixed base
frequencies) and T C A and G are numbers specifying the frequencies of
bases T, C, A and G, respectively. An example with fixed base
frequencies is as follows:</p>
<pre><code>    basefreqs = 1  0.1 0.2 0.2 0.5
</code></pre>
<p>In most cases, users will want to use the variable base frequencies
setting which specifies that the posterior distribution of base
frequencies is estimated from the data. An example with variable base
frequencies is as follows:</p>
<pre><code>    0  10 10 10 10
</code></pre>
<p>Here the 4 variables T C A G specify the alpha parameters of a Dirichlet
prior distribution on the nucleotide frequencies. So, the prior mean
frequency of base A in this example is <span class="arithmatex">\(10/40 = 1/4\)</span> and the variance is
<span class="arithmatex">\((1/4)(3/4)/41 = 4.6 \times 10^{-3}\)</span>. As noted above when discussing the
Dirichlet prior for exchangeability parameters the absolute magnitude of
the variables determines the variance of the Dirichlet prior. If a GTR
model is chosen and <code>Qrates</code> and <code>basefreqs</code> are not specified in the
control file default values of <code>Qrates = 0 1 2 1 1 2 1</code> and
<code>basefreqs = 0 1 1 1 1</code> will be used.</p>
<h4 id="other-dna-substitution-models">Other DNA substitution models</h4>
<p>A range of models are available that fall between the JC69 model and the
GTR model in terms of complexity and number of parameters. The
additional models available include: K80, F81, HKY, T92, TN93, and F84.
The priors on the parameters of these models are pre-specified by the
program.</p>
<h4 id="amino-acid-substitution-models">Amino acid substitution models</h4>
<p>There are many amino acid substitution models available. These models
completely specify the substitution matrix so there are no free
parameters. Many of the models are described in <a href="https://global.oup.com/academic/product/molecular-evolution-9780199602605?lang=es&amp;cc=gb">Yang 2014</a>. The available models are: DAYHOFF, LG, DCMUT, JTT, MTREV, WAG, RTREV, CPREV,
VT, BLOSUM62, MTMAM, MTART, MTZOA, PMB, HIVB, HIV,W, JTTDCMUT, FLU, and
STMTREV.</p>
<h3 id="custom-among-locus-model-variation">Custom among-locus model variation</h3>
<p>If a single substitution model is specified, all loci are assigned that
model. The control file option <code>model = Custom filename</code> can be used to
assign different models among loci, where the models are specified in
the file <code>filename</code>. This file should contain one or more lines with the
format:</p>
<pre><code>loci_indices data_type model
</code></pre>
<p>where <code>loci_indices</code> is an index or range of indexes for loci (indexed
according to their order of occurrence in the sequence data file),
<code>data_type</code> is either <code>DNA</code> or <code>AA</code>, and model specifies the
substitution model variable for the indexed loci as described above. For
example, the entry</p>
<pre><code>1-10 DNA JC69
11 AA DAYHOFF
</code></pre>
<p>specifies that loci 1 to 10 comprise DNA sequences that will have the
<code>JC69</code> model applied to them, while locus 11 is an amino acid sequence
that will have the <code>DAYHOFF</code> model applied to it.</p>
<h2 id="models-of-substitution-rate-variation">Models of Substitution Rate Variation</h2>
<h3 id="among-site-rate-variation">Among-site rate variation</h3>
<p>Substitution rates vary among sites due to many factors, including
purifying selection in coding regions. BPP allows among-site rate
variation to be accomodated using the canonical Gamma distributed rate
variation model (<a href="https://doi.org/10.1007/BF00178256">see Yang 1994</a>). In this model, the mean rate among
sites is constrained to be 1 so that the expected number of
substitutions on a branch is proportional to its length (in units of
expected substitutions). If the mean of the Gamma distribution is fixed,
the only free parameter is the shape parameter <span class="arithmatex">\(\alpha\)</span>. The variance of
rates among sites is <span class="arithmatex">\(1/\alpha\)</span> and therefore a smaller <span class="arithmatex">\(\alpha\)</span>
specifies more rate variation. Gamma distributed rate variation is
specified using the control file variable <code>alphaprior</code>. If this variable
is not set the default is <span class="arithmatex">\(\alpha = \infty\)</span> so that the variance is zero
(rates are constant among sites). The prior on <span class="arithmatex">\(\alpha\)</span> is also a Gamma
distribution with parameters <span class="arithmatex">\(\alpha_p\)</span> and <span class="arithmatex">\(\beta_p\)</span>. The control file
specification is as follows:</p>
<pre><code>alphaprior = a b ncatG
</code></pre>
<p>where <span class="arithmatex">\(a = \alpha_p\)</span> and <span class="arithmatex">\(b = \beta_p\)</span> are the parameters of the Gamma
distributed prior on the parameter <span class="arithmatex">\(\alpha\)</span> and <code>ncatG</code> is the number of
rate categories used to approximate the Gamma distribution. A discrete
distribution is used to efficiently approximate the Gamma distribution
and ncatG is the number of rate categories used in the approximation --
more rate categories provide a better approximation but incur greater
computational expense. A value for ncatG of 4 is recommended. The
variance and mean of the prior on <span class="arithmatex">\(\alpha\)</span> are: </p>
<p><span class="arithmatex">\(\begin{aligned}
  \textrm{Mean}(\alpha) &amp; = &amp; \frac{\alpha_p}{\beta_p}, \nonumber \\
  \textrm{Var}(\alpha) &amp; = &amp; \frac{\alpha_p}{\beta_p^2} \nonumber\end{aligned}\)</span></p>
<p>As an example, the setting <code>alphaprior = 1 1 4</code> specifies a prior on the
shape parameter <span class="arithmatex">\(\alpha\)</span> (with 4 rate categories) that has a mean and
variance of 1, thus the expected variance of <span class="arithmatex">\(\alpha\)</span> is also 1. This is
a fairly diffuse (uninformative) prior.</p>
<h3 id="among-locus-rate-variation">Among-locus rate variation</h3>
<p>Empirical evidence suggests that substitution rates can vary
substantially among loci, probably reflecting the influences of
selection. BPP incorporates a hierarchical prior to model rate variation
among loci. The models implemented in BPP allow each locus to have a
different substitution rate. For <span class="arithmatex">\(L\)</span> loci (alignments), there are <span class="arithmatex">\(L\)</span>
substitution rate parameters, where <span class="arithmatex">\(\mu_i\)</span> is the rate at the <span class="arithmatex">\(i\)</span>th
locus.</p>
<h4 id="prior-distribution-of-the-mean-rate-among-loci">Prior distribution of the mean rate among loci</h4>
<p>The mean rate across loci is</p>
<p><span class="arithmatex">\(\bar{\mu} = \frac{1}{L} \sum_{i=1}^L \mu_i\)</span>.</p>
<p>The prior distribution for the mean rate across loci <span class="arithmatex">\(\bar{\mu}\)</span> is a Gamma distribution with
parameters <span class="arithmatex">\(\alpha_{\bar{\mu}}\)</span> and <span class="arithmatex">\(\beta_{\bar{\mu}}\)</span> and mean and
variance, </p>
<p><span class="arithmatex">\(\begin{aligned}
  \textrm{Mean}(\bar{\mu}) &amp; = &amp; \frac{\alpha_{\bar{\mu}}}{\beta_{\bar{\mu}}}, \nonumber \\
  \textrm{Var}(\bar{\mu}) &amp; = &amp; \frac{\alpha_{\bar{\mu}}}{\beta_{\bar{\mu}}^2}. \nonumber\end{aligned}\)</span></p>
<p>If there are no fossil calibrations in the species tree, the rates
should all be relative. In this case, to avoid overparameterization we
suggest fixing the mean substitution rate across loci to be
<span class="arithmatex">\(\bar\mu = 1\)</span>. This is done by setting
<span class="arithmatex">\(\alpha_{\bar{\mu}} = \beta_{\bar{\mu}} = \infty\)</span> (see discussion of
control file settings below), or setting both these parameters to large
values such as <span class="arithmatex">\(\alpha_{\bar{\mu}} = \beta_{\bar{\mu}} = 100\)</span>, in which
case <span class="arithmatex">\(\bar{\mu}\)</span> is nearly fixed.</p>
<h4 id="prior-distribution-for-locus-specific-rates-given-mean-rate-among-loci">Prior distribution for locus-specific rates given mean rate among loci</h4>
<p>Conditioning on the mean rate <span class="arithmatex">\(\bar\mu\)</span>, two priors are available to
specify the probability distribution of locus-specific rates, <span class="arithmatex">\(\mu_i\)</span>.
The control file variable argument <code>prior</code> (see discussion of control
file settings below) can take two values <code>dir</code> (for Gamma-Dirichlet
rates for loci), and <code>iid</code> (for conditional i.i.d. or hierarchical
prior). If <code>prior = dir</code> (Gamma-Dirichlet distribution of locus rates),
the total rate <span class="arithmatex">\(L\bar\mu\)</span>, given the mean rate (<span class="arithmatex">\(\bar\mu\)</span>), is
partitioned into locus rate variables , <span class="arithmatex">\(\mu_i\)</span>, that are constrained to
sum to <span class="arithmatex">\(L\bar\mu\)</span> and have a Dirichlet distribution with concentration
parameter <span class="arithmatex">\(\alpha_\mu\)</span>. The model and notation follow 
<a href="https://doi.org/10.1093/molbev/msn148">Burgess and Yang 2008</a> eq.4
and <a href="https://doi.org/10.1093/sysbio/syu020">Dos Reis et al 2014</a> eqs.3-5. 
If <code>prior = iid</code> (Conditional-i.i.d. or hierarchical prior of locus rates), the locus
rates, <span class="arithmatex">\(\mu_i\)</span>, are independent and identically distributed i.i.d. given
the mean rate <span class="arithmatex">\(\bar\mu\)</span>, so the prior distribution of <span class="arithmatex">\(\mu_i\)</span> is:</p>
<p><span class="arithmatex">\(\mathtt{P}(\mu_i) = \textrm{Gamma}(\alpha_{\mu_i}, \alpha_{\mu_i/\bar{\mu}})\)</span>.</p>
<p>This model is described in <a href="https://doi.org/10.1093/sysbio/syu109">Zhu et al 2015</a> eq.8 and implemented in
[mcmctree]{.smallcaps}. In both the Gamma-Dirichlet and the Conditional
i.i.d. models, the parameter <span class="arithmatex">\(\alpha_{\mu_i}\)</span> is inversely related to
the extent of rate variation among loci, with a large <span class="arithmatex">\(\alpha_{\mu_i}\)</span>
meaning similar rates among loci (less rate variation). If all loci are
noncoding, the rates are probably similar, so <span class="arithmatex">\(\alpha_{\mu_I} =\)</span> 10 or
20 may be reasonable, while for coding loci or exons,
<span class="arithmatex">\(\alpha_{\mu_i} = 2\)</span> or 1 may be appropriate. The <span class="arithmatex">\(\alpha_{\mu_i}\)</span>
parameter may affect the estimates of the population size parameter
(<span class="arithmatex">\(\theta\)</span>) for the root node on the species tree (<a href="https://doi.org/10.1093/molbev/msn148">Burgess and Yang 2008</a>). The <span class="arithmatex">\(L\)</span>
locus rates (<span class="arithmatex">\(\mu_i\)</span>) are parameters in the model. If
<span class="arithmatex">\(\alpha_{\bar{\mu}} &gt; 0\)</span>, the mean rate, <span class="arithmatex">\(\bar\mu\)</span>, is a parameter as
well.</p>
<h4 id="control-file-settings-for-locusrate">Control file settings for locusrate</h4>
<p>The <code>locusrate</code> variable takes from 1 and 5 arguments. In the case of 5
arguments the form is shown below, with the last argument (prior) being
optional.</p>
<pre><code>locusrate = rate_variation a_mubar b_mubar a_mui prior
</code></pre>
<p>The rate_variation variable is either 0 (no rate variation, the
default), 1 (rate variation with rates inferred) or 2 (rate variation
with fixed user-specified rates). If rate_variation = 0 no other
variables are specified, if rate_variation = 2 then a second argument
specifies the name of a file containing <span class="arithmatex">\(L\)</span> specified fixed rates, one
for each locus. If rate_variation = 1 the additonal variables <code>a_mubar</code>
and <code>b_mubar</code> specify the shape and rate parameters
(<span class="arithmatex">\(\alpha_{\bar{\mu}}\)</span> and <span class="arithmatex">\(\beta_{\bar{\mu}}\)</span>) of the gamma distribution
that is the prior for the mean rate across loci (<code>mubar</code> or <span class="arithmatex">\(\bar\mu\)</span>),
and variable <code>a_mui</code> (<span class="arithmatex">\(\alpha_{\mu_i}\)</span>) and <code>prior</code> are used to specify
parameter of the prior on the locus rates (<span class="arithmatex">\(\mu_i\)</span>) given the mean rate
(<span class="arithmatex">\(\bar\mu\)</span>) and the form of that prior (<code>dir</code> or <code>iid</code>). Some example
configuration file settings are:</p>
<pre><code>#  (0: No rate variation among loci.  This is default)
locusrate = 0
# (1: estimate locus rates mui)
locusrate = 1 10 10 5.0 iid
# (1: estimate locus rates mui)
locusrate = 1  0  0 5.0 iid
# (2: locus rates from file)
locusrate = 2 LocusRateFileName
</code></pre>
<p>Note that <code>locusrate = 2 LocusRateFileName</code> specifies the fixed-rates
model of locus-rate variation (<a href="https://doi.org/10.1093/molbev/msn148">Burgess and Yang 2008</a>). This is the strategy used
by <a href="https://doi.org/10.1093/genetics/162.4.1811">Yang 2002</a>, with the relative rates estimated by the distance to an
outgroup species. The relative locus rates are listed in the file: there
should be as many numbers in the file, separately by spaces or line
returns, as the number of loci (<code>nloci</code>). The program re-scales those
rates so that the average across all loci is 1 and then use those
relative rates as fixed constants. The following example nearly fixes
(first line), or fixes (second line) the mean rate <span class="arithmatex">\(\bar{\mu}\)</span>. This is
the recommended practice when no fossil calibrations are available:</p>
<pre><code># locusrate = 1 a_mubar b_mubar a_mui &lt;prior&gt;
# (1: estimate locus rates mui)
locusrate = 1 10 10 5 iid
# (1: estimate locus rates mui, mubar=1 fixed)
locusrate = 1  0  0 5 iid
</code></pre>
<p>Note that <code>a_mubar = 0</code> and <code>b_mubar = 0</code> (with 0 meaning <span class="arithmatex">\(\infty\)</span>)
cause the program to fix <span class="arithmatex">\(\bar{\mu}\)</span>, while equal and large values such
as <code>a_mubar = 100</code> and <code>b_mubar = 100</code> cause cause <span class="arithmatex">\(\bar\mu\)</span> to be
nearly fixed at 1.</p>
<p><strong>The locus-rate prior in IMa.</strong> The model of variable rates among loci
implemented here has some differences from a similar model implemented
in the IMa program <a href="https://doi.org/10.1534/genetics.103.024182">Hey and Nielsen 2004</a>. The biggest difference
appears to be the parameterization. BPP defines mutation
rate on a per-nucleotide basis, so the prior specifies that the
expectation of the mutation rate per site is constant among loci.
IMa defines mutation rate on a per-locus basis, so its
prior specifies that the expectation of the mutation rate per locus is
the same among loci. If locus one has 100 sites and locus two has 1000
sites, then IMa assumes that the per-site rate for locus
one is 10 times for locus two, while BPP assumes the same
per-site rate. Also IMa constrains the geometric mean of
rates across loci to be one, while BPP constrains their
arithmetic mean to be one.</p>
<h3 id="among-species-rate-variation">Among-species rate variation</h3>
<p>The strict molecular clock model of evolution assumes that rates of
substitution for a particular gene, or genomic region, are the same in
different species. Early applications of relative rates tests suggested
that this assumption is often violated in empirical data, probably due
to factors such as selection and in more distantly related species
evolving DNA repair and proof-reading mechanisms.</p>
<p>One can allow rates of substitution to vary in an arbitrary way be
inferring unrooted phylogenetic trees but this does not allow species
divergence times to be estimated. Relaxed clock models allow the rate of
substitution to vary among contemporary and ancestral species but with
some constraints which allow species divergence times to be estimated.
The BPP program implements two relaxed clock models, as well as a strict
molecular clock model. The two relaxed clock models differ in terms of
their assumptions about rates in ancestor and descendent species. The
autocorrelated rates model assumes that descendents have rates that are
correlated with the rate in the ancestor, while independent rates model
assumes that the rates on each branch are iid from a common
distribution. For the independents rates model, the variance of the
distribution of rates among branches determines the degree of departure
from a molecular clock - a low variance means that the rates are almost
clocklike (vary little among branches).</p>
<h4 id="variation-in-the-degree-of-departure-from-a-molecular-clock-among-loci">Variation in the degree of departure from a molecular clock among loci</h4>
<p>For a given set of species some loci may have rates that are highly
clock-like while other loci may have much greater variation in
substitution rates among species (branches of the species tree). To
accomodate this, the variance of the substitution rate among species is
allowed to vary among loci. The prior distribution of the average
variance <span class="arithmatex">\(\bar{\nu}\)</span> for all loci is</p>
<p><span class="arithmatex">\(P(\bar{\nu}) = \textrm{Gamma}(\alpha_{\bar{\nu}},\beta_{\bar{\nu}})\)</span>,</p>
<p>so that the expected average variance is
<span class="arithmatex">\(\alpha_{\bar{\nu}}/\beta_{\bar{\nu}}\)</span>. If this value is larger loci
depart more from the molecular clock on average. Conditional on the
average variance parameter <span class="arithmatex">\(\bar{\nu}\)</span> the prior for the variance
<span class="arithmatex">\(\nu_i\)</span> at locus <span class="arithmatex">\(i\)</span> can be either a Gamma-Dirichlet distribution or a
Conditional i.i.d distribution (controlfile variables <code>dir</code> and <code>iid</code>,
respectively). Both distributions have a parameter <span class="arithmatex">\(\alpha_{\nu_i}\)</span> that
determines the variance of the variance in rates across loci. The
specification of the distribution of <span class="arithmatex">\(\nu_i\)</span> given the mean <span class="arithmatex">\(\bar\nu\)</span>
follows the same procedure as the specification of the distribution of
the locus-specific substitution rate <span class="arithmatex">\(\mu_i\)</span> given the mean rate
<span class="arithmatex">\(\bar\mu\)</span> in the previous section; see notes above about the locusrate
variable. Note that <span class="arithmatex">\(\alpha\)</span> (control file variable <code>a_vi</code>) is inversely
related to the variance in <span class="arithmatex">\(\nu_i\)</span>: use small values of <span class="arithmatex">\(\alpha\)</span> (2, 1,
or 0.5) if you believe that <span class="arithmatex">\(\nu_i\)</span> varies among loci (meaning that the
clock nearly holds at some loci but is seriously violated at others).</p>
<h4 id="relaxed-clock-with-iid-rates-among-branches">Relaxed clock with i.i.d rates among branches</h4>
<p>The independent rates model of the relaxed molecular clock assumes that
for locus <span class="arithmatex">\(i\)</span> each branch has a rate that is independent and identically
distributed (i.i.d). The distribution has mean <span class="arithmatex">\(\mu_i\)</span> specified by the
<code>locusrate</code> model and variance <span class="arithmatex">\(\nu_i\)</span> specified by the prior
distribution for rates among loci discussed above. Two distributions are
available: a Gamma distribution or a log-normal distribution (control
file variables <code>LN</code> and <code>G</code>, respectively). Given the locus rate <span class="arithmatex">\(\mu_i\)</span>
(specified using the <code>locusrate</code> variable) and the variance <span class="arithmatex">\(\nu_i\)</span>
(specified using the <code>clock</code> variable) for locus <span class="arithmatex">\(i\)</span>, the different
lineages may have different rates at the locus, and those rates are
independent among loci. If distribution = G (for gamma), the rate for
species-tree branch <span class="arithmatex">\(j\)</span> at locus <span class="arithmatex">\(i\)</span> has the following gamma
distribution</p>
<p><span class="arithmatex">\(r_{ij} | \mu_i, \nu_i \sim G(\mu_i^2/\nu_i, \mu_i/\nu_i)\)</span>.</p>
<p>This has mean <span class="arithmatex">\(\mu_i\)</span> and variance <span class="arithmatex">\(\nu_i\)</span>.</p>
<p>Alternatively if distribution = <code>LN</code> (for log-normal), the rate for
species-tree branch <span class="arithmatex">\(j\)</span> at locus <span class="arithmatex">\(i\)</span> has the following the log-normal
distribution <span class="arithmatex">\(<span class="arithmatex">\(r_{ij} | \mu_i, \nu_i \sim \mathrm{LN}(\mu_i, \nu_i),\)</span>\)</span>
where <span class="arithmatex">\(\mu_i\)</span> is the mean of the LN distribution and <span class="arithmatex">\(\nu_i\)</span> is the
variance parameter of the lognormal.</p>
<h4 id="relaxed-clock-with-autocorrelated-rates-among-branches">Relaxed clock with autocorrelated rates among branches</h4>
<p>The autocorrelated-rates model (control file option <code>clock = 3</code>) with
the log-normal distribution (LN) specifies the geometric Brownian motion
model of <a href="https://doi.org/10.1080/10635150701420643">Rannala and Yang 2007</a>. 
This assigns a rate to each species-tree branch,
that is, to the mid-point of the branch. Given the rate at the
species-tree root (<span class="arithmatex">\(\mu_i\)</span> at locus <span class="arithmatex">\(i\)</span>), the rates for the two branches
around the root are specified. Then given the rate for each ancestral
branch, the rates for its two daughter branches are specified, by
integrating over the rate at the internal node that is ancestral to the
daughter branches. See figure 1 and equations 3-8 in <a href="https://doi.org/10.1080/10635150701420643">Rannala and Yang 2007</a>. The
rates for all species-tree branches are thus assigned through a
pre-order tree traversal, starting from the root moving to the tips,
until all branches are visited.</p>
<p>If the gamma distribution (G) is instead used the model works as
follows, using a similar pre-order tree traversal. First the two
branches at the species-tree root have the gamma distribution with mean
<span class="arithmatex">\(\mu_i\)</span> (<code>mu_i</code>) and variance <span class="arithmatex">\(\nu_i\)</span> (<code>nu_i</code>). Then given the rate for
each ancestral branch, the rates for its two daughter branches are
specified as independent gamma variables with the mean to be the rate of
the parental branch and with the variance to be <span class="arithmatex">\(\nu_i\)</span>. The
autocorrelated relaxed clock model is only implemented for the MSC model
and is not available when using the MSC-I model.</p>
<p><strong>Rates vary among species tree branches</strong>\
The rate-drift model specifies rates for branches on the species tree
(rather than on the gene tree) for each locus, and gene-tree branches
residing in the same population or species have the same rate. For
example, if all sequences at a locus are from the same species and all
coalescent events occur in that species (before reaching an ancestral
species), all branches on the gene tree will have the same rate even if
the relaxed-clock model allows rates to vary among species. In contrast,
if a gene-tree branch passes several species or populations, the
different segments of the branch will have different rates. In that
case, the branch length on the gene tree is the sum of the lengths of
those segments (with the length of each segment being the product of the
rate and the time duration for the segment).</p>
<h4 id="control-file-settings-for-clock">Control file settings for clock</h4>
<p>The <code>clock</code> variable takes 6 arguments when using a relaxed clock as
follows:</p>
<pre><code>clock = clock_type a_vbar b_vbar a_vi prior dist
</code></pre>
<p>the argument <code>clock_type</code> takes on of three values: 1 (strict clock,
default), 2 (relaxed clock with i.i.d rates) and 3 (relaxed clock with
autocorrelated rates). In the case of the strict clock there are no
additional arguments. The next two variables <code>a_vbar</code> and <code>b_vbar</code>
specify the parameters (<span class="arithmatex">\(\alpha_{\bar{\nu}}\)</span> and <span class="arithmatex">\(\beta_{\bar{\nu}}\)</span>) of
the Gamma prior specifying the average rate variance across loci (see
above). The variable <code>a_vi</code> specifies the variance parameter of the
prior on the locus specific variance, <span class="arithmatex">\(\alpha_{\nu_i}\)</span> (see above). The
variable <code>prior</code> is either <code>iid</code> or <code>dir</code> and specifies the prior on
<span class="arithmatex">\(\nu_i\)</span> given <span class="arithmatex">\(\bar{\nu}\)</span> (see above). The variable <code>dist</code> is either
<code>LN</code> or <code>G</code> specifying either a Gamma distribution of a log-normal
distribution for the branch rates conditional on the locus rate <span class="arithmatex">\(\mu_i\)</span>
and among-branch variance <span class="arithmatex">\(\nu_i\)</span>.</p>
<p>Several example control file entries for the <code>clock</code> variables are
provided below:</p>
<pre><code>    # (1: strict clock, default)
    clock = 1 
    # (2: independent-rates)
    clock = 2 10.0 100.0 5.0 iid G
    # (3: correlated-rates)
    clock = 3 10.0 100.0 5.0 iid G
</code></pre>
<p>The specification (<code>clock = 2 10.0 100.0 5.0 iid G</code>) means the
following. First <span class="arithmatex">\(\bar\nu\)</span> is assigned a gamma distribution
<span class="arithmatex">\(G(10.0, 100.0)\)</span>, with mean 0.1. Given <span class="arithmatex">\(\bar\nu\)</span>, the conditional
i.i.d. prior means that <span class="arithmatex">\(\nu_i \sim G(\alpha, \alpha/\bar\nu)\)</span> with
shape parameter <span class="arithmatex">\(\alpha = 5.0\)</span> (<code>a_vi</code>) and mean <span class="arithmatex">\(\bar\nu\)</span>, for
<span class="arithmatex">\(i = 1, 2, \cdots, L\)</span>. As another example, the specification
(<code>clock = 2 10.0 100.0 5.0 dir LN</code>) means that <span class="arithmatex">\(\bar\nu\)</span> is assigned a
gamma prior <span class="arithmatex">\(G(10.0, 100.0)\)</span>, and then the sum <span class="arithmatex">\(L\bar\nu\)</span> is partitioned
into <span class="arithmatex">\(\nu_i\)</span> (for <span class="arithmatex">\(i = 1, 2, \cdots, L\)</span>) according to the Dirichlet
distribution (<code>prior = dir</code>) with concentration parameter <span class="arithmatex">\(\alpha = 5.0\)</span>
(<code>a_vi</code>). Again large <span class="arithmatex">\(\alpha\)</span> means the same extent of clock violation
at different loci.</p>
<p><strong>Choosing <span class="arithmatex">\(\alpha_{\bar{\nu}}\)</span> and <span class="arithmatex">\(\beta_{\bar{\nu}}\)</span></strong><br />
A larger variance <span class="arithmatex">\(\nu_i\)</span> represents a greater violation of the
molecular clock at locus <span class="arithmatex">\(i\)</span>. Note that <span class="arithmatex">\(\nu_i\)</span> will be similar to
<span class="arithmatex">\(\bar\nu\)</span>, especially if <span class="arithmatex">\(\alpha_{\bar{\nu}}\)</span> (<code>a_vi</code>) is large, and
<span class="arithmatex">\(\bar\nu\)</span> has prior mean <span class="arithmatex">\(\alpha_{\bar{\nu}}/\beta_{\bar{\nu}}\)</span>. If
(<code>a_vbar = 10 b_vbar = 100</code>), the prior mean will be 0.1. For the
log-normal model <span class="arithmatex">\(\nu = 0.5\)</span> suggests a serious violation of the clock
while <span class="arithmatex">\(\nu &lt; 0.1\)</span> represents only a slight violation.</p>
<h2 id="introgression-and-migration-models">Introgression and Migration Models</h2>
<p>Another important factor influencing the results of species tree
inference is introgression (gene flow) between species (or populations).
BPP also implements a model of episodic introgression between species or
populations <a href="https://doi.org/10.1093/molbev/msz296">Flouri et al 2020</a> as well as a model of ongoing gene flow (MSC-M).
Since version 4.1, BPP implements the MSC-I model
<a href="https://doi.org/10.1093/molbev/msz296">Flouri et al 2020</a> and since version 4.4 it implements the MSC-M model.
Currently, introgression (MSC-I) and ongoing gene flow (MSC-M) analyses can
only be performed using the A00 method of analysis (fixed species tree
and delimitation, see <a href="#methods-of-analysis">Methods of Analysis</a>). In other words, the user has to specify
the number of introgression events (or bands of gene flow), their
directions, and the populations involved on a fixed species tree. The
program will then estimate the parameters of the MSC-I (or MSC-M) models
using MCMC. We hope to implement MCMC moves that change the MSC-I (or MSC-M)
models and species tree topology in the future.</p>
<h3 id="the-msc-i-model">The MSC-I Model</h3>
<p>The MSC-I model involves three types of parameters: the species
divergence or hybridization times (<span class="arithmatex">\(\tau\)</span>s), the population size
parameters (<span class="arithmatex">\(\theta\)</span>s), and the introgression probabilities
(<span class="arithmatex">\(\varphi\)</span>s). The control file must specify an extended Newick format
tree with one or more introgression events as well as a prior
distribution for the introgression probability <span class="arithmatex">\(\varphi\)</span>. The MSC-I model
is specified in the <code>species&amp;tree</code> block using the extended Newick
notation (<a href="https://doi.org/10.1186/1471-2105-9-532">Cardona et al 2008</a>)</p>
<pre><code>    species&amp;tree = 3  A  B  C
    10 10 10
    ((A, (C)H[&amp;phi=0.5,&amp;tau-parent=yes])S, (H[&amp;tau-parent=yes], B) T)R;
</code></pre>
<p>The rules for the extended Newick format are as follows. '(A, B)S'
specifies two branches from S to A and from S to B, while '(A)H'
specifies one branch from H to A. Every branch in the tree model is
represented once. Each tip species occurs once. Internal nodes for
speciation nodes may and may not be labeled, but hybridization nodes
must be labeled. The extended Newick notation is not unique.</p>
<p>Each hybrid node has two parental species. Models A-C are distinguished
using a meta-data variable <code>tau-parent</code>. The value yes means that the
parental population exists and has a separate age parameter <span class="arithmatex">\(\tau\)</span>, with
an associated <span class="arithmatex">\(\varphi\)</span> parameter, while the value no means that such
parameters do not exist. For example, in model (A), <span class="arithmatex">\(\tau_S\)</span>, <span class="arithmatex">\(\tau_T\)</span>,
<span class="arithmatex">\(\theta_{Hl}\)</span> (for H-left), and <span class="arithmatex">\(\theta_{Hr}\)</span> (for H-right) are all
parameters; in model (B), <span class="arithmatex">\(\tau_S\)</span> and <span class="arithmatex">\(\theta_{Hl}\)</span> are not parameters
in the model while <span class="arithmatex">\(\tau_T\)</span> and <span class="arithmatex">\(\theta_{Hr}\)</span> are parameters; and
finally in model (C), none of those four parameters exists in the model.
Model (D) specifies a bidirectional introgression event between species
A and B.</p>
<p><img alt="MSCI-models" src="https://github.com/bpp/bpp-manual/blob/191d107413566b145360597fdfdfef3e68ce9173/figures/fig-msci-models.png?raw=true" /></p>
<p><strong>Four different types of MSC-I models implemented in BPP</strong><br />
The models are specified using the extended Newick format, as follows:</p>
<pre><code>(A): ((A, (C)H[&amp;phi=0.5,&amp;tau-parent=yes])S, (H[&amp;tau-parent=yes], B) T)R; 
(B): ((A, (C)H[&amp;phi=0.5,&amp;tau-parent=no])S, (H[&amp;tau-parent=yes], B) T)R; 
(C): ((A, (C)H[&amp;phi=0.5,&amp;tau-parent=no])S, (H[&amp;tau-parent=no], B) T)R; 
(D): ((A, (B) Y[&amp;phi=0.3])X, (X[&amp;phi=0.1])Y)R; 
(D): ((A, (B)Y)X, (X)Y)R; 
(D): ((A, Y)X, (B, X)Y) R; 
(D): ((A, y)x, (B, x) y) r;
</code></pre>
<p>The control file variable <code>phiprior</code> specifies the prior probability for
the <span class="arithmatex">\(\varphi\)</span> parameter, which is a beta distribution with parameters
<span class="arithmatex">\(a\)</span> and <span class="arithmatex">\(b\)</span>. The syntax is</p>
<pre><code>    phiprior = a b
</code></pre>
<p>where <span class="arithmatex">\(a\)</span> and <span class="arithmatex">\(b\)</span> are positive numbers. For example, <code>phiprior = 1 1</code>
specifies the beta prior beta(<span class="arithmatex">\(a, b\)</span>) with <span class="arithmatex">\(a = 1\)</span> and <span class="arithmatex">\(b = 1\)</span> for
<span class="arithmatex">\(\varphi\)</span>, the introgression probability under the MSC-I model.</p>
<h4 id="autogeneration-of-extended-newick-graphs">Autogeneration of extended newick graphs</h4>
<p>Starting with BPP release 4.4.0, we have changed the definition of introgression probability (<span class="arithmatex">\(\varphi\)</span>) so that it is assigned to the horizontal (introgression) branch.  This note illustrates the <code>msci-create</code> feature in bpp, which generates extended Newick notation for the MSC-I model from a data file which contains a binary species tree with introgression events specified using source and target branches on the tree.
The <code>--msci-create</code> option has been available since <a href="https://github.com/bpp/bpp/issues/114">BPP 4.2.1</a>.  The user prepares a input file (named <code>msci.txt</code> in our example), which includes a binary species tree in Newick format and specifies the introgression events by identifying the source and target branches involved in the introgression.  Run bpp using the following command:</p>
<pre><code>$ bpp --msci-create msci.txt
</code></pre>
<p>This generates extended Newick notation for the MSC-I model, which can be copied into a bpp control file. 
The input file <code>msci.txt</code> uses four commands: <code>tree</code>, <code>define</code>, <code>hybridization</code>, and <code>bidirection</code>.</p>
<ul>
<li><code>tree</code> defines the binary species tree, in Newick format.</li>
<li><code>define</code> defines an ancestral species or internal node label, as the most recent common ancestor of the tip species.  Alternatively you can label the internal nodes on the Newick tree.</li>
<li><code>hybridization</code> defines a hybridization/introgression event by specifying the source and target branches (which represent the source and target populations involved).</li>
<li><code>bidirection</code> defines a bidirectional introgression (BDI) event.</li>
</ul>
<p>One may think of the binary species tree as describing the history of species divergences and add introgression events onto it as new (horizontal) branches.  The introgression probability is assigned to the newly created introgression branch. </p>
<h4 id="examples-using-msci-create">Examples using --msci-create</h4>
<p>A series of examples follow. The introgression graph is displayed in a figure followed by the <code>msci=create</code> encoding and the resulting extended Newick format.
<img alt="Fig-MSCI2" src="https://github.com/bpp/bpp-manual/blob/36012225be26f3b82593c9a35eaaa3e7217a5fd0/figures/msci_example2.png?raw=true" /></p>
<pre><code># Model A, version 1
tree (A,(B,C));
define T as B,C
define R as A,B
hybridization R A, T C as S H tau=yes, yes phi=0.10

# Model A, version 2
tree (A,(B,C)T)R;
hybridization R A, T C as S H tau=yes, yes phi=0.10

# The generated Newick notation for model A is 
# ((H[&amp;phi=0.1,tau-parent=yes],A)S, (B,(C)H[&amp;phi=0.9,tau-parent=yes])T)R;

# Model B1
tree (A,(B,C)T)R;
hybridization R A, T C as S H tau=no, yes phi=0.10

# The generated Newick notation for model B1 is 
# ((H[&amp;phi=0.1,tau-parent=no],A)S, (B,(C)H[&amp;phi=0.9,tau-parent=yes])T)R;

# Model B2
tree ((A,C)S,B)R;
hybridization R B, S C as T H tau=no, yes phi=0.10

# The generated Newick notation for model B2 is 
# ((A,(C)H[&amp;phi=0.9,tau-parent=yes])S, (H[&amp;phi=0.1,tau-parent=no],B)T)R;

# Model C
tree (A,(B,C)T)R;
hybridization R A, T C as S H tau=no, no phi=0.40

# The generated the Newick notation for model C is 
# ((H[&amp;phi=0.4,tau-parent=no],A)S, (B,(C)H[&amp;phi=0.6,tau-parent=no])T)R;

# Model D
tree (A,B)R;
bidirection  A R, B R as X Y phi=0.1,0.2
</code></pre>
<p>The above figure illustrates the specifications of models A, B, C, and D of <a href="https://doi.org/10.1093/molbev/msz296">Flouri et al 2020</a>.  In models A, B1, and C, the introgression event is from the source branch RA to the target branch TC, with a new introgression branch SH created.  The introgression probability (<span class="arithmatex">\(\varphi = 0.1\)</span>) is assigned to the newly added introgression branch (SH) while <span class="arithmatex">\(1 – \varphi\)</span> is assigned to the other parent branch (TH).  Models A, B1, and C are distinguished by using the keyword <code>tau</code> (which means the same as <code>tau-parent</code> in the Newick notation of the MSci model). Model A assumes that the two parental species S and T of the hybridization node H have distinct ages from H, with <span class="arithmatex">\(\tau_S &gt; \tau_H\)</span> and <span class="arithmatex">\(\tau_T &gt; \tau_H\)</span>.  Thus we have <code>hybridization R A, T C as S H tau=yes, yes phi=0.10</code>: the first <code>yes</code> means that parental node S on the source branch RA has a distinct tau from node H and that the second <code>yes</code> means that parental node T on the target branch TC has a distinct tau from node H.  Note that there are often multiple ways of specifying the same MSC-I model and here you can specify model A by starting with the binary tree <code>((A, C), B)</code> and adding branch TH as a introgression event.  Model A might appear to be nonsensical biologically since only contemporary species can exchange migrants.  Nevertheless, model A may be used to represent introgressions from a ghost species not sampled in the sequence dataset.  Suppose at time <span class="arithmatex">\(\tau_S\)</span> a speciation event generated two species SA and SH, and at <span class="arithmatex">\(\tau_H\)</span>, species SH contributed migrants into species THC, but species SH has since become extinct or is otherwise not sampled in the data.  This scenario matches model A.
Model B1 assumes <span class="arithmatex">\(\tau_S = \tau_H\)</span> and <span class="arithmatex">\(\tau_T &gt; \tau_H\)</span> so there is no new tau for parent S of the hybridization node H, and there is a new tau for parent T.  Thus we have <code>tau=no, yes</code>.  Model B2 works similarly.  In model C, none of parents S and T has a distinct age from H, so we have <code>tau=no, no</code>.  One interpretation is that species C is a hybrid species. In model D for the bidirectional introgression (BDI) model, there is no distinction of source and target branches.  We specify two phi values, assigned to the introgression (horizontal) branches: <span class="arithmatex">\(\varphi_X = 0.1\)</span> for node X (or into node X) and <span class="arithmatex">\(\varphi_Y = 0.2\)</span> for node Y (see above figure).</p>
<h3 id="the-msc-m-model">The MSC-M Model</h3>
<p>The multispecies-coalescent-with-migration model (or isolation-with-migration or IM model) is activated in BPP using the keyword migration.
A binary species tree is specified using an extended Newick format that includes labels for internal nodes; these are subsequently used in the control file to specify the source and target populations involved in migration.</p>
<h4 id="control-file-specification-of-msc-m">Control file specification of MSC-M</h4>
<p>The basic model is specified in the control file as follows, using a species tree for four species (A, B, C, D) as an example:</p>
<pre><code>      ((A, B)S, (C, D)T)R;

     migprior = 2 200 
    migration = 2
                A C
                S C
</code></pre>
<p>Here S is the AB common ancestor, T is the CD common ancestor, while R is the ABCD ancestor.  Not all internal nodes need to be labeled but those involved in migration have to be.  The migration line specifies 2 migration connections: one connecting source population A to target population C, and another connecting source population S to target population C, with migration rates <span class="arithmatex">\(M_{AC}\)</span> and <span class="arithmatex">\(M_{SC}\)</span>.  Note that the migration rate </p>
<p><span class="arithmatex">\(M_{AC} = m_{AC} N_C\)</span></p>
<p>is the expected number of immigrants in population C that are from population A per generation under the real-world view with time running forward, where <span class="arithmatex">\(m_{AC}\)</span> is the proportion of immigrants in population C that are from population A.  See Appendix A for definitions of migration rates used in different programs. </p>
<h4 id="prior-on-migration-rates">Prior on migration rates</h4>
<p>The control file variable <code>migprior = alpha beta</code> specifies a default gamma prior density with parameters <code>alpha</code> and <code>beta</code> for all migration rates.  In the example above, both <span class="arithmatex">\(M_{AC}\)</span> and <span class="arithmatex">\(M_{SC}\)</span> are assigned the gamma prior <span class="arithmatex">\(G(2, 200)\)</span>, with prior mean <span class="arithmatex">\(2/200 = 0.01\)</span>.
A separate gamma prior can instead be defined for each migration rate (where the migration connection is specified using the source and target populations) and this specification takes precedence.  For example if we specify</p>
<pre><code>     migprior = 2 200
    migration = 2
                A C  2 100
                S C
</code></pre>
<p>the priors are <span class="arithmatex">\(M_{AC} ~ G(2, 100)\)</span> with the prior mean <span class="arithmatex">\(0.02\)</span> and <span class="arithmatex">\(M_{SC} ~ G(2, 200)\)</span> from the default prior specified by <code>migprior</code>.</p>
<h4 id="variable-migration-rates-among-loci">Variable migration rates among loci</h4>
<p>In this model, the migration rate <span class="arithmatex">\(M_i\)</span> at each locus <em>i</em> varies according to a gamma distribution <span class="arithmatex">\(M_i \approx G(\alpha_M,\alpha_M/\overline{M})\)</span>, with shape parameter <span class="arithmatex">\(\alpha_M\)</span> while the mean rate <span class="arithmatex">\(\overline{M}\)</span> is assigned the gamma prior <span class="arithmatex">\(\overline{M} \approx G(\alpha,\beta)\)</span>.  Here <span class="arithmatex">\(\alpha_M\)</span> is a parameter that characterizes the variation of <span class="arithmatex">\(M_i\)</span> among loci, with a small <span class="arithmatex">\(\alpha_M\)</span> (e.g., 0.5 or 1) indicating highly variable rates among loci and a large <span class="arithmatex">\(\alpha_M\)</span> indicating nearly constant migration rates among loci (when <span class="arithmatex">\(\alpha_M = \infty\)</span> all loci have the same rate).</p>
<pre><code>     migprior = 2 200
    migration = 2
                A C  2 100 5
                S C
</code></pre>
<p>In the above example, <span class="arithmatex">\(M_{SC}\)</span> is applied to all loci with the prior <span class="arithmatex">\(G(2, 200)\)</span>, but <span class="arithmatex">\(M_{AC}\)</span> varies among loci according to the shape parameter 
<span class="arithmatex">\(\alpha_M = 5\)</span>, and the mean rate for all loci <span class="arithmatex">\(\overline{M}_{AC}\)</span> is assigned the gamma prior <span class="arithmatex">\(G(2, 100)\)</span>.</p>
<h4 id="priors-and-pseudo-priors">Priors and pseudo-priors</h4>
<p>When the gene flow involves ancestral species, migration may become impossible because of changes to the species divergence times in the MCMC.  In the example above, if <span class="arithmatex">\(\tau_S &lt; \tau_T\)</span> migration from S to C is possible during the time period <span class="arithmatex">\((\tau_S, \tau_T)\)</span>, but if <span class="arithmatex">\(\tau_S &gt; \tau_T\)</span> migration from S to C is impossible as the two populations were never contemporary.  Thus the MCMC proposal to change species divergence times <span class="arithmatex">\(\tau_S\)</span> or <span class="arithmatex">\(\tau_T\)</span> may cause the migration rate parameter <span class="arithmatex">\(M_{SC}\)</span> to disappear or reappear.  When <span class="arithmatex">\(M_{SC}\)</span> is absent from the model (that is, when <span class="arithmatex">\(\tau_S &gt; \tau_T\)</span>), the MCMC algorithm treats it as a pseudo-parameter (written as <span class="arithmatex">\(M^*_{SC}\)</span>) and assigns it a pseudo-prior to facilitate the trans-dimensional move.  The choice of pseudo-priors affects the mixing efficiency of the MCMC, but not the correctness of the algorithm.  For good mixing, one should choose the pseudo-prior to be close to the posterior of the parameter <span class="arithmatex">\(M_{SC}\)</span> (which can be generated by running short chains).</p>
<pre><code>     migprior = 2 200 
    migration = 2
                A C  2 100 
                S C  2 200   100 250
</code></pre>
<p>In the example above, the prior <span class="arithmatex">\(M_{SC} \approx G(2, 200)\)</span> is applied when <span class="arithmatex">\(\tau_S &lt; \tau_T\)</span>, and the pseudo-prior <span class="arithmatex">\(M^*_{SC} \approx G(100, 250)\)</span> with mean 0.4, is applied when <span class="arithmatex">\(\tau_S &gt; \tau_T\)</span>.  Note that in this example the pseudo-prior is far more concentrated than the prior (with shape parameter 100 versus 2).
In order to summarize the posterior from the MCMC samples, only those samples taken when the migration rate is defined should be used (that is, only samples collected when <span class="arithmatex">\(\tau_S &lt; \tau_T\)</span> in our example).  Samples collected when the migration rate does not exist in the model (that is, when <span class="arithmatex">\(\tau_S &gt; \tau_T\)</span> in our example) approximate the pseudo-prior.</p>
<p>In the case of four species, the following awk commands may be used to split the samples into two files (this assumes that tau_S and tau_T are in column 10 and 11)</p>
<pre><code>awk 'BEGIN {cnt=0} NR&gt;1{if ($10 &gt; $11) {cnt=cnt+1;}}} END {print cnt}' mcmc.txt
# split sample file into ab.txt (samples where t_AB &gt; t_CD) and cd.txt (t_CD &gt; t_AB)
head -n 1 mcmc.txt &gt; cd.txt; awk 'NR&gt;1 {if ($10&lt;$11) print $0}' mcmc.txt &gt;&gt; cd.txt
head -n 1 mcmc.txt &gt; ab.txt; awk 'NR&gt;1 {if ($10&gt;$11) print $0}' mcmc.txt &gt;&gt; ab.txt
</code></pre>
<p>Use something like this if you understand the syntax. The running mean of the migration rate printed on the monitor during the MCMC run is the posterior mean after the filtering.
Note that this problem of the migration rate appearing and disappearing during the MCMC may exist when a migration event involves ancestral populations, and is not limited to the balanced species tree for four species.  Later we should automate the summary of the MCMC sample.<br />
Under the model of variable <span class="arithmatex">\(M\)</span> among loci, all migration rates for all loci may appear and disappear if the migration involves ancestral species.  The following five ways of specifying the model and priors and pseudo-priors are accepted (with <span class="arithmatex">\(M_{SC}\)</span> as an example)</p>
<pre><code>(a)  S C
(b)  S C a_M
(c)  S C a b
(d)  S C a b a_M
(e)  S C a b pseudo_a pseudo_b
(f)  S C a b a_M pseudo_a pseudo_b
</code></pre>
<p>In options (a) and (b), the default prior specified with migprior is assigned on <span class="arithmatex">\(M_{SC}\)</span>.</p>
<h2 id="odds-and-ends">Odds and Ends</h2>
<pre><code>    heredity = 0       # (0: No variation)
    heredity = 1 4 4   # (1: estimate, &amp; a_gamma b_gamma)
    heredity = 2 heredity.txt      # (2: from file)
</code></pre>
<p><code>heredity = 0</code> is the default and means that <span class="arithmatex">\(\theta\)</span> is the same for
all loci. <code>heredity = 1</code> or <code>2</code> specifies two models that allow <span class="arithmatex">\(\theta\)</span>
to vary among loci, which may be useful for combined analysis of data
from autosomal, mitochondrial, X and Y loci. With such mixed data, the
effective population sizes are different among loci, so that a heredity
multiplier [inheritance scalars @Hey2004] should be applied. Other
factors such as natural selection may also cause <span class="arithmatex">\(\theta\)</span> to deviate
from the neutral expectation. BPP implements two options
for this. The first option (<code>heredity = 1</code>) is to estimate the
multipliers from the data, using a gamma prior with parameters <span class="arithmatex">\(\alpha\)</span>
and <span class="arithmatex">\(\beta\)</span> specified by the user. In the example above, a gamma prior
<span class="arithmatex">\(G(4, 4)\)</span>, with mean <span class="arithmatex">\(4/4 = 1\)</span>, is specified for the multiplier for each
locus. The MCMC should then generate a posterior for the multiplier for
each locus. The second option (<code>heredity = 2</code>) is for the user to
specify the multipliers in a file, and the multipliers will then be used
as fixed constants in the MCMC run. The file simply contains as many
numerical values as the number of loci, separated by spaces or line
breaks.</p>
<hr />
<p>Genome           Heredity scalar</p>
<p>Nuclear            autosome 1
  X chromosome          0.75
  Y chromosome          0.25
  Mitochondrial         0.25</p>
<hr />
<p>: Heredity scalars</p>
<p>[[table-heredity-scalar]]{#table-heredity-scalar
label="table-heredity-scalar"}</p>
<p>Note.--- The effect of the locus-specific mutation rates and the
locus-specific heredity multipliers are different. A locus rate is used
to multiply all <span class="arithmatex">\(\theta\)</span>s and <span class="arithmatex">\(\tau\)</span>s for the locus, while a heredity
multiplier is used to multiply all <span class="arithmatex">\(\theta\)</span> parameters for the locus but
not the <span class="arithmatex">\(\tau\)</span>s. Nevertheless, those parameters are quite likely to be
strongly correlated, especially when the species tree is small.</p>
<h2 id="methods-of-analysis">Methods of Analysis</h2>
<p>Here we describe the specifics of the four different types of analyses
that BPP can perform and illustrate them using several of the example
control files provided in the <code>examples</code> subdirectory of the BPP
distribution. Briefly, BPP is capable of analyzing sequence data using 4
distinct phylogeographical models. The 4 models are determined by the
combinations of two possible settings for each of the two Boolean
control file variables <code>speciesdelimitation</code> (0 = no species
delimitation, 1 = species delimitation) and <code>speciestree</code> (0 = fixed
species tree, 1 = inferred species tree). Therefore, analysis A00 infers
the biogeographical parameters <span class="arithmatex">\(\theta\)</span> and <span class="arithmatex">\(\tau\)</span> on a fixed (user
specified) tree with a fixed (delimited) number of species/populations
as described in (<a href="https://doi.org/10.1093/genetics/164.4.1645">Rannala and Yang 2003</a>. Analysis A10 delimits the number of
species and infers biogeographical parameters using a fixed species tree
(guide tree) as described in <a href="https://doi.org/10.1073/pnas.0913022107">Yang and Rannala 2010</a>. Analysis A01 infers the
species tree and biogeographical parameters using a fixed number of
species (delimitation) as described in <a href="https://doi.org/10.1093/sysbio/syw119">Rannala and Yang 2017</a>. Analysis A11
jointly infers the species tree, species delimitation and
biogeographical parameters as described in <a href="https://doi.org/10.1093/molbev/msu279">Yang and Rannala 2014</a>. 
Under Model A00 it is also possible to include a model of either instantaneous
introgression or ongoing gene flow -- examples of such analyses are
given in <a href="#introgression-and-migration-models">Introgression and Migration Models</a>. For each of these 4 major types of analyses
the control file specifications will differ as well as the form of
output printed to the screen and output files. Therefore, we will
include two subsections in our discussion of each analysis method: input
and output. Here we only consider the control file variables specific to
the methods of analysis under consideration. For a general discussion of
other control file variables see the example control file discussed in <a href="#control-file">Control File</a>.
Note that for most analyses the output printed to screen during the run
and that printed to the specified output file will be identical.</p>
<p>It is important to note that only analysis A00 applies within-model
inference. There is one specified model in that case (the MSC with a
fixed species tree) and the parameters are well defined, with the
objective being to estimate those parameters. In contrast, analyses A01,
A10, and A11 all apply trans-model inference. They move between
different models, and the main objective is to calculate the posterior
probabilities of the models. Each is an instance of the MSC model, but
the species delimitation (the number and nature of the species) and/or
the species phylogeny may differ between models. In analyses A01, A10,
and A11, the prior specified using control file variable <code>thetaprior</code>
applies to all <span class="arithmatex">\(\theta\)</span> parameters in all models. Similarly, all MSC
models specifying two or more delimited species have a parameter
<span class="arithmatex">\(\tau_0\)</span> (the divergence time of the root), and these parameters (for
different models) are assigned the same prior, specified by the
<code>tauprior</code> control file variable.</p>
<h3 id="a00-demography-and-divergence-time-estimation">A00: Demography and Divergence Time Estimation</h3>
<p>Analysis A00<code>(speciesdelimitation = 0, speciestree = 0)</code>, requires a
user-specified species tree topology, given by variable <code>species&amp;tree</code>
in the control file, and infers the <span class="arithmatex">\(\theta\)</span> and <span class="arithmatex">\(\tau\)</span> parameters on
this fixed tree. The theory underlying estimation of species divergence
times and population sizes (<span class="arithmatex">\(\tau\)</span>s and <span class="arithmatex">\(\theta\)</span>s) under the MSC model
when the species phylogeny is given is described in (<a href="https://doi.org/10.1093/genetics/164.4.1645">Rannala and Yang 2003</a>. Note
that if there are <span class="arithmatex">\(s\)</span> species in the species tree, the model will
minimally contain the following parameters: <span class="arithmatex">\(s – 1\)</span> species divergence
times (<span class="arithmatex">\(\tau\)</span>s) and <span class="arithmatex">\(s – 1\)</span> ancestral species <span class="arithmatex">\(\theta\)</span>s. A <span class="arithmatex">\(\theta\)</span>
parameter will also exist for each contemporary species with at least
two sequences present at some locus. If the data are for a single
species/population, the model will contain one parameter only, <span class="arithmatex">\(\theta\)</span>
for that species. If there is only one species, the MSC model becomes
the single-population Kingman coalescent. We will provide
examples of both a single population analysis (using the control file
<code>yu2001.bpp.ctl</code>) and a multi-species analysis (using the control file
<code>A00.bpp.ctl</code>).</p>
<p><strong>How many <span class="arithmatex">\(\theta\)</span> and <span class="arithmatex">\(\tau\)</span> parameters exist?</strong>\
The number of parameters that BPP includes in the MSC or
MSC-I models depends on the data configuration. The program always
includes <span class="arithmatex">\(\theta\)</span> and <span class="arithmatex">\(\tau\)</span> parameters for each internal node
(ancestral species) on the species tree, but includes a <span class="arithmatex">\(\theta\)</span> for an
extant species (tip) if and only if that species has at least 2
sequences at some loci. If two or more sequences for a species are
specified in the control file but there are at most 0 or 1 sequence per
locus for that species in the sequence data file, <span class="arithmatex">\(\theta\)</span> for that
species will not be identifiable or estimable. In that case, the
posterior for the parameter will be the prior. Nevertheless, other
results (such as the posterior distribution for other parameters or
posterior probabilities for species trees and species delimitations)
will still be correct. The same applies to other more complex cases of
missing data and parameter non-identifiability. As an example, suppose
the species tree and the numbers of sequences for two kinds of loci are
as follows:\
((A,B), (C,D))\
locus configuration 1: 1 1 0 0\
locus configuration 2: 2 0 0 1\
In this case, <span class="arithmatex">\(\theta_A\)</span>, <span class="arithmatex">\(\theta_{AB}\)</span>, <span class="arithmatex">\(\theta_{ABCD}\)</span>, <span class="arithmatex">\(\tau_{AB}\)</span>,
and <span class="arithmatex">\(\tau_{ABCD}\)</span> are identifiable while <span class="arithmatex">\(\theta_B\)</span>, <span class="arithmatex">\(\theta_C\)</span>,
<span class="arithmatex">\(\theta_D\)</span>, <span class="arithmatex">\(\theta_{CD}\)</span> and <span class="arithmatex">\(\tau_{CD}\)</span> are not. Ii is impossible to
estimate <span class="arithmatex">\(\theta_D\)</span> , <span class="arithmatex">\(\theta_{CD}\)</span> and <span class="arithmatex">\(\tau_{CD}\)</span> as no data are
available from species C.</p>
<h4 id="input-a00-single-population">Input A00 (single population)</h4>
<p>Our first example is a dataset of human sequence data generated by
<a href="https://doi.org/10.1093/oxfordjournals.molbev.a003795">Yu et al 2001</a>. 
The control file is found in the BPP distribution at location:</p>
<pre><code>    examples/yu2001/yu2001.bpp.ctl
</code></pre>
<p>The data comprise 61 phased sequences for a single locus that will be
analyzed to estimate the single parameter <span class="arithmatex">\(\theta\)</span>. There is no need to
specify an Imap file, or tag the sequence names in the sequence file
(<code>yu2001.txt</code>); the sequence names are read and then ignored. Multiple
loci may be included in the sequence file. The contents of the control
file <code>yu2001.bpp.ctl</code> are shown below:</p>
<pre><code>    seed = -1

    seqfile = yu2001.txt
    outfile = out.txt
    mcmcfile = mcmc.txt

    # fixed species delimitation and species tree
    speciesdelimitation = 0 
    speciestree = 0

    species&amp;tree = 1  H
                      61  

    # 0: no data (prior); 1:seq like
    usedata = 1

    # number of data sets in seqfile
    nloci = 1    

    # remove sites with ambiguity data (1:yes, 0:no)?
    cleandata = 0    

    # gamma(a, b) for theta
    thetaprior = gamma 2 2000

    # auto (0 or 1): finetune for GBtj, GBspr, theta, tau, mix, locusrate, seqerr
    finetune = 1: 2 0.00001 0.0001  0.0005 0.5 0.2 1.0  

    # MCMC samples, locusrate, heredityscalars, genetrees, locussubstitution
    print = 1 0 0 0  

    burnin = 4000
    sampfreq = 2
    nsample = 10000
</code></pre>
<p>The two lines:</p>
<pre><code>    speciesdelimitation = 0 
    speciestree = 0
</code></pre>
<p>specify that the species delimitation (number of species) is fixed as
well as the species tree topology. In this case, there is no species
tree and only a single species exists. There is no need for a species
tree, so the block for specifying species names and species tree looks
like this:</p>
<pre><code>    species&amp;tree = 1  H
                      61
</code></pre>
<p>The line:</p>
<pre><code>    # gamma(a, b) for theta
    thetaprior = gamma 2 2000
</code></pre>
<p>specifies a gamma distribution for the prior on <span class="arithmatex">\(\theta\)</span> with parameters
<span class="arithmatex">\(\alpha = 2\)</span> and <span class="arithmatex">\(\beta = 2000\)</span> which has mean <span class="arithmatex">\(0.001\)</span> and variance
<span class="arithmatex">\(5 \times 10^{-7}\)</span>.</p>
<h4 id="output-a00-single-population">Output A00 (single population)</h4>
<p>When BPP is run using the above control file a summary of the progress
of the MCMC analysis is printed to screen. Here, we briefly explain what
this output means. The different analyses presented in this chapter all
produce a similar form of output when the program is running, but the
number of columns and their specific content varies somewhat. Omitting
some mostly irrelevant output related to the adjustment of proposal
moves, the output to screen when this control file is run appears as
follows:</p>
<pre><code>         |  Acceptance proportions  |
    Prgs | Gage Gspr thet  tau  mix | mthet1       log-PG         log-L
    -------------------------------------------------------------------
     -15%  0.71 0.18 0.30 0.00 0.38   0.0004    467.11472  -12721.22305  0:01
     (some output omitted here)
       5%  0.71 0.29 0.33 0.00 0.24   0.0004    428.40742  -12721.09413  0:05
      10%  0.71 0.29 0.31 0.00 0.25   0.0003    480.92848  -12720.75185  0:06
      15%  0.71 0.29 0.30 0.00 0.24   0.0003    476.37432  -12720.65199  0:08
      20%  0.71 0.29 0.31 0.00 0.23   0.0003    441.36960  -12720.92326  0:09
      25%  0.71 0.29 0.32 0.00 0.23   0.0003    443.51740  -12720.85802  0:10
      30%  0.71 0.29 0.31 0.00 0.23   0.0003    441.93616  -12720.86033  0:11
      35%  0.71 0.29 0.31 0.00 0.23   0.0003    495.50716  -12720.84565  0:12
      40%  0.71 0.29 0.31 0.00 0.23   0.0003    452.82049  -12720.87810  0:13
      45%  0.71 0.29 0.31 0.00 0.23   0.0003    461.93346  -12720.94582  0:15
      50%  0.71 0.29 0.32 0.00 0.23   0.0004    463.85955  -12721.01005  0:16
      55%  0.71 0.29 0.32 0.00 0.23   0.0004    463.42744  -12720.98006  0:17
      60%  0.71 0.29 0.32 0.00 0.23   0.0004    478.52797  -12721.03353  0:18
      65%  0.71 0.29 0.32 0.00 0.23   0.0004    499.97705  -12720.96168  0:19
      70%  0.71 0.29 0.32 0.00 0.23   0.0004    455.32489  -12720.93611  0:20
      75%  0.71 0.29 0.32 0.00 0.23   0.0004    439.21544  -12720.99190  0:22
      80%  0.71 0.29 0.32 0.00 0.23   0.0004    461.56954  -12721.00180  0:23
      85%  0.71 0.29 0.32 0.00 0.23   0.0004    469.53443  -12720.99823  0:24
      90%  0.71 0.29 0.32 0.00 0.23   0.0004    455.60303  -12720.99213  0:25
      95%  0.71 0.29 0.32 0.00 0.23   0.0004    454.07141  -12721.00745  0:26
     100%  0.71 0.29 0.32 0.00 0.23   0.0004    461.01044  -12720.97549  0:27

    0:27 spent in MCMC

              theta_1H  lnL
    mean      0.000352  -12720.995649
    median    0.000337  -12720.726000
    S.D       0.000116  2.936769
    min       0.000089  -12735.535000
    max       0.001293  -12712.774000
    2.5%      0.000170  -12727.454000
    97.5%     0.000617  -12716.086000
    2.5%HPD   0.000145  -12726.920000
    97.5%HPD  0.000574  -12715.718000
    ESS*      757.838791  1180.434956
    Eff*      0.075784  0.118043
</code></pre>
<p>The line:</p>
<pre><code>         |  Acceptance proportions  |
    Prgs | Gage Gspr thet  tau  mix | mthet1       log-PG         log-L
    -------------------------------------------------------------------
</code></pre>
<p>is a header that explains the content of each column printed during the
run. The <code>Prgs</code> (progress) column will appear in all analyses and
indicates the percentage of the MCMC iterations that have been
completed. If this number is negative it indicates that the program is
still running the burn-in iterations. For example, <code>-15%</code> means that 15
percent of the burn-in iterations remain to be completed. The next 5
columns are the current acceptance proportions for different parameter
proposals in the MCMC. The proposals are defined as follows:</p>
<pre><code>-   `Gage`: proposal to change ages of nodes on gene trees

-   `Gsbr`: proposal to change gene tree topology using SPR move

-   `thet`: proposal to change $\theta$ parameter

-   `tau`: proposal to change $\tau$ parameter

-   `mix`: mixing step proposal
</code></pre>
<p>In the output the acceptance proportions are stable. The fifth column
(the <span class="arithmatex">\(\tau\)</span> proposal acceptance proportion) is zero; this is expected
because there are no species divergence times (<span class="arithmatex">\(\tau\)</span>s) in the model
when only a single species exists. Column 7, labeled <code>mthet1</code> gives the
current average (mean) value of theta in the MCMC. Columns 8 and 9 are
the log probability of the coalescent model (<code>log-PG</code>) and the
log-likelihood of the sequence data on the gene tree (<code>log-L</code>). Check
that the acceptance proportions are in a reasonable range (about 20% to
40% for continuous parameters such as <span class="arithmatex">\(\theta\)</span>). If you need to
terminate the program before it finishes, you can do so using the key
combination <code>Ctrl-C</code> (e.g., hold down the Control key and then press the
C key). The final screen output, printed after the MCMC is completed,
summarizes the posterior parameter estimates for <span class="arithmatex">\(\theta\)</span> (<code>theta_1H</code>)
in the second column. The posterior mean in the first row, for example,
is <span class="arithmatex">\(0.000352\)</span> and the lower and upper limits of the highest posterior
density (HPD) set of values are in rows 8 and 9, respectively. In this
case, the HPD credibility interval for <span class="arithmatex">\(\theta\)</span> is
<span class="arithmatex">\((0.000145,0.000574)\)</span>.</p>
<h4 id="input-a00-multiple-populations">Input A00 (multiple populations)</h4>
<p>Our second example is a dataset of frog sequence data generated by
<a href="https://doi.org/10.1111/j.1365-294X.2011.05411.x">Zhou et al 2012</a>. 
This same dataset will also be used to illustrate analyses
A01, A10 and A11. The control file is found in the BPP distribution at
location:</p>
<pre><code>    examples/frogs/A00.bpp.ctl
</code></pre>
<p>The data comprise 5 loci for 4 populations, with the number of sequences
at each locus varying between 21 and 30. We will be estimating <span class="arithmatex">\(\theta\)</span>s
for each of the 4 contemporary and 3 ancestral populations as well as
the 3 species divergence times (<span class="arithmatex">\(\tau\)</span>s). The contents of the control
file <code>A00.bpp.ctl</code> are shown below:</p>
<pre><code>    seed =  -1

    seqfile = frogs.txt
    Imapfile = frogs.Imap.txt
    outfile = out.txt
    mcmcfile = mcmc.txt

    # fixed number of species/populations
    speciesdelimitation = 0 

    # fixed species tree
    speciestree = 0

    species&amp;tree = 4  K  C  L  H
                        9  7 14  2
                     (((K, C), L), H);

    # unphased data for all 4 populations
    phase =   1  1  1  1

    # use sequence likelihood                  
    usedata = 1

    nloci = 5  

    # do not remove sites with ambiguity data
    cleandata = 0    

    # gamma(a, b) for theta (estimate theta)
    thetaprior = gamma 2 2000 

    # invgamma(a, b) for root tau &amp; Dirichlet(a) for other tau's
    tauprior = invgamma 3 0.002 

    # finetune for GBtj, GBspr, theta, tau, mix, locusrate, seqerr
    finetune =  1: 5 0.001 0.001  0.001 0.3 0.33 1.0  

    # MCMCsamples, locusrate, heredityscalars, genetrees, substitutionparams
    print = 1 0 0 0 0   

    burnin = 8000
    sampfreq = 2
    nsample = 100000
</code></pre>
<p>As in the previous analysis, the two lines:</p>
<pre><code>    speciesdelimitation = 0 
    speciestree = 0
</code></pre>
<p>specify that the species delimitation (number of species) is fixed as
well as the species tree topology. In this case, a species tree is
specified along with population labels. The block for specifying species
names and the species tree looks like this:</p>
<pre><code>    species&amp;tree = 4  K  C  L  H
                        9  7 14  2
                     (((K, C), L), H);
</code></pre>
<p>The first line above lists the number of species (populations) which is
4 in this example. This is followed by the label for each species (these
should match the labels in the map file):</p>
<pre><code>    K  C  L  H
</code></pre>
<p>The next line specifies the maximum number of sequences per species in
the same order as the species labels:</p>
<pre><code>    9  7 14  2
</code></pre>
<p>the actual number of sequences at any locus must be less than this
value. The third line is a rooted phylogenetic tree topology specified
in Newick format:</p>
<pre><code>    (((K, C), L), H);
</code></pre>
<p>The next line specifies that the loci for each of the 4 populations are
unphased:</p>
<pre><code>    phase = 1 1 1 1
</code></pre>
<p>The specified phase variables should include a 0 or 1 (Boolean) entry
for each population, where the populations are assumed to be in the same
order as specified above (in the <code>species&amp;tree</code> block) and the Boolean
variable indicates that the sequence data for a population are either
unphased (1) or phased (0). Ambiguity characters are used to represent
genotypes for unphased sequences (see <a href="#bpp-control-file-variables">BPP control file variables</a>. 
In the example data the sequences
are unphased for all 4 populations. Currently, it is not possible to
have partially phased data (e.g., only some loci unphased in a
particular population). The program will infer the phase as part of the
analysis.</p>
<p>Unlike the earlier single population analysis, in which we estimated a
single <span class="arithmatex">\(\theta\)</span> parameter, in the multipopulation analysis we will be
estimating multiple <span class="arithmatex">\(\theta\)</span>s (one for each population) and divergence
times (<span class="arithmatex">\(\tau\)</span>s) between populations. Thus, we have prior distributions
for both <span class="arithmatex">\(\theta\)</span> and <span class="arithmatex">\(\tau_0\)</span>. The lines below specify a gamma
distribution as the prior for <span class="arithmatex">\(\theta\)</span> with parameters <span class="arithmatex">\(\alpha = 2\)</span> and
<span class="arithmatex">\(\beta = 2000\)</span> (the prior mean and variance of <span class="arithmatex">\(\theta\)</span> are <span class="arithmatex">\(0.001\)</span> and
<span class="arithmatex">\(0.0000005\)</span>):</p>
<pre><code>    # gamma(a, b) for theta
    thetaprior = gamma 2 2000 
</code></pre>
<p>The lines:</p>
<pre><code>    # invgamma(a, b) for root tau &amp; Dirichlet(a) for other tau's
    tauprior = invgamma 3 0.002 
</code></pre>
<p>specify an inverse gamma distribution as a prior for the root age and
the remaining <span class="arithmatex">\(\tau\)</span>s have a uniform Dirichlet distribution conditional
on the root age (the prior mean and variance of the root age are <span class="arithmatex">\(0.001\)</span>
and <span class="arithmatex">\(0.000001\)</span>,respectively):</p>
<h4 id="output-a00-multiple-populations">Output A00 (multiple populations)</h4>
<p>When BPP is run using the above control file a summary of the progress
of the MCMC is again printed to screen as follows:</p>
<pre><code>         |  Acceptance proportions  |
    Prgs | Gage Gspr thet  tau  mix | mthet1 mthet2 mthet3   mtau1  mtau2  mtau3       log-PG        log-L
    ------------------------------------------------------------------------------------------------------
      -3%  0.65 0.10 0.43 0.04 0.12   0.0031 0.0095 0.0068  0.0021 0.0015 0.0014   1204.63468  -4458.38814
     (some output omitted here)
       5%  0.64 0.26 0.30 0.28 0.27   0.0035 0.0096 0.0067  0.0019 0.0018 0.0017   1225.51650  -4443.85539  1:09
      10%  0.65 0.27 0.30 0.28 0.28   0.0035 0.0095 0.0068  0.0019 0.0018 0.0018   1202.62497  -4445.50020  1:47
      15%  0.64 0.27 0.30 0.28 0.28   0.0034 0.0095 0.0068  0.0019 0.0018 0.0017   1211.95125  -4446.35694  2:25
      20%  0.64 0.26 0.31 0.28 0.28   0.0034 0.0095 0.0067  0.0018 0.0018 0.0017   1211.09206  -4445.74292  3:05
      25%  0.64 0.26 0.31 0.28 0.28   0.0034 0.0096 0.0067  0.0018 0.0018 0.0017   1222.97603  -4446.12020  3:43
      30%  0.64 0.26 0.31 0.28 0.28   0.0034 0.0096 0.0067  0.0018 0.0018 0.0017   1216.60217  -4445.55424  4:21
      35%  0.64 0.26 0.31 0.28 0.28   0.0034 0.0096 0.0067  0.0018 0.0018 0.0017   1260.59722  -4445.22113  5:00
      40%  0.64 0.26 0.31 0.28 0.27   0.0034 0.0096 0.0067  0.0018 0.0018 0.0017   1269.45020  -4445.20340  5:38
      45%  0.65 0.26 0.31 0.28 0.27   0.0034 0.0096 0.0067  0.0018 0.0018 0.0017   1243.07668  -4445.00207  6:16
      50%  0.65 0.26 0.31 0.28 0.27   0.0034 0.0096 0.0067  0.0018 0.0018 0.0017   1229.59858  -4445.04036  6:55
      55%  0.64 0.26 0.31 0.28 0.27   0.0034 0.0096 0.0067  0.0018 0.0018 0.0017   1209.31273  -4445.43161  7:33
      60%  0.64 0.26 0.31 0.28 0.27   0.0034 0.0096 0.0067  0.0018 0.0018 0.0017   1206.22162  -4445.53994  8:11
      65%  0.64 0.26 0.31 0.28 0.27   0.0034 0.0096 0.0067  0.0018 0.0018 0.0017   1190.78837  -4445.21618  8:50
      70%  0.64 0.26 0.31 0.28 0.27   0.0034 0.0096 0.0067  0.0018 0.0018 0.0017   1203.82684  -4445.04124  9:30
      75%  0.64 0.26 0.31 0.28 0.27   0.0034 0.0096 0.0067  0.0018 0.0018 0.0017   1186.32176  -4445.00578  10:09
      80%  0.64 0.26 0.31 0.28 0.27   0.0034 0.0096 0.0067  0.0018 0.0018 0.0017   1222.75718  -4444.94315  10:47
      85%  0.64 0.26 0.31 0.28 0.27   0.0034 0.0096 0.0067  0.0018 0.0018 0.0017   1228.39106  -4444.68671  11:26
      90%  0.64 0.26 0.31 0.28 0.27   0.0034 0.0096 0.0067  0.0018 0.0018 0.0017   1228.72385  -4444.54844  12:04
      95%  0.64 0.26 0.31 0.28 0.27   0.0034 0.0096 0.0067  0.0018 0.0018 0.0017   1223.25921  -4444.25455  12:43
     100%  0.64 0.26 0.31 0.28 0.27   0.0034 0.0096 0.0067  0.0018 0.0018 0.0017   1199.52731  -4444.17857  13:21
    }
    13:21 spent in MCMC

              theta_1K  theta_2C  theta_3L  theta_4H  theta_5KCLH theta_6KCL theta_7KC tau_5KCLH tau_6KCL tau_7KC lnL
    mean      0.003422  0.009588  0.006683  0.003031  0.003720  0.001577  0.001648  0.001823  0.001787  0.001710  -4444.16
    median    0.003351  0.009492  0.006599  0.002932  0.003642  0.001412  0.001463  0.001809  0.001773  0.001700  -4443.69
    S.D       0.000747  0.001461  0.001118  0.000819  0.000944  0.000899  0.000954  0.000234  0.000230  0.000238  11.615936
    min       0.001309  0.004913  0.003258  0.000866  0.001131  0.000029  0.000039  0.001047  0.001043  0.000772  -4496.72
    max       0.008147  0.017569  0.012778  0.009724  0.010355  0.008461  0.008987  0.003140  0.003113  0.002847  -4403.99
    2.5%      0.002171  0.006999  0.004743  0.001729  0.002100  0.000325  0.000340  0.001404  0.001375  0.001271  -4468.01
    97.5%     0.005084  0.012720  0.009104  0.004919  0.005780  0.003751  0.003991  0.002319  0.002280  0.002209  -4422.74
    2.5%HPD   0.002071  0.006802  0.004595  0.001566  0.001949  0.000142  0.000164  0.001396  0.001342  0.001252  -4467.06
    97.5%HPD  0.004928  0.012469  0.008903  0.004650  0.005573  0.003312  0.003512  0.002308  0.002237  0.002185  -4421.90
    ESS*      2721.25   8954.56   5279.29   11762.38  1199.63   4543.27   3299.32   850.33    970.61    932.00    420.40
    Eff*      0.027213  0.089546  0.052793  0.117624  0.011996  0.045433  0.032993  0.008503  0.009706  0.009320  0.004204
</code></pre>
<p>There are again 5 acceptance proportions for the 5 parameter proposals,
but now the current mean values of 6 parameters are printed. The current
mean values of <span class="arithmatex">\(\theta\)</span> for nodes 1, 2 and 3 (<code>mthet1 mthet2 mthet3</code>)
are printed -- which are <span class="arithmatex">\(\theta\)</span>s of the contemporary species K, C and
L, and those of the 3 <span class="arithmatex">\(\tau\)</span>s (<code>mtau1 mtau2 mtau3</code>) which are the
divergence times for ancestral species KCLH, KCL and KC, respectively.
If there are many populations/species only the first few <span class="arithmatex">\(\tau\)</span>s will be
printed to screen but all the <span class="arithmatex">\(\tau\)</span>s will be available in the summary
created at the end of the run (see below). When the run finishes, a
final block is written summarizing the results. The first two rows give
the mean and median of the posterior distribution for each parameter.
This is followed by summaries of the statistical uncertainty: the
standard deviation (S.D.) of the posterior distribution; the minimum
(min) and maximum (max) values; the lower and upper bounds of the 95%
<a href="https://en.wikipedia.org/wiki/Credible_interval">Credible Interval</a>
(2.5% and 97.5%); the lower and upper bounds of the <a href="https://en.wikipedia.org/wiki/Credible_interval">Highest Posterior
Density (HPD) Interval</a>
(2.5%HPD and 97.5%HPD); the Effective Sample Size (ESS*); and the
Efficiency (Eff*). Larger values for ESS* and Eff* indicate better
mixing and more reliable estimates. The above information is also
printed (along with other technical details of the run) to the output
file specified by the <code>outfile</code> variable in the control file (in our
example, <code>out.txt</code>). A tree file in NEXUS/Newick format is also created
and placed in the file named FigTree.tre which is formatted for
viewing/printing using the
<a href="http://tree.bio.ed.ac.uk/software/figtree/">Figtree</a> program</p>
<h4 id="mcmc-output-file">MCMC output file</h4>
<p>A file will be produced (with a name specified by the control file
variable <code>mcmcfile</code>) that contains the mcmc samples from the run for all
the <span class="arithmatex">\(\theta\)</span> and <span class="arithmatex">\(\tau\)</span> parameters. When running under analysis method
A00 the contents of this file can be analyzed using a program such as
<a href="">Tracer</a> to visually examine the convergence and mixing of the MCMC
run. However, the other three analysis methods (A01, A10 and A11) are
trans-model analyses in which the number of parameters and or/the
meaning of the parameters changes as the chain runs so it is not correct
to examine the trace plot without conditioning on a particular model.</p>
<h4 id="adjusting-step-lengths-for-mcmc-moves-finetune">Adjusting step lengths for MCMC moves (finetune)</h4>
<p>You can use the automatic adjustment of the finetune variable (MCMC step
lengths), which appears to be reliable, but make sure that the
acceptance proportions are neither too small nor too large. Below are
notes for manual adjustments of the step lengths. Often I use automatic
adjustments to generate good step lengths and then copy them into the
control file, so that the next time the automatic procedure starts with
already good step lengths.</p>
<p>Below are some notes about adjusting the step lengths manually (finetune
= 0).</p>
<p>First note the following line in the control file ChenLi2001.bpp.ctl.
Here finetune = 0 means that MCMC step lengths will not be adjusted
automatically, and they are used as given.</p>
<pre><code>    finetune = 0: .01 .02 .03 .04 .05 .01 .01  # auto (0 or 1): MCMC step lengths
</code></pre>
<p>There are seven finetune steplengths here. They are in a fixed order and
always read by the program even if the concerned proposal is not used.
The first five of them are <span class="arithmatex">\(\epsilon_1\)</span>, <span class="arithmatex">\(\epsilon_2\)</span>, <span class="arithmatex">\(\epsilon_3\)</span>,
<span class="arithmatex">\(\epsilon_4\)</span>, and <span class="arithmatex">\(\epsilon_5\)</span>, described in (<a href="https://doi.org/10.1093/genetics/164.4.1645">Rannala and Yang 2003</a>. These are
the step lengths used in the MCMC proposals that (1) change internal
node ages in the gene tree, (2) prune and re-graft nodes in the gene
tree, (3) update <span class="arithmatex">\(\theta\)</span>, (4) update <span class="arithmatex">\(\tau\)</span>s using the rubber-band
algorithm, and (5) implements the mixing step. The 6th and 7th are for
the proposals that change the locus rates or heredity multipliers and
that change the sequencing errors, respectively. If the model assumes
the same rate for all loci and does not use heredity multipliers, the
6th proposal step is not used. If the model assumes no sequencing
errors, the 7th step is not used. The acceptance proportions for the
first five proposals are always printed out on the screen, but those for
the 6th and 7th are printed out only if the concerned proposal is used
in the model. In the example above, only the first five proposals are
used in the algorithm and we will change the step lengths so that the
acceptance proportions become close to 30%. If the acceptance proportion
is too small (say, \&lt;0.10)), decrease the corresponding finetune
parameter. If the acceptance proportion is too large (say, &gt;0.80),
increase the finetune parameter.</p>
<p>Run the program for a small number of iterations and look at the screen
output for the acceptance proportions.</p>
<pre><code>    lnpG0 = 2440.0952  lnL0 = -43946.6769
     0% 0.13 0.00 0.00 0.04 0.43  0.0009 0.0009 0.0008  0.0150 0.0067 0.0057 2399.70 -43752.8348  0:05
     5% 0.14 0.00 0.00 0.04 0.44  0.0009 0.0009 0.0008  0.0148 0.0064 0.0057 2340.71 -43738.0172  0:09
    10% 0.14 0.00 0.00 0.04 0.44  0.0009 0.0009 0.0008  0.0149 0.0064 0.0056 2383.91 -43739.3830  0:14
</code></pre>
<p>Here the second acceptance proportion, at 0.00, is much too small, which
means that the corresponding step-length (0.02 in the control file) is
much too large. Terminate the run (Ctrl-C) and decrease the value in the
control file. Then run the program again (use the up and down arrow keys
to retrieve past commands). Repeat this process a few times until every
acceptance proportion is neither too small nor too large. In this
example, changing 0.02 to 0.002 brings the acceptance proportion to
 24%, which is slightly too small but already good enough.</p>
<p>Those MCMC proposals are used in all four analyses (A00, A01, A10, A11),
so that the description here applies to all of them. Note that the
finetune parameters affect the efficiency of the MCMC or how fast one
can obtain reliable results. In theory they do not change the results if
all runs using different finetune parameters are long enough to generate
reliable results.</p>
<p><code>A01</code> means species tree estimation when the assignments and
delimitation are fixed.</p>
<pre><code>    speciesdelimitation = 0    *
    speciestree = 1    * NNI/SPR over species trees
    speciesmodelprior = 1  * 0: uniform LH; 1:uniform rooted trees; 2: uniformSLH; 3: uniformSRooted
</code></pre>
<p>This invokes the NNI or SPR algorithm to change the species tree
topology, while species delimitation is fixed (so that the number of
species and the assignment of individuals to species are fixed).</p>
<p>A10, for species delimitation using a user-specified guide tree
[@Yang2010; @Rannala2013], is specified using</p>
<pre><code>    speciesdelimitation = 1 0 2   * speciesdelimitation algorithm0 and finetune(e)
    speciesdelimitation = 1 1 2 1 * speciesdelimitation algorithm1 finetune (a m)
    speciesmodelprior = 1  * 0: uniform LH; 1:uniform rooted trees
</code></pre>
<p>The first line specifies rjMCMC algorithm 0, with <span class="arithmatex">\(\epsilon = 2\)</span> in
equations 3 and 4 of [@Yang2010]. Reasonable values for <span class="arithmatex">\(\epsilon\)</span> are
1, 2, 5, etc.\
The second line specifies rjMCMC algorithm 1, with <span class="arithmatex">\(\alpha = 2\)</span> and
<span class="arithmatex">\(m = 1\)</span> in equations 6 and 7 of [@Yang2010]. Reasonable values are
<span class="arithmatex">\(\alpha = 1, 1.5, 2\)</span>, etc. and <span class="arithmatex">\(m = 0.5, 1, 2\)</span>, etc.</p>
<p>The two algorithms in theory should produce identical results. The
variable <code>speciesmodelprior</code> specifies Priors 0 and 1. Prior 0 means
equal probabilities for labeled histories (which are rooted trees with
internal nodes ordered by their age). This is the prior used by
[@Yang2010 eq. 2]. Prior 1 means equal probabilities for rooted trees.
This is now the default. The prior with the user specified probabilities
for nodes described by [@Rannala2013] is deleted since version 2.3(?).</p>
<p><code>A11</code>, for joint species delimitation and species tree inference or for
unguided species delimitation [@Yang2014a], is specified as follows.</p>
<pre><code>    speciesdelimitation = 1 0 2    * rjMCMC speciesdelimitation algorithm0(e)
    *speciesdelimitation = 1 1 2 1  * rjMCMC speciesdelimitation algorithm1(a m)
    speciestree = 1    * NNI over species trees
    speciesmodelprior = 1  * 0: uniform LH; 1:uniform rooted trees; 2: uniformSLH; 3: uniformSRooted
</code></pre>
<p>In this case, BPP will use the rjMCMC algorithm (either
algorithm 0 or algorithm 1 of [@Yang2010] to change the species
delimitation model and the NNI/SPR move to change the species tree
topology.</p>
<p>For A11, <code>speciesmodelprior</code> can take the four values 0, 1, 2, 3, which
mean Priors 0, 1, 2, 3, respectively. As mentioned above, Prior 1 (which
is the default) assigns equal probabilities to the rooted species trees,
while Prior 0 means equal probabilities for the labeled histories
(rooted trees with the internal nodes ordered by age). Priors 2 and 3
assign equal probabilities for the numbers of species (1/s each for 1,
2, ..., s species given s populations) and then divided up the
probability for any specific number of species among the compatible
models (of species delimitation and species phylogeny) either uniformly
[Prior 3] or in proportion to the labeled histories [Prior 2].
Priors 2 and 3 are mentioned by [@Yang2014a] and implemented by
[@Yang2015]. Prior 3 may be suitable when there are many populations.</p>
<h3 id="a01-species-tree-estimation">A01: Species Tree Estimation</h3>
<h4 id="input-a01">Input A01</h4>
<h4 id="output-a01">Output A01</h4>
<h4 id="mcmc-output-file_1">MCMC output file</h4>
<p>Suppose we use the control file bpp.4s.ctl to run analysis A01: species
tree estimation with species delimitation and assignment fixed
(speciesdelimitation = 0, speciestree = 1).</p>
<p>The screen output will look like this:</p>
<pre><code>     5% 0.28 0.28 0.28 0.35 0.29  0.0000  0.0033 0.0138 2403.43 -43721.0651  0:13
    10% 0.27 0.27 0.27 0.36 0.30  0.0000  0.0028 0.0142 2304.08 -43723.0804  0:21
    15% 0.26 0.27 0.26 0.36 0.30  0.0000  0.0025 0.0143 2308.11 -43723.8564  0:27
      ^^ Pjump for MCMC moves ^^  PSPR   theta  tau     lnpG    E(lnL)
</code></pre>
<p>Here the five Pjump values are the acceptance proportions for the five
conventional MCMC moves, as discussed above. PSPR is the acceptance
proportion for the SPR/Nodeslider moves that change the species
phylogeny [@Rannala2017]. The next two numbers are posterior means of
<span class="arithmatex">\(\theta\)</span> for the root population (-1 is printed if <span class="arithmatex">\(\theta\)</span>s are
integrated out) and <span class="arithmatex">\(\tau_0\)</span> for the root age. The last two numbers are
the log MSC gene-tree density (<a href="https://doi.org/10.1093/genetics/164.4.1645">Rannala and Yang 2003</a> and the average log
sequence likelihood [@Felsenstein1981].</p>
<p><strong>The MCMC sample</strong> of species trees is collected in the file mcmc.txt.
Below are two lines from that file. The numbers after : are the branch
lengths (<span class="arithmatex">\(\tau\)</span>s), while those after # are <span class="arithmatex">\(\theta\)</span>s. I hope to change
the output format to be more consistent with the Newark format.</p>
<pre><code>    (((H #0.000708: 0.004728, C #0.000740: 0.004728) #0.002814: 0.001190, G #0.000574: 0.005918) #0.004132: 0.007626, O #0.002036: 0.013544) #0.003835;
    (((H #0.000708: 0.004728, C #0.000740: 0.004728) #0.002814: 0.001190, G #0.000574: 0.005918) #0.003740: 0.007351, O #0.002786: 0.013268) #0.003506;
</code></pre>
<p>The BPP summary of the sample will look like the
following.</p>
<pre><code>    Read tree sample, count trees &amp; splits
    tree 20000  (((H, C), G), O);
    20001 trees read, 1 distinct trees.

    Species in order:
    1. H
    2. C
    3. G
    4. O

    (A) Best trees in the sample (1 distinct trees in all)
    20001  1.00000  1.00000  (O, (G, (H, C)));

    (B) Best splits in the sample of trees (2 splits in all)
    20001   1.00000  1100
    20001   1.00000  1110

    (C) Majority-rule consensus tree
    (O, (G, (H, C) #1.000000) #1.000000);

    (D) Best tree (or trees from the mastertree file) with support values
    (O, (G, (H, C) #1.000000) #1.000000);  [P = 1.00000]
</code></pre>
<p>Section (A) lists the species trees in decreasing order of posterior
probabilities. From this you can easily identify the 95% or 99%
credibility set of species trees. Section (B) lists the splits (or
bipartitions) and their posterior probabilities. The splits here take
into account the location of the root, and may be different from the
splits for unrooted trees. Section (C) prints the majority-rule
consensus tree, with posterior probabilities for nodes.</p>
<h3 id="a10-species-delimitation">A10: Species Delimitation</h3>
<h4 id="input-a10">Input A10</h4>
<h4 id="output-a10">Output A10</h4>
<h4 id="mcmc-output-file_2">MCMC output file</h4>
<p>We apply rjMCMC algorithm to the frogs data (A10:
<code>speciesdelimitation = 1, speciestree = 0</code>).</p>
<pre><code>    cd frogs\r1
    ..\..\bin\bpp ..\A10.bpp.ctl
</code></pre>
<p><strong>The screen output</strong> will look like the following. The species
delimitation models that can be generated from the fixed guide tree are
listed, together with their prior probabilities calculated by
BPP. (As a check, if you use <code>usedata = 0</code>, the MCMC
should be sampling from this prior distribution.) The species
delimitation model is represented using four 0-1 flags for the four
interior nodes 6, 7, 8, 9 in the guide tree, with 0 for 'collapsed' and
1 for 'resolved'. Note that the tips in the guide tree are numbered 1,
2, <span class="arithmatex">\(\cdots, s\)</span> for <span class="arithmatex">\(s\)</span> potential species, while the interior (ancestral)
nodes are numbered <span class="arithmatex">\(s + 1, s + 2, \cdots, 2s – 1\)</span>, with <span class="arithmatex">\(s + 1\)</span> to be
the root of the guide tree. The numbering is through a tree-traversal
algorithm, fixed by the program. This same order is used to specify the
divergence time parameters (<span class="arithmatex">\(\tau\)</span>s), so you can work out the order by
looking at the list of nodes in the screen output (look at the
"population by population table", "# species divergence times in the
order:", etc.).</p>
<pre><code>    Number of species-delimitation models =  5
    delimitation model   1: 000  prior  0.20000
    delimitation model   2: 100  prior  0.20000
    delimitation model   3: 101  prior  0.20000
    delimitation model   4: 110  prior  0.20000
    delimitation model   5: 111  prior  0.20000

    [Note: Ancestral nodes in order:   5 KCLH  6 KC  7 LH]

    MCMC settings: 8000 burnin, sampling every 2, 100000 samples
    Approximating posterior, using sequence data
    (Settings: cleandata=1 print=1 saveconP=1 moveinnode=1)

    Starting rjMCMC...
    PrSplit = 0.500000
    rj algorithm 1: new theta from G(a=2.00, m=1.00)

    Starting species-delimitation model: 111

    root dist = 0.00095

    Initial parameters, np = 3.
    Genetrees generated from the MSC density.
    0.00108  0.00074  0.00075
    lnpG0 =  752.4533  lnL0 = -3402.2670
    -3% 0.70 0.13 0.00 0.20 0.36   3 0.0198 111 P[5]=0.8750  -1.0000 0.0007  787.23 -3127.1399
    (nsteps = 5)
    Current Pjump:      0.69618  0.13189  0.00000  0.19758  0.35650
    Current finetune:   5.00000  0.00100  0.00100  0.00100  0.30000
    New     finetune:  18.97702  0.00041  0.00001  0.00063  0.36913

    -2% 0.70 0.30 0.00 0.19 0.27   3 0.0209 111 P[5]=0.8575  -1.0000 0.0007  775.05 -3125.7744
    (nsteps = 5)
    Current Pjump:      0.69565  0.29861  0.00000  0.18742  0.27050
    Current finetune:  18.97702  0.00041  0.00001  0.00063  0.36913
    New     finetune:  71.87933  0.00041  0.00000  0.00037  0.32779

    -1% 0.70 0.31 0.00 0.20 0.33   3 0.0126 111 P[5]=0.8990  -1.0000 0.0007  732.14 -3126.0675
    (nsteps = 5)
    Current Pjump:      0.69528  0.30537  0.00000  0.20442  0.33250
    Current finetune:  71.87933  0.00041  0.00000  0.00037  0.32779
    New     finetune:  99.00000  0.00042  0.00000  0.00024  0.37030

     0% 0.69 0.30 0.00 0.23 0.26   3 0.0165 111 P[5]=0.9215  -1.0000 0.0007  723.36 -3125.7568  0:11

    (nsteps = 5)
    Current Pjump:      0.69354  0.29812  0.00000  0.23367  0.26400
    Current finetune:  99.00000  0.00042  0.00000  0.00024  0.37030
    New     finetune:  99.00000  0.00042  0.00000  0.00018  0.31993

     5% 0.70 0.30 0.00 0.29 0.34   2 0.0260 110 P[5]=0.8731  -1.0000 0.0007  773.34 -3125.8363  0:25
    10% 0.70 0.30 0.00 0.30 0.34   3 0.0234 111 P[5]=0.8778  -1.0000 0.0007  745.39 -3125.8263  0:39
    15% 0.70 0.30 0.00 0.30 0.34   2 0.0225 110 P[5]=0.8879  -1.0000 0.0007  762.56 -3125.9276  0:53

    ^^ Pjump for MCMC moves ^^       Prj      P[model 5]     theta    tau   lnpG     E(lnL)
</code></pre>
<p>The starting species delimitation model is generated by choosing at
random one of the models defined by the guide tree or starting tree.</p>
<p>After the chain has started, the five ratios after the % sign are the
acceptance proportions for the conventional MCMC moves discussed above.</p>
<p>Next there are three numbers related to the rjMCMC move, highlighted in
red above. The rest of the line shows the posterior mean for <span class="arithmatex">\(\theta\)</span>
for the root and posterior mean for <span class="arithmatex">\(\tau_0\)</span> (which exist in all
species-tree or species delimitation models), the current log MSC
density and average log sequence likelihood The three numbers related to
the rjMCMC move, "2 0.0225 110" in the example, mean that the current
model is 110 (using the flags of fig. 1 in Yang and Rannala, 2010), and
the rjMCMC move has the acceptance proportion 0.0225. [What does the
number 2 mean?] In general the larger this proportion, the more
efficient the rjMCMC algorithm is. However there is no optimal
acceptance proportion for the rjMCMC move, and a value close to 0 may
not necessarily mean a problem. If one model has posterior probability
close to 1, the acceptance proportion should be near 0 as well. Thus
both poor mixing of the rjMCMC algorithm and extreme posterior model
probabilities can cause the acceptance proportion for the rjMCMC to be
close to 0. It has been noted that if the rjMCMC algorithm is suffering
from poor mixing, different starting species trees often lead to
different results.</p>
<p>Next the posterior probability for the best species-tree model (the most
frequently visited tree model up to now) is printed. In the example,
"P[5]=0.8879" means tree model 5 (111) is the most favoured model,
with the posterior at 0.8879.</p>
<p>After the MCMC is finished, the program will summarize the sample. The
output looks like the following. The seven delimitation models are
listed again, together with their posterior and prior probabilities.</p>
<pre><code>    Summarizing the species-delimitation sample in file mcmc.txt

    Number of species-delimitation models =  5
    model    prior  posterior
    1  000   0.20000   0.00000
    2  100   0.20000   0.00000
    3  101   0.20000   0.00038
    4  110   0.20000   0.10019
    5  111   0.20000   0.89943

    [Note: Ancestral nodes in order:   5 KCLH  6 KC  7 LH]

    Guide tree with posterior probability for presence of nodes
    ((K, C) #0.999620, (L, H) #0.899810) #1.000000;
</code></pre>
<p>The MCMC sample file mcmc.txt. As the number of parameters changes when
the rjMCMC moves between models, the mcmc sample file may not be very
useful, so you can ignore it. Right now the header line is generated
using the starting species tree and should be ignored. After the header
line, each line of output has the following numbers, separated by TABs:
iteration number, the number of parameters, the tree, the sampled
parameter values, and lnL.</p>
<pre><code>    Gen np  tree    tau_5KCLH   tau_6KC tau_7LH lnL
    2   3   111 0.00079675  0.0007372   0.00065989  -3122.248
    4   3   111 0.00074995  0.00067776  0.00059457  -3122.427
    ...
    786 3   111 0.00037556  0.00010047  2.5634e-05  -3129.184
    788 3   111 0.00065693  7.9848e-05  3.6489e-05  -3116.988
    790 2   110 0.00044831  5.7933e-05  -3125.574
    792 2   110 0.00060611  0.00029978  -3122.610
    794 2   110 0.00060611  0.00048884  -3122.912
    796 2   110 0.00041675  0.00027494  -3130.812
    798 2   110 0.00034214  0.0002618   -3142.615
    800 3   111 0.00044038  0.00033698  9.8583e-05  -3129.386
    802 3   111 0.00060202  0.00047807  7.7594e-05  -3121.725
</code></pre>
<p>If you know the unix command grep, you can retrieve the lines for the
same tree model to summarize the posterior for parameters in that model.</p>
<pre><code>    grep &quot;3 111&quot; mcmc.txt &gt; result.Tree111.txt
</code></pre>
<p>In theory this should give you the same posterior as if you run analysis
A00 with the species tree fixed at tree 111. In practice I think it is
simpler that you edit the Imap file and the control file to run analysis
A00.</p>
<p><strong>Running rjMCMC Algorithms.</strong></p>
<ul>
<li>
<p>The rjMCMC algorithms for species delimitation allow the chain to
    move from one model to another but can have mixing problems. Check
    that similar results are produced from multiple runs using
    Algorithm0 and Algorithm1 regardless of the starting species tree.</p>
<p>The starting tree is chosen by the program at random and will vary
among runs. Make sure that some of your runs are started with the
one-species model, some from the fully resolved tree, and some from
other trees in between. If you get consistent results among runs
with different starting trees and using the two algorithms, you are
unlikely to have a convergence or mixing problem.</p>
<p>If you have a computer with multicore, you can run those different
combinations or replicates in different folders at the same time.</p>
</li>
<li>
<p>In medium-sized or large datasets with multiple loci, we have come
    across cases where the chain is stuck at the one-species model, or
    have difficulty moving into the one-species model. The problem is
    less common when there are only 1 or 2 loci. If you have a similar
    problem, you may start the analysis with 1 locus, then 2 loci, etc.
    to observe how the results change (you can do this by changing nloci
    in the control file as there is no need to change the sequence
    file).</p>
</li>
<li>
<p>The program used the burnin to adjust the step lengths for the
    proposals in the MCMC algorithm. If the rjMCMC stays in species
    model 0, which does not have any <span class="arithmatex">\(\tau\)</span> parameter, no information is
    collected during the burnin about the proposals to change <span class="arithmatex">\(\tau\)</span> and
    the automatically adjusted step length for <span class="arithmatex">\(\tau\)</span> can be very poor.</p>
</li>
<li>
<p>You probably need to evaluate the impact of the priors on <span class="arithmatex">\(\theta\)</span>s
    and <span class="arithmatex">\(\tau\)</span>. See the note about the gamma distribution later in this
    document.</p>
</li>
</ul>
<h3 id="a11-joint-species-delimitation-and-species-tree-estimation">A11: Joint Species Delimitation and Species Tree Estimation</h3>
<h4 id="input-a11">Input A11</h4>
<h4 id="output-a11">Output A11</h4>
<h4 id="mcmc-output-file_3">MCMC output file</h4>
<p>Again, we use the frogs example (<code>A11.bpp.ctl</code>) to illustrate joint
species delimitation and species tree estimation (A11: <code>speciesdelimitation = 1, speciestree = 1</code>).</p>
<p>The screen output looks like this:</p>
<pre><code>    Initial parameters, np = 3.
    Genetrees generated from the MSC density.

    lnpG0 =  711.7015  lnL0 = -3350.8540
    -3% 0.70 0.13 0.00 0.22 0.35   4  3 0.0039 0.1455 P(4)=0.9965  -1.0000 0.0007  775.17 -3126.4790
    -2% 0.70 0.31 0.00 0.21 0.28   4  3 0.0100 0.1260 P(4)=0.9925  -1.0000 0.0008  745.60 -3125.8524
    -1% 0.70 0.30 0.00 0.22 0.35   4  3 0.0039 0.1000 P(4)=0.9980  -1.0000 0.0007  750.41 -3126.2339
     0% 0.70 0.30 0.00 0.23 0.25   4  3 0.0077 0.0920 P(4)=0.9845  -1.0000 0.0009  775.04 -3126.2217  0:13
     5% 0.70 0.30 0.00 0.24 0.34   4  3 0.0231 0.0967 P(4)=0.9368  -1.0000 0.0007  750.67 -3125.9098  0:30
    10% 0.70 0.30 0.00 0.24 0.34   4  3 0.0189 0.0954 P(4)=0.9474  -1.0000 0.0007  732.19 -3125.9363  0:46
      ^^ Pjump for MCMC moves ^^   S  p   Prj   PSPR  P(S=4)       theta   tau0    lnpG    E(lnL)
</code></pre>
<p>Here S is the number of species in the current model, and p is the
number of parameters in the current model. Prj is acceptance proportion
for the rjMCMC moves, while PNNI is acceptance proportion for the NNI or
SPR moves. P(4)=0.9474 means that based on the states sampled, the
number of species 4 has the highest posterior probability (compared with
1 species, 2 species, etc.) and the probability is 0.9474. The next two
numbers are posterior means of <span class="arithmatex">\(\theta\)</span> for the root population (-1 is
printed if <span class="arithmatex">\(\theta\)</span>s are integrated out) and <span class="arithmatex">\(\tau_0\)</span> for the root age.
These are followed by the log MSC gene-tree density and the average log
likelihood.</p>
<p>The summary of the sample by BPP has four sections, as
follows.</p>
<pre><code>    Summarizing the species-tree sample in file mcmc.txt
    read tree 100000  ((K, C), (H, L));
    (A) List of best models (count postP #species SpeciesTree)
    16131  0.16131  0.16131   4 (K C L H)   ((L, H), (K, C));     0011 1100
    12736  0.12736  0.28867   4 (K C L H)   (K, (C, (L, H)));     0011 0111
     9552  0.09552  0.38419   4 (K C L H)   (H, (L, (K, C)));     1100 1110
     8009  0.08009  0.46428   4 (K C L H)   (L, (H, (K, C)));     1100 1101
     7901  0.07901  0.54329   4 (K C L H)   (C, (K, (L, H)));     0011 1011
     6638  0.06638  0.60967   4 (K C L H)   (K, (L, (C, H)));     0101 0111
    ...

    (B)  5 species delimitations &amp; their posterior probabilities
    95530   0.95530   4 (K C L H)
     3638   0.03638   3 (K C LH)
      533   0.00533   3 (KH C L)
      291   0.00291   3 (K CH L)
        8   0.00008   3 (KC L H)

    (C)  8 delimited species &amp; their posterior probabilities
    99701   0.99701  C
    99459   0.99459  K
    96362   0.96362  L
    95538   0.95538  H
     3638   0.03638  LH
      533   0.00533  KH
      291   0.00291  CH
        8   0.00008  KC

    (D) Posterior probability for # of species
    P[ 1] =   0.0000  prior[ 1] =  0.23810
    P[ 2] =   0.0000  prior[ 2] =  0.23810
    P[ 3] =   0.0447  prior[ 3] =  0.28571
    P[ 4] =   0.9553  prior[ 4] =  0.23810
</code></pre>
<p>Section (A) lists the best models in the decreasing order of the
posterior probabilities. Here a model is a full MSC model that species
both the species delimitation and species phylogeny. From this section,
one can easily construct the 95% or 99% credibility sets of models. This
section is further summarized to produce sections B, C, and D. Section
(B) gives the posterior probabilities for the top few delimitations.
Section (C) lists the delimited species and their posterior
probabilities, and section (D) lists the posterior probability for the
number of species, together with the prior probabilities calculated by
BPP.</p>
<h2 id="advanced-features-of-bpp">Advanced Features of BPP</h2>
<h3 id="threading-and-checkpointing">Threading and Checkpointing</h3>
<h4 id="threads">Threads</h4>
<p>BPP4 can use Single-Instruction-Multiple-Data (SIMD)
instruction sets (also called vector instructions) available on modern
processors. It automatically detects and uses the best instruction set
available so you don't have to do anything about this. However, you can
manually specify and force an instruction set in the control file using
the arch tag. For instance, arch=SSE forces BPP to use the
SSE instruction set even if AVX (which has twice the register size of
SSE and thus in theory yields twice the speedup) is present on the
system. To completely disable vector instruction, one can use the
arch=CPU option. The available values for the arch tag are 'CPU', 'SSE',
'AVX' and 'AVX2'.</p>
<p><img alt="Fig-Numa" src="https://github.com/bpp/bpp-manual/blob/191d107413566b145360597fdfdfef3e68ce9173/figures/fig-numa.png?raw=true" /></p>
<p>Modern computer architecture using Non-Uniform momery access (NUMA).</p>
<p>BPP4 can use multiple cores/threads. The control variable threads has the following syntax.</p>
<pre><code>    Threads = 8 19 1
</code></pre>
<p>This means BPP will use 8 threads, starting from
core/thread 19, with increment 1; in other words hardware threads 19-36
will be used. The software threads are pinned onto the hardware threads
and they not allowed to migrate. We found that this helps with
performance.</p>
<p>Modern clusters based on microprocessors use a common shared-memory
configuration called NUMA (for non-uniform memory access). The cluster
typically consists of two or four microprocessors interconnected on a
local bus to a shared memory on a single motherboard.</p>
<p>You can use commands like <code>lscpu</code> to see the machine architecture. For
example, on our server lscpu gives the following output:</p>
<pre><code>    Architecture:          x86_64
    CPU op-mode(s):        32-bit, 64-bit
    Byte Order:            Little Endian
    CPU(s):                144
    On-line CPU(s) list:   0-143
    Thread(s) per core:    2
    Core(s) per socket:    18
    Socket(s):             4
    NUMA node(s):          4
    ...
    NUMA node0 CPU(s):     0-17,72-89
    NUMA node1 CPU(s):     18-35,90-107
    NUMA node2 CPU(s):     36-53,108-125
    NUMA node3 CPU(s):     54-71,126-143
</code></pre>
<p>This means that hardware threads 1-18 are on CPU0, while threads 73-90
are the hyperthreads on CPU0. The specification Threads = 8 19 1 thus
uses 8 threads on CPU1. Our recommendation is that you do some tests
yourself using threads = 2 or 4 or 8, and use options that work well for
your servers. Adjust the step lengths first. Then use burnin=0 and a
small nsample so that the running time is about 1-2 minutes. Change
threads and record running time.</p>
<p>Make sure all threads for the same BPP job are on the same
CPU. Use lscpu to see the machine configuration and htop to examine the
load on the hardware threads.</p>
<h4 id="checkpointing">Checkpointing</h4>
<p>This option instructs BPP to create a checkpoint file
(save the progress) after a specified number of MCMC iterations. The
MCMC loop in BPP consists of
<code>N = burnin + sampfreq*nsample</code> iterations, numbered as <span class="arithmatex">\(1,2,\cdots,N\)</span>.
There are two formats, where X, Y are whole numbers:</p>
<pre><code>    checkpoint = X
    checkpoint = X Y
</code></pre>
<p>In the first format, a single checkpoint is created after X MCMC
iterations (including burnin).</p>
<p>In the second format, a checkpoint is created after X steps, and then
additional checkpoints are created every Y steps. For example, if X =
10000 and Y = 10000, the first checkpoint is created at MCMC iteration
10000, and further checkpoints at iterations 20000, 30000, etc.</p>
<p>The checkpoint files are named OUTFILE.Z.chk, where OUTFILE is the label
specified for the outfile option, and Z is the number of the checkpoint
file, starting from 1 and incrementing with every new checkpoint file.</p>
<p>To resume from a checkpoint file, use the resume switch, i.e.</p>
<pre><code>    bpp --resume checkpoint-file.chk
</code></pre>
<h3 id="marginal-likelihood-calculations">Marginal Likelihood Calculations</h3>
<p>The C program BFdriver generates control files and job subscription
scripts for running MCMC (using BPP or MCMCtree) to calculate the
marginal likelihood (or the Bayes factor), as described in @Rannala2017.</p>
<p>The program takes a control file you provide (such as bpp.ctl) and
generates <span class="arithmatex">\(K = 16\)</span> control files with different beta values, which are
used to run bpp to sample from the different power posterior
distributions. The program also generates job submission scripts and
submit the jobs using qsub. All generated control files and output files
are in the same current directory. The frogs dataset in the bpp release
[@Yang2015] is used as the example.</p>
<p>You need the following: a linux system with SUN grid engine managing job
submission (including commonds such as <code>qsub</code>, <code>qstat</code>, <code>qdel</code>, etc.),
and a C compiler. If you don't have this job submission system, you can
use BFdriver to generate the control files and run the MCMC jobs from
the command line.</p>
<h4 id="compiling-and-running-bfdriver">Compiling and running BFdriver</h4>
<pre><code>    cc -o BFdriver -O3 BFdriver.c tools.c -lm
    BFdriver &lt;controlfilename&gt; &lt;npoints&gt; &lt;scriptname.sh&gt;
    BFdriver A00.ctl 16 tmp.sh
</code></pre>
<p>You may need to edit the following two lines inside BFdriver.c, and if
you do, remember to compile the program after editing. Here bpp is
assumed to be on your search path. You can use a full path for the
executable prorgam, such as <code>/home/gooduser/bin/bpp3.3</code>. Also the second
line is for submitting the jobs using <code>qsub</code>. Here the limits are set to
4G of RAM and 360 hours of running time. Check those values if necessary
(and recompile).</p>
<pre><code>    fprintf(fcommand, &quot;     echo \&quot;bpp %s.b$I.ctl &gt; log.b$I.txt\&quot; &gt; %s\n&quot;, ctlf, scriptf);
    fprintf(fcommand, &quot;     qsub -S /bin/bash -l h_vmem=4G -l tmem=4G -l h_rt=360:0:0 -cwd %s\n&quot;, scriptf);
</code></pre>
<h4 id="running-the-program">Running the program</h4>
<p>Create a folder inside <code>frogs/</code>, say <code>bf1</code>:</p>
<pre><code>    mkdir frogs/bf1
    cd frogs/bf1
</code></pre>
<p>Prepare a control file (A00.ctl, say) for the A00 analysis in the
folder. Check that it works. This should have a fixed species tree,
which is ((K, C), (L, H)). Species delimitation and species tree
estimation should be turned off. The control file specifies the priors
for theta, tau, and also specifies burnin, nsample, sampfreq etc. Run
bpp to confirm that the control file works. Then run BDdriver as
follows:</p>
<pre><code>    BFdriver A00.ctl 16 tree1.sh
</code></pre>
<p>Here A00.ctl is the control file we have prepared. <span class="arithmatex">\(K = 16\)</span> is the
number of points in the Gauss-Legendre qradrature algorithm for
numerical integration. You can use 8 for testing, and 16 or 32 for real
calculation. <code>tree1.sh</code> is the temporary script file for job submission
using <code>qsub</code>. The BFdriver command does a few things. First it reads the
control file specified (<code>A00.ctl</code>) and creates 16 control files with
names like <code>A00.b01.ctl</code>, ..., <code>A00.b16.ctl</code>. Each of those control
files has the same content as <code>A00.ctl</code> except that one extra line is
inserted at the beginning, like the following</p>
<pre><code>    BayesFactorBeta = 0.122298 *  w=0.124629.ctl
</code></pre>
<p>This specifies the beta value when the control file is used to run
BPP.</p>
<p>Second <code>BFdriver</code> creates a file named <code>betaweights.txt</code>, which lists
the beta values and Gauss-Legendre weights. I have copied those values
into an excel file in <code>frogs/BFdriver.frogs.xls</code>.</p>
<p>Third <code>BFdriver</code> creates a file named commands, which has the bash shell
scripts for submitting the 16 jobs using <code>qsub</code>. You use the following
to submit the jobs.</p>
<pre><code>    source commands
</code></pre>
<p>You can look at the content of <code>tree1.sh</code> to see the script for the last
job:</p>
<pre><code>    more tree1.sh
</code></pre>
<p>which should have the content like the following:</p>
<pre><code>    bpp --cfile bpp.b16.ctl &gt; log.b16.txt
</code></pre>
<p>You can use <code>qstat</code> to check the status of the 16 jobs you have
submitted. When the jobs are running, they generate output files in the
current folder, such as <code>mcmc.b01.txt</code>, <code>out.b01.txt</code>, and <code>log.b01.txt</code>
(which logs the screen output). After all jobs are finished, you can use
grep to extract the line with <code>BFbeta</code> from screen log (this command is
at the bottom of the file commands).</p>
<pre><code>    grep BFbeta log.b*.txt
</code></pre>
<p>Then copy the ElnfX values into the excel file, and estimate the
logarithm of the marginal likelihood by summing (<code>weights * ElnfX / 2</code>)
over the 16 points. This gives the log marginal likelihood to be
<span class="arithmatex">\(-3185.93\)</span>.</p>
<h4 id="exercise-and-results">Exercise and results</h4>
<p>Duplicate the calculation for the alternative species tree (tree2) in
the folder /bf2. Change the species tree topology to (((L, H), C), K),
and use tree2.sh as the temporary job script file. Everything else
should be the same as described above.</p>
<p>My runs gave the log marginal likelihood for tree 2 to be <span class="arithmatex">\(-3186.14\)</span>.
The ratio of the posterior probabilities for the two trees is then
estimated to be
<span class="arithmatex">\(P_1/P_2 = \exp\{-3185.93 - (-3186.14)\} = exp\{0.21\} = 1.24\)</span>. With
BPP4.0 and the inverse-gamma priors on <span class="arithmatex">\(\theta\)</span> and
<span class="arithmatex">\(\tau\)</span>s, the MCMC runs (A01) gave the posterior probabilities for trees
1 and 2 as 0.167 and 0.137, with the ratio 1.22. The MCMC runs and the
marginal likelihood calculations seem to agree with each other. (Note
that with [bpp3]{.smallcaps} and the gamma priors on <span class="arithmatex">\(\theta\)</span> and
<span class="arithmatex">\(\tau\)</span>s, the posteriors are 0.16 and 0.13 with the ratio 1.2 [@Yang2015
fig. 4].</p>
<h4 id="common-errors-and-problems">Common errors and problems</h4>
<p>Check the bpp control file by running bpp at the command line before
submitting the jobs. Make sure that the bpp program is on your path or
use a full path. You may have to edit the source file <code>BFdriver.c</code> and
recompile.</p>
<h3 id="simulation-using-bpp">Simulation Using BPP</h3>
<p>The simulation option of [bpp4]{.smallcaps} is the MCCOAL program in
version 3.4 and earlier. This can be used to simulate gene trees and
sequence alignments at multiple loci under the MSC (<a href="https://doi.org/10.1093/genetics/164.4.1645">Rannala and Yang 2003</a> and
MSC-I <a href="https://doi.org/10.1093/molbev/msz296">Flouri et al 2020</a> models. While the inference program
BPP implements the JC model only, the simulation program
allows GTR+G (and its special cases) and allows among-loci heterogeneity
in the process of sequence evolution. The different loci can have
different exchangeability parameters (<span class="arithmatex">\(a, b, c, d, e, f\)</span>) and base
compositions in the GTR model, different overall evolutionary rates,
different extent of among-site rate variation (reflected in the alpha
parameter for the gamma model), and different among-branch rate drifts
(violation of the clock). See @Shi2018 for examples of such simulations.
The simulation option also includes a continuous migration model, with a
user-specified migration rate matrix between species/populations.</p>
<p>Look at the <code>README.txt</code> file for compiling the program. To run the
program, type one of the following. The default control file name is
<code>MCcoal.ctl</code>. Always look at the screen output to confirm that the
program reads the control file correctly.</p>
<pre><code>    bpp --simulate MCcoal.ctl
    bpp --simulate MCcoalMigration.ctl
</code></pre>
<h4 id="b1-control-file-options-for-simulating-without-migration">B1 Control-file options for simulating without migration</h4>
<p>Here is a sample control file.</p>
<pre><code>    seed = 12345
    seqfile = MySeq.txt 0 * comment out this line if you don't want seqs
    treefile = MyTree.tre  * comment out this line if you don't want trees
    Imapfile = MyImap.txt
    *   concatfile = concatenatedfile.txt  * concatenated alignment
    modelparafile = modelparas.txt   * comment out this line if you don't want seqs

    species&amp;tree = 4  A  B  C  D
    3  2  1  1
    ((A #0.01, B #0.01) :0.01 #.01, (C, D) :0.011 #0.01) : 0.012 #0.01;
    phase =    1  1  1  0   * 0: do not phase (fully phased seqs), 1: diploid unphased seqs

    loci&amp;length = 100 1000 * number of loci &amp; number of sites at each locus
    *    locusrate = 0 (default: same rate for all loci)
    *    locusrate = mu_bar a_mui prior
    *        clock = 1 (default: strict clock)
    *        clock = 2 v_bar a_vi prior dist  (independent-rates)
    *        clock = 3 v_bar a_vi prior dist  (correlated-rates)

    model = 7       * model: 0:JC69, 7:REV (GTR)
    Qrates = 0  10 5 5 5 5 10     * 1: fixed; 0: dirichlet, for TC TA TG CA CG AG
    basefreqs = 0  10  10  10  10   * 0: random, Dirichlet(aT,aC,aA,aG), for base frequencies
    alpha_siterate = 0  100  20  5  * G(a, b) for alpha for sites &amp; K for discrete gamma
</code></pre>
<p><code>seed</code> Use <span class="arithmatex">\(-1\)</span> to simulate different datasets every time the program is
run. If you use a positive integer the data will stay exactly the same,
which is a bad idea.</p>
<p><code>seqfile, treefile, Imapfile, concatfile, modelparafile.</code> These are
names of files to be generated. If you want to simulate gene trees and
not sequence files, you can comment out the line for the seqfile.
concatfile will have the sequence alignment concatenated across loci.
The sequence files can become very large. modelparafile records the
parameter values for the GTR model, as well as the locus rate and the
rates for nodes on the species tree. There is one line of output per
locus.</p>
<p>By default MCcoal prints out the sequence alignment at each locus, but
if you use file format 1 (<code>seqfile = MySeq.txt 1</code>), the program will
print out site pattern counts instead. The file may then be smaller.
This format is readable by BPP and [3s]{.smallcaps}, but
not by other programs.</p>
<p><code>phase.</code> This variable is used in the same way as in the inference (data
analysis) mode. Each species is assigned a switch/flag, with 0 meaning
fully phased (haplotype) sequences and 1 for unphased diploid sequences.
With the specifications in the example (sampling configuration 3 2 1 1
and phase = 1 1 1 0), the program will simulate 6, 4, 2, 1 sequences for
A, B, C, D, respectively, and then combine every two sequences from A,
B, C into one single diploid sequence with heterozygote sites
represented using ambiguity characters (YRMKSW). If you use the same
random number seed, the diploid data generated using the sampling
configuration (3 2 1 1) with phase = 1 1 1 0 should match the fully
resolved sequences generated using the sampling configuration (6 4 2 1)
with phase = 0 0 0 0. The program also prints out the fully resolved
sequences in a file with the suffix <code>_full</code> (or <code>MySeq_full.txt</code> in the
example).</p>
<p><code>species&amp;tree.</code> This block specifies the number and names of species,
the species tree, and the parameters under the multispecies coalescent
model, including the species divergence times (<span class="arithmatex">\(\tau\)</span>s) and population
size parameters (<span class="arithmatex">\(\theta = 4N\mu\)</span>). For the example above, 100 loci,
each of 1000 sites, will be simulated, on the species tree ((A, B), (C,
D)), with 3, 2, 1, 1 sequences for A, B, C, D, so that there are 7
sequences at each locus. The divergence time parameters (<span class="arithmatex">\(\tau\)</span>s) are
after ':' in the tree, while the population size parameters (<span class="arithmatex">\(\theta\)</span>s)
are after '#'. Thus we have 0.01 for all <span class="arithmatex">\(\theta\)</span>s, and
<span class="arithmatex">\(\tau_{ABCD} = 0.012\)</span>, <span class="arithmatex">\(\tau_{AB} = 0.01\)</span>, and <span class="arithmatex">\(\tau_{CD} = 0.011\)</span>. We
need <span class="arithmatex">\(\theta_A\)</span> and <span class="arithmatex">\(\theta_B\)</span> because 2 or more sequences are sampled
from species A and B. Parameters <span class="arithmatex">\(\theta_C\)</span> and <span class="arithmatex">\(\theta_D\)</span> are
unnecessary in this case as only one sequence is sampled from C and D:
if you specify them, they will be ignored by the program. Note that both
<span class="arithmatex">\(\theta\)</span>s and <span class="arithmatex">\(\tau\)</span>s are measured by the expected number of mutations
per site, so that <span class="arithmatex">\(\theta = 0.01\)</span> means that two sequences sampled from
the population are 1% different. For reference, the estimate for the
human species is <span class="arithmatex">\(\theta_H = 0.0006\)</span>.</p>
<p><code>model.</code> This can take two possible values: 0 for JC69 and 7 for GTR.
The next few lines are relevant for the GTR model.</p>
<p><code>Qrates</code> specifies the exchangeability parameters in the GTR model
(<span class="arithmatex">\(a, b, c, d, e, f\)</span> for TC, TA, TG, CA, CG, and AG, in @Yang1994a).
There are two options, either to have those rates fixed for all loci or
to have them sampled at random. The first has the following format (note
that the first 0-1 number is a switch):</p>
<pre><code>Qrates = 1  2 1 1 1 1 2   *  1: fixed; for TC TA TG CA CG AG
</code></pre>
<p>The input is one integer value (0 or 1) which acts as a switch followed
by six real numbers. The line above fixes the rates at
<span class="arithmatex">\(a = 2, b = c = d = e = 1\)</span> and <span class="arithmatex">\(f = 2\)</span>, so that the transition rate is
twice as high as the transversion rate, and the model corresponds to K80
or HKY. Note that only the relative rates matter, because the rate
matrix (<span class="arithmatex">\(Q\)</span>) is scaled so that the average rate (over the base
frequencies) is 1 and branch lengths are measured in the expected number
of mutations/substitutions per site. Nevertheless the program expects
six rate parameters in the input here.</p>
<pre><code>Qrates = 0  10 5 5 5 5 10   *  0: dirichlet, for TC TA TG CA CG AG
</code></pre>
<p>The second option, above, is to sample <span class="arithmatex">\(a, b, c, d, e\)</span> from the
Dirichlet distribution for every locus. The above specifies Dir(10, 5,
5, 5, 5, 10), with parameters
<span class="arithmatex">\(\alpha_a = 10, \alpha_b = \alpha_c = \alpha_d = \alpha_e = 10\)</span>, and
<span class="arithmatex">\(\alpha_f = 10\)</span>. The rates generated from the Dirichlet sum to 1, and
they are rescaled. Note that larger values for those parameters mean
less variance, while the mean for <span class="arithmatex">\(a\)</span>, say, is given by
<span class="arithmatex">\(\alpha_a/\alpha\)</span> with
<span class="arithmatex">\(\alpha = \alpha_a + \alpha_b + \alpha_c + \alpha_d + \alpha_e +
\alpha_f\)</span>. The specification here gives on average a
transition/transversion rate ratio of 2 (if the base frequencies are
fixed at <span class="arithmatex">\(\frac{1}{4}\)</span> each), but it varies among loci.</p>
<p><code>basefreqs.</code> The base frequency parameters
(<span class="arithmatex">\(\pi_T, \pi_C, \pi_A, \pi_G\)</span>) can also be fixed or sampled from the
Dirichlet distribution, like the <code>Qrates</code>. The two options are as
follows.</p>
<pre><code>    basefreqs = 1 0.15 0.35 0.15 0.35  * 1: fixed; Base frequencies are for TCAG.
    basefreqs = 0  10  10 10  10       * 0: random, Dirichlet(aT, aC, aA, aG), for base frequencies
</code></pre>
<p><code>alpha_siterate.</code> The gamma shape parameter (<span class="arithmatex">\(\alpha\)</span>) for rate
variation among sites of the same locus can be fixed or sampled from a
gamma distribution. The possible options are as follows. When alpha is
fixed, the same value is used for all loci. Otherwise different
<span class="arithmatex">\(\alpha\)</span>s are used for different loci. Note that given alpha, the rates
for sites at the same locus have a G((<span class="arithmatex">\(\alpha, \alpha\)</span>) distribution
with mean 1 [@Yang1993; @Yang1994]. The fourth option below will allow
the program to sample <span class="arithmatex">\(\alpha\)</span> from <span class="arithmatex">\(G(100, 20)\)</span>, with mean 5, for each
locus and then use <span class="arithmatex">\(G(\alpha, \alpha)\)</span> to describe the among-site rate
variation for the locus.</p>
<pre><code>    alpha_siterate = 1 0            * 1: alpha fixed at 0(inf), one rate for all sites at the locus
    alpha_siterate = 1  5.6 0       * 1: alpha fixed at 5.6, K = 0(inf) for continuous gamma
    alpha_siterate = 1  5.6 5       * 1: alpha fixed at 5.6, K = 5 categories for discrete gamma
    alpha_siterate = 0  100  20  5  * 0: alpha sampled from G(100, 20), with mean 5, K = 5 for discrete gamma
</code></pre>
<p><code>locusrate.</code> The variables <code>locusrate</code> and <code>clock</code> are used to specify
relaxed-clock models, with rate variation among branches and among loci.
These are the same models as implemented in the inference program.
Please see notes in section ?? for details.</p>
<p><code>locusrate</code> specifies variable rates for different loci. The default is
0, meaning that all loci have the same rate. To specify variable rates
for loci, use the following format.</p>
<pre><code>    locusrate = 0 (default)
    locusrate = mu_bar a_mui prior
    locusrate = mu_bar a_mui
    locusrate = 1.0 5.0 dir
</code></pre>
<p>Here <code>mu_bar</code> is the mean rate across loci.</p>
<p>If <code>prior = dir</code> (gamma-Dirichlet), the total rate (<span class="arithmatex">\(L\bar\mu\)</span>) is
partitioned to generate the locus-rates (<span class="arithmatex">\(\mu_i\)</span>) according to the
Dirichlet distribution with concentration parameter <span class="arithmatex">\(\alpha\)</span> (<code>a_mui</code>).
Large <span class="arithmatex">\(\alpha\)</span> (10 or 100, say) means that the rates are similar among
loci, while small values (1 or 0.5, say) mean that rates are highly
variable among loci. If <code>prior = iid</code> (conditional i.i.d.), the locus
rates (<span class="arithmatex">\(\mu_i\)</span>) are i.i.d. given the mean rate (<span class="arithmatex">\(\bar\mu\)</span>):</p>
<pre><code>    mu_i ~ Gamma(a_mui, a_mui/mubar)
</code></pre>
<p><code>clock.</code> This is used to specify strict-clock or relaxed-clock models.
The default is <code>clock = 1</code> for strict clock, while <code>clock = 2</code> is the
independents-rates model, and <code>clock = 3</code> is the correlated-rates model
(which is implemented for the MSC model with no introgression and is
unavailable for the MSC-I model).</p>
<pre><code>    clock = 1 (default: strict clock)
    clock = 2 v_bar a_vi prior dist  * independent-rates model
    clock = 2 0.1 10 dir G    * gamma-dirichlet for loci &amp; gamma for branches
    clock = 2 0.1 10 iid LN   * iid for loci &amp; log-normal for branches
    clock = 3 v_bar a_vi prior dist  * correlated-rates model
    clock = 3 0.1 10 iid G    * iid for loci &amp; gamma for branches
    clock = 3 0.1 10 iid LN   * iid for loci &amp; log-normal for branches
</code></pre>
<p>The specification (<code>clock = 2 0.1 10 dir G</code>) means the following. First
the average variance parameter is = 0.1. Then the total <span class="arithmatex">\(L\bar\mu\)</span> is
partitioned into <span class="arithmatex">\(\mu_i\)</span> (for <span class="arithmatex">\(i = 1, 2, \cdots, L\)</span>) according to the
Dirichlet distribution with concentration parameter <span class="arithmatex">\(\alpha = 10\)</span>
(<code>a_vi</code>).</p>
<p>In contrast, the specification (<code>clock = 2 0.1 10 iid LN</code>) means that
the variance parameter for locus <span class="arithmatex">\(\nu_i\)</span> is generated as
<span class="arithmatex">\(\nu_i|\bar\nu \sim G(\alpha, \alpha/\bar\nu)\)</span> with shape parameter
<span class="arithmatex">\(\alpha = 10\)</span> (<code>a_vi</code>) and mean <span class="arithmatex">\(\bar\nu = 0.1\)</span>.</p>
<p>In both prior models (<code>prior=dir</code> and <code>iid</code>), <span class="arithmatex">\(\alpha\)</span> (<code>a_vi</code>) is
inversely related to the variance in <span class="arithmatex">\(\mu_i\)</span> among loci: use small
values of <span class="arithmatex">\(\alpha\)</span> (2, 1, or 0.5) if the clock nearly holds at some loci
but is seriously violated at others, and large values (such as 10 or
100) if the clock is violated in the same extent at different loci.</p>
<p><code>clock = 2.</code> Given the locus rate <span class="arithmatex">\(\mu_i\)</span> (specified using the
<code>locusrate</code> variable) and the variance parameter <span class="arithmatex">\(\nu_i\)</span> (specified
using the <code>clock</code> variable) for locus <span class="arithmatex">\(i\)</span>, the rates for species-tree
branches are specified as follows. If <code>dist = G</code> (for gamma), the rate
for branch <span class="arithmatex">\(j\)</span> at locus <span class="arithmatex">\(i\)</span> has the following gamma distribution with
mean <span class="arithmatex">\(\mu_i\)</span> and variance <span class="arithmatex">\(\nu_i\)</span>
<span class="arithmatex">\(<span class="arithmatex">\(r_{ij} | \mu_i, \nu_i \sim G(\mu_i^2/\nu_i, \mu_i/\nu_i).\)</span>\)</span></p>
<p>If <code>dist = LN</code> (for log-normal), the rate has the log-normal
distribution <span class="arithmatex">\(<span class="arithmatex">\(r_{ij} | \mu_i, \nu_i \sim LN(\mu_i, \nu_i).\)</span>\)</span></p>
<p><code>clock = 3.</code> The correlated-rates model is specified similarly, and for
the MSC model only. Given the locus rate <span class="arithmatex">\(\mu_i\)</span> (specified using the
<code>locusrate</code> variable) and the variance parameter <span class="arithmatex">\(\nu_i\)</span> (specified
using the <code>clock</code> variable) for locus <span class="arithmatex">\(i\)</span>, the rates for species-tree
branches are specified recursively, starting from the root towards the
tips.</p>
<p>If <code>dist = G</code> (for gamma), the rate for branch <span class="arithmatex">\(j\)</span> at locus <span class="arithmatex">\(i\)</span> has the
following gamma distribution with shape parameter
<span class="arithmatex">\(\alpha = \mu_i^2/\nu_i\)</span> and with the mean to be the rate for the
parental branch of <span class="arithmatex">\(j\)</span> or <span class="arithmatex">\(r_{i,p(j)}\)</span>:
<span class="arithmatex">\(<span class="arithmatex">\(r_{ij} | \mu_i, \nu_i, r_{i,p(j)} \sim G(\frac{\mu_i^2}{\nu_i}, \frac{\mu_i^2}{\nu_i r_{i,p(j)}}).\)</span>\)</span>
If <code>dist = LN</code> (for log-normal), the rate for branch <span class="arithmatex">\(j\)</span> at locus <span class="arithmatex">\(i\)</span>
has the log-normal distribution with the mean to be the parental rate
and variance parameter <span class="arithmatex">\(\nu_i\)</span>:
<span class="arithmatex">\(<span class="arithmatex">\(r_{ij} | \nu_i, r_{i,p(j)} \sim LN(r_{i,p(j)}, \nu_i).\)</span>\)</span></p>
<p>After the rates for branches (populations) on the species tree are
generated for the locus, they are used to calculate the gene-tree branch
lengths. Gene-tree branches residing in the same species/populations
have the same rate (the rate of that species at the locus), while those
residing in different species may have different rates. A branch on a
gene tree may have segments residing in different species/populations,
and the length of the branch (measured by the expected number of
mutations/substitutions per site) is calculated by adding up those
segments.</p>
<h4 id="b2-simulating-with-migration-under-the-im-model">B2 Simulating with migration (under the IM model)</h4>
<pre><code>    migration = 7   * number of pops (order fixed by program)

    A     B     C     D    ABCD    AB   CD
    A      0     1.1   1.2   1.3    0     0    -1
    B      0.1   0     1.4   1.5    0     0    -1
    C      0.2   0.4   0     1.6    0     1.7   0
    D      0.3   0.5   0.6   0      0     1.8   0
    ABCD   0     0     0     0      0     0     0
    AB     0     0     0.7   0.8    0     0     1.9
    CD    -1    -1     0     0      0     0.9   0
</code></pre>
<p>The control file for simulating migration as well as coalescence
includes a block as above (see the file <code>MCcoalMigration.ctl</code>). The line
<code>migration = 7</code>, where 7 is the number of populations on the species
tree, tells the program to simulate migrations. This number is fixed by
the species tree and is used here for error checking. This is then
followed by a migration matrix, of size <span class="arithmatex">\(7\times 7\)</span>. The names of the
populations are read and ignored by the program, and the order of the
populations is fixed. You should run the program without the migration
matrix first and then use the screen output to use the correct order of
populations to specify the migration matrix.</p>
<p>In the migration matrix, values 0 and <span class="arithmatex">\(-1\)</span> mean that migration is either
impossible (for example, if the two populations did not live at the same
time) or not allowed (if the two populations were contemporary but no
migration between them is assumed to occur). Use 0 for both cases here.
Positive values are scaled migration rates, with the element on the
<span class="arithmatex">\(i\)</span>th row <span class="arithmatex">\(j\)</span>th column to be <span class="arithmatex">\(M_{ij} = N_j m_{ij}\)</span>, where <span class="arithmatex">\(m_{ij}\)</span> is
the migration rate per generation in population <span class="arithmatex">\(j\)</span> from population <span class="arithmatex">\(i\)</span>,
or the proportion of individuals in population <span class="arithmatex">\(j\)</span> that are immigrants
from population <span class="arithmatex">\(i\)</span>, and where <span class="arithmatex">\(\mu\)</span> is the mutation rate per site per
generation. In the example above, <span class="arithmatex">\(M_{A\to B} = 1.1\)</span>, which means that
on average 1.1 individuals are immigrants from population <span class="arithmatex">\(A\)</span> to
population <span class="arithmatex">\(B\)</span>.</p>
<p>Note that the <span class="arithmatex">\(\theta\)</span> parameter for a modern species has to be
specified in the tree file even if only one sequence is sampled from
that species but migrations into that species are allowed by the
migration matrix.</p>
<h3 id="features-new-to-bpp-40">Features New To BPP 4.0</h3>
<ul>
<li>
<p>i. Multiple threads</p>
</li>
<li>
<p>ii. Introgression model (MSC-I)</p>
</li>
<li>
<p>iii. Local-clock and locus-rates models (clock and locusrate)</p>
</li>
<li>
<p>iv. Topological constraints (constraint and outgroup)</p>
</li>
</ul>





                
              </article>
            </div>
          
          
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    <script id="__config" type="application/json">{"base": "..", "features": ["content.code.copy"], "search": "../assets/javascripts/workers/search.74e28a9f.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../assets/javascripts/bundle.220ee61c.min.js"></script>
      
        
          <script src="../mathjax.js"></script>
        
      
        
          <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
        
      
        
          <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
        
      
    
  </body>
</html>